---
title: "HPA-gene-clustering"
author: "Maria Bueno Alvez"
date: "2/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r load packages and functions, include=FALSE}

# Packages
library(tidyverse)
library(magrittr)
library(broom)
library(readxl)
library(NOISeq)
library(uwot)
library(MASS)
library(viridis)
library(pcaMethods)
library(factoextra)
library(cluster)
library(kohonen)
# library(ClusterR)
# library(fclust)
library(flashClust)
library(Seurat)
# library(dbscan)
# library(igraph)
library(kmed)
library(aricode)
library(pheatmap)
library(clValid)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(annotate)
library(GO.db)
library(ggthemes)
library(ggrepel)
library(readr)
library(clusterProfiler)
library(ClusterJudge)
library(mixOmics)
library(ggsci)
library(scales)

map <- purrr::map
pca <- pcaMethods::pca
#concaveman, ggalt, sf

# Wd, scripts
setwd("/Users/maria.bueno/Desktop/HPA-gene-clustering")
source("scripts/functions_preprocessing.R")
source("scripts/functions_dimensionalityreduction.R")
source("scripts/functions_clustering.R")
source("scripts/functions_evaluation.R")

map <- purrr::map
select <- dplyr::select
pca <- pcaMethods::pca
rank <- base::rank

```

```{r load data, include=FALSE}
# Data
data <- 
  read_delim("data/pig_filtered_sample_data.tab", delim = "\t")

gene_info <-
  read_delim("data/meta/pig_consensus_category_piggenome_92.tsv", delim = "\t")

genes_to_filter <-
  gene_info %>%
  filter(distribution_category == "not detected") %>%
  pull(ensg_id) 

tissue_info <-
  gene_info %>%
  filter(!ensg_id %in% genes_to_filter) %>%
  separate_rows("enhanced_tissues", sep = ',') %>%
  select(ensg_id,enhanced_tissues) %>%
  remove_missing()
  
# Color palette
colors <-
  read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors <- colors  %>%
  select(organ_color, consensus_tissue_name) %>%
  distinct() %$%
  set_names(organ_color, consensus_tissue_name)

# Information of orthologs
ortholog_info <- 
  read_delim("data/meta/gene_orthologs.csv", delim = ",") %>%
  select(enssscg_id, ensg_id) %>%
  dplyr::mutate(entrez = mapIds(org.Hs.eg.db,
                  keys= ensg_id, 
                  column="ENTREZID",
                  keytype="ENSEMBL"))
```

```{r plot theme}

```


#1- Data pre-processing

```{r pre-procesing}
filter <- dplyr::filter
# Filtering genes
filtered_data <-
  data %>% 
  filter(!enssscg_id %in% genes_to_filter)

# Save gene names in order
gene_names <- 
  filtered_data %>%
  select(enssscg_id, sample_ID, tpm) %>%
  spread(key = sample_ID , value = tpm) %$%
  enssscg_id

#save(gene_names,
 #    file = "data/processed/gene_names.Rdata")

# Normalize
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


tmm_data %>%
  spread(key= sample_ID, value = tmm) %>% 
  column_to_rownames("enssscg_id") %>% 
  pheatmap(show_colnames = F, show_rownames = F,
           cluster_rows = F, cluster_cols = F,
           color = RColorBrewer::brewer.pal(7, "YlGnBu"))
# Scale data
scaled <- list("zscore","max") %>% #"min-max",
  map(~data_scaling(df = tmm_data, col_value = "tmm",
                    col_gene = "enssscg_id", col_sample = "sample_ID",
                    log_transform = F, m = .x))

scaled[[2]]$scaled %>% 
  column_to_rownames("enssscg_id") %>% 
  pheatmap(show_colnames = F, show_rownames = F,
           cluster_rows = F, cluster_cols = F,
           color = RColorBrewer::brewer.pal(7, "YlGnBu"))


```

#2- Dimensionality reduction

```{r dimensionality reduction}

#load("data/processed/scaled_data.Rdata")

pcas <- scaled %>%
  map(~pca_calc(data = .x$scaled, npcs = 200, m = .x$method))

pcas[[2]]$scores %>%
  as_tibble() %>% 
  dplyr::select(c(1:30)) %>% 
  pheatmap(show_colnames = F, show_rownames = F,
           cluster_rows = F, cluster_cols = F,
           color = RColorBrewer::brewer.pal(7, "YlGnBu"))
# save(pcas,
#      file = "data/processed/pcas.Rdata")

```

#3- Distance calculation

```{r distance calculation}

#All distances
distances <- lapply(list("euclidean","pearson"), function(i) #"manhattan"
  map(pcas, ~dist_calc(df = .x$scores, comp = 30, m = i, 
                       id = .x$method)))

#save(distances,
 #    file = "data/processed/all_distances.Rdata")
```

#4- Clustering

```{r Clustering analysis}
#load("data/processed/distances_4.Rdata")
#load("data/processed/gene_names.Rdata")

#d <- list(distances[[1]][[1]],distances[[1]][[2]],distances[[2]][[1]],distances[[2]][[2]])

# Run with 1 k 
res_kmeans <- 
  distances %>%
  map(~clust(dist = .x$distance, 
             k = 100, 
             m = "fastkmeans", 
             genes = gene_names, 
             id = .x$id))

# Run multi k
hclust_multk <- 
  lapply(list(50,70,90,110,130,150), function(i)
    map(d, ~clust(dist = .x$distance, 
                  k = as.numeric(i), 
                  m = "fasthclust", 
                  genes = gene_names, 
                  id = .x$id, 
                  mult_k = T)))

# Run with 1 r 
res_louvain <- 
  distances %>%
  map(~clust(dist = .x$distance, 
             r = 4, 
             m = "louvain", 
             genes = gene_names, 
             id = .x$id))

# Run multi r
res_louvain_multk <- 
  lapply(list(4,6,8,10,12,14), function(i)
    map(d, ~clust(dist = .x$distance, 
                  r = as.numeric(i), 
                  m = "louvain", 
                  genes = gene_names, 
                  id = .x$id)))

#save(hclust_multk,
 #    file = "data/processed/hclust_multk.Rdata")

```


#5- Evaluation

```{r Evaluation (connectivity, dunn, asw, BHI, BP_enriched, time)}
load("data/processed/distances_4.Rdata")
load("data/processed/leiden_multk.Rdata")


eval_somt_multk <- lapply(list(1,2,3,4,5,6), function(i) 
  map(som_multk[[i]], 
      ~eval(clustering = .x, 
            genes= ortholog_info,
            dis = d
      )))

#save(kmeans_multk, file = "data/processed/kmeans_multk.Rdata")
# load("data/processed/eval_som_multk.Rdata")
# load("data/processed/som_multk.Rdata")

eval_som_df <- eval_to_df(eval_somt_multk)


all_som <- add_clusterings(df = eval_som_df$res, 
                        dist = d, 
                        pca = pcas[[1]],
                        ref_clust = som_multk[[1]][[1]]$cluster,
                        gene_names = gene_names, 
                        clust_m = "SOM",
                        orthologs = ortholog_info)

# bestk_leiden <- find_bestk(all_leiden)
# 
# save(eval_leiden_df,all_leiden,bestk_leiden, file = "data/processed/evaluation_leiden.Rdata")

# df with all evaluations
load("data/processed/evaluation_kmeans.Rdata")
load("data/processed/evaluation_medoids.Rdata")
load("data/processed/evaluation_hclust.Rdata")
load("data/processed/evaluation_louvain.Rdata")
load("data/processed/evaluation_leiden.Rdata")
load("data/processed/evaluation_som.Rdata")


all_eval <- 
  bind_rows(all_kmeans,
            all_medoids,
            all_hclust,
            all_louvain,
            all_leiden,
            all_som)





eval_to_plot <- 
  all_eval %>%
  filter(!grepl("2D",clustering_id)) %>%
  group_by(clustering_id) %>%
  rename(id = clustering_id) %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))



test <- 
  eval_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    #m <- unique(data$method)
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  })


test %>%
  select(id,clustering,k, BHI_index) %>%
  ggplot(aes(x=id, y = BHI_index, fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~(clustering), scale="free") +
 scale_fill_viridis(discrete = F)+
  labs(y = "Value", x = "Clustering strategy")


```

```{r Run enrichments}
load("data/processed/som_multk.Rdata")


to_df <- function(multk, add_random = T) {
  lapply(c(1:4),
         function(i) 
           
           if(add_random == T) {
             random <- as.data.frame(multk[[1]][[i]]$cluster) %>%
               mutate(value = sample(value))
             df <- random %>%
               mutate(id = 
                        paste((gsub('[[:digit:]]+', '', 
                                    multk[[1]][[i]]$id)),"random", 
                              sep = ""))
             
             for (j in c(1:6)) {
               df <- bind_rows(df, as.data.frame(multk[[j]][[i]]$cluster) %>% 
                                 mutate(id = multk[[j]][[i]]$id))
             }
             return(df)
           }
         
         else {
           df <- data.frame()
           for (j in c(1:6)) {
             df <- bind_rows(df, as.data.frame(multk[[j]][[i]]$cluster) %>% 
                               mutate(id = multk[[j]][[i]]$id))
           }
           return(df)
         }
         
  )
}

enrich <- function (clustering,ontology_type="BP",orthologs, BP_index=F) {
  clustering <-
    clustering %>%
    left_join(orthologs, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(entrez,value)
  
  cluster_list <- c()
  num_clus <- tail(sort(clustering$value), n=1)
  num_genes <- length(clustering$value)
  for (i in c(0:num_clus)) {
    genes <- clustering %>%
      filter(value==i) %>%
      pull(entrez) %>%
      as.character()
    cluster_list[[as.character(i)]] <- genes
  }
  
  # Profile clusters (with universe)
  ck <- try(compareCluster(geneClusters = cluster_list, 
                           fun = "enrichGO", 
                           OrgDb = org.Hs.eg.db, 
                           ont = ontology_type, 
                           universe = as.character(clustering$entrez),
                           pAdjustMethod = "BH",
                           pvalueCutoff = 1, 
                           qvalueCutoff = 1))
  
  ck_tibble <- as_tibble(ck) 
  
  if(BP_index == T) {
    bio_index <-
      length(unique(as.numeric(ck_tibble$Cluster)))/(num_clus)
    
    return(list(enriched_terms = ck_tibble,
                enrichment_score = bio_index))
  }
  
  else {
    return(ck_tibble)
  }
  
}
enrichments_df <- function(df, grouping, orthologs) {
  
  res <- df %>%
    group_by(.data[[grouping]]) %>%
    do({
      subset <- .
      
      enrichment <-
        enrich(clustering = subset, orthologs = orthologs) %>%
        mutate(id = unique(subset$id))
    })
  
}



df_som <- 
  to_df(som_multk, add_random = T)

Sys.time() #5.37
enrichments_som <- df_som %>%
  map(~enrichments_df(df = .x, grouping = "id", orthologs = ortholog_info))

save(enrichments_som, file="data/processed/enrichments_som.Rdata")

load("data/processed/enrichments_louvain.Rdata")


enrichment_scores <- function(df,grouping) {
  
  df %>%
    group_by(.data[[grouping]]) %>%
    do ({
      data <- .
      min_significance <- data %>%
        mutate(log_qvalue = -log10(qvalue)) %>%
                                group_by(Cluster) %>% 
                                summarize(max(log_qvalue)) %>% 
                                pull(`max(log_qvalue)`) %>% mean() 
      annotability <- length(data %>%
               filter(qvalue < 0.05) %>% pull(Cluster) %>% sort() %>% unique())/
                           length(data %>% pull(Cluster) %>% sort() %>% unique())
      data.frame(id = unique(data$id), min_significance = min_significance, annotability = annotability)
        
    })
}

# to_plot_kmeans <- 
#   enrichment_scores(df = enrichments_kmeans[[1]], grouping = "id") %>%
#   gather(key = "measure", value = "value", -id) 
# 
# 
# to_plot_kmeans$id <- factor(to_plot_kmeans$id,
#                                             levels = c("zscore euclidean fastkmeans 50","zscore euclidean fastkmeans 70",
#                                                        "zscore euclidean fastkmeans 90","zscore euclidean fastkmeans 110",
#                                                        "zscore euclidean fastkmeans 130","zscore euclidean fastkmeans 150",
#                                                        "zscore euclidean fastkmeans random")
#                             )
# 
# to_plot_kmeans %>%
#   ggplot(aes(x= id, y = value, fill = id)) +
#   geom_bar(stat = "Identity") +
#   theme_minimal() +
#   facet_wrap(~measure, scales="free_y") +
#   scale_fill_viridis(discrete = T) +
#   theme(text = element_text(size=15)) + 
#   labs(y = "Value", x = "Clustering") +
#       theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
  
  
  
  plot_res <- function(x,grouping) {
  x %>%
    group_by(.data[[grouping]]) %>%
    do ({
      data <- .
      curr_id <- strsplit(data$id, " ")
      title <- paste(curr_id[[1]][1],curr_id[[1]][2],curr_id[[1]][3])

      data$new_id <- curr_id[[1]][4]
            data$group <- title

      data
    })

}

# All 
to_plot_louvain <-
  map(enrichments_louvain,
  ~enrichment_scores(.x, grouping = "id"))
  
list_to_plot <- 
  map(to_plot_louvain,
  ~.x %>% 
  plot_res(grouping = "id")) 
  

df_to_plot <- data.frame()
for(i in c(1:4)){
  df_to_plot <- bind_rows(df_to_plot, list_to_plot[[i]])
}
  
df_to_plot$new_id <-  factor(df_to_plot$new_id,levels = c("50","70","90","110","130","150","random"))
df_to_plot$new_id <-  factor(df_to_plot$new_id,levels = c("69","87","104","118","127","138","random"))

df_to_plot%>%
  ggplot(aes(x= new_id, y = value, fill = new_id)) +
  geom_bar(stat = "Identity") +
  theme_minimal() +
  facet_wrap(group~measure, scales="free_y", ncol = 2) +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=10)) + 
  labs(y = "Value", x = "Clustering") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  
}
```

```{r Evaluations based on enrichments (clusterProfiler)}

# Calculate enrichments

load("data/processed/louvain_multk.Rdata")

df_kmeans <- 
  to_df(kmeans_multk, add_random = T)

Sys.time()
enrichments_kmeans <- df_kmeans %>%
  map(~enrichments_df(df = .x, grouping = "id", orthologs = ortholog_info))


# Calculate enrichment scores

load("data/processed/enrichments_kmeans.Rdata")
load("data/processed/enrichments_medoids.Rdata")
load("data/processed/enrichments_hclust.Rdata")
load("data/processed/enrichments_louvain.Rdata")
load("data/processed/enrichments_leiden.Rdata")
load("data/processed/enrichments_som.Rdata")

enrichment_scores <- function(df,grouping) {
  
  df %>%
    group_by(.data[[grouping]]) %>%
    do ({
      data <- .
      min_significance <- data %>%
        mutate(log_qvalue = -log10(qvalue)) %>%
                                group_by(Cluster) %>% 
                                summarize(max(log_qvalue),na.rm = T) %>% 
                                pull(`max(log_qvalue)`) %>% na.omit() %>% mean() 
      annotability <- length(data %>%
               filter(qvalue < 0.05) %>% pull(Cluster) %>% sort() %>% unique())/
                           length(data %>% pull(Cluster) %>% sort() %>% unique())
      data.frame(id = unique(data$id), min_significance = min_significance, annotability = annotability)
        
    })
}

enrichment_scores_kmeans <-
  map(enrichments_kmeans,
  ~enrichment_scores(df = .x, grouping = "id")) #%>%
 # gather(key = "measure", value = "value", -id))
 

# df with all enrichment scores
load("data/processed/enrichment_scores_kmeans.Rdata")
load("data/processed/enrichment_scores_kmeans.Rdata")

load("data/processed/enrichment_scores_kmeans.Rdata")

all_bio <- bind_rows(c(enrichment_scores_kmeans, enrichment_scores_medoids, enrichment_scores_hclust, 
                      enrichment_scores_louvain, enrichment_scores_leiden, enrichment_scores_som))


save(all_bio, file="data/processed/all_bio.Rdata")
bio_to_plot <- 
  all_bio %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))



test <- 
  bio_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    #m <- unique(data$method)
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  })


test %>%
  select(-annotability) %>%
  ggplot(aes(x=id, y = min_significance, fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~(clustering), scale="free") +
 scale_fill_viridis(discrete = F)+
  labs(y = "Value", x = "Clustering strategy")












######
enrichment_scores_kmeans <- bind_rows(enrichment_scores_kmeans) %>%
  mutate(title = strsplit(id, " ")[[1]][3],
         subtitle  =paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2]),
           k = strsplit(id, " ")[[1]][4])


l <- sort(unique(as.numeric(enrichment_scores_kmeans$k)))
l[7] <- "random"

enrichment_scores_kmeans$k <- factor(enrichment_scores_kmeans$k,
                                            levels = l)

enrichment_scores_kmeans %>%
  ggplot(aes(x= k, y = value, fill = k)) +
  geom_bar(stat = "Identity") +
  theme_minimal() +
  facet_wrap(subtitle~measure, scales="free_y", ncol = 2) +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=15)) + 
  labs(y = "Value", x = "Clustering") +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
  
  
save(enrichment_scores_kmeans, file= "enrichment_scores_kmeans.Rdata")




# Test alternatives for biological index
clustering <- louvain_multk[[1]][[1]]$cluster

clustering_random <-
  clustering %>%
  mutate(value = sample(value))

louvain_k <- list(louvain_multk[[1]][[1]], louvain_multk[[2]][[1]], louvain_multk[[3]][[1]],
               louvain_multk[[4]][[1]], louvain_multk[[5]][[1]], louvain_multk[[6]][[1]])

enrichments <- map(louvain_k, 
                   ~enrich(clustering = .x$cluster, orthologs = ortholog_info))

enrichment_random <- enrich(clustering = clustering_random, orthologs = ortholog_info)
# 
# enrichment_random$enriched_terms %>%
#   ggplot(aes(x=p.adjust)) +
#   geom_histogram()
#   geom_bar(stat = "identity")

enrichments_louvain <-
  map(enrichments,
      ~.x$enriched_terms)

enrichments_louvain[[7]] <- enrichment_random$enriched_terms



# ennrichments <- function(clusterings,orthologs) {
#   
#   clustering_random <-
#     clusterings[[1]][[1]]$cluster %>%
#     mutate(value = sample(value))
#   
#   clusterings_k <- list(clusterings[[1]][[1]], clusterings[[2]][[1]], clusterings[[3]][[1]],
#                     clusterings[[4]][[1]], clusterings[[5]][[1]], clusterings[[6]][[1]])
#   
#   enrichments <- map(clusterings_k, 
#                      ~enrich(clustering = .x$cluster, orthologs = ortholog_info))
#   
#   enrichment_random <- enrich(clustering = clustering_random, orthologs = ortholog_info)
# 
#   enrichment_terms <-
#     map(enrichments,
#         ~.x$enriched_terms)
#   
#   enrichment_terms[[7]] <- enrichment_random$enriched_terms
#   
#   return(enrichment_terms)
# 
# }



# Calculate scores
significance_overall <- map(enrichments_louvain, ~.x  %>% pull(qvalue) %>% mean())
significance_filtered <- map(enrichments_louvain, ~.x %>% filter(qvalue < 0.05) %>% pull(qvalue) %>% mean())
significance_overall_perclust <- map(enrichments_louvain, ~.x  %>% group_by(Cluster) 
                                      %>% summarize(mean(qvalue)) %>% pull(`mean(qvalue)`) %>%
                                       mean())
significance_filtered_perclust <- map(enrichments_louvain, ~.x %>% filter(qvalue < 0.05)  %>%
                                        group_by(Cluster) %>% summarize(mean(qvalue)) %>%
                                        pull(`mean(qvalue)`) %>% mean())

num_sig_terms_total <- map(enrichments_louvain, ~.x %>% filter(qvalue < 0.05) %>% pull(qvalue) %>% length)
num_sig_terms_per_enriched_cluster <- map(enrichments_louvain, ~.x %>% 
                                            filter(qvalue < 0.05) %>% group_by(Cluster) %>%
                                            summarize(length(qvalue)) %>% pull(`length(qvalue)`) %>%
                                            mean())

avg_minq_cluster <- map(enrichments_louvain, ~.x %>% group_by(Cluster) %>% 
                  summarize(min(qvalue)) %>% pull(`min(qvalue)`) %>% mean() )

avg_min_logq_cluster <- map(enrichments_louvain, ~.x %>% 
                              mutate(log_qvalue = -log10(qvalue)) %>%
                              group_by(Cluster) %>% 
                              summarize(max(log_qvalue)) %>% 
                              pull(`max(log_qvalue)`) %>% mean() )

perc_enrichmemt <- map(enrichments_louvain, ~
                       length(.x %>%filter(qvalue < 0.05)%>% pull(Cluster) %>% sort() %>% unique())/
                         length(.x %>% pull(Cluster) %>% sort() %>% unique()))

# Calculate BHI
calc_BHI <- function(c, genes) {
  r <- 
    c%>%
    select(gene,value) %>%
    left_join(genes, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(-gene) 
  
  clusters <- as.numeric(as.character(r$value))
  
  bio_index <- 
    if(require("Biobase") && require("annotate") && require("GO.db") &&
       require("org.Hs.eg.db")) {
      BHI(clusters, annotation="org.Hs.eg.db", names=r$entrez, category="BP") 
    }
  return(bio_index)
}
  

all_clusters <- list(louvain_multk[[1]][[1]]$cluster,
                     louvain_multk[[2]][[1]]$cluster, louvain_multk[[3]][[1]]$cluster, 
                     louvain_multk[[4]][[1]]$cluster, louvain_multk[[5]][[1]]$cluster, 
                     louvain_multk[[6]][[1]]$cluster, clustering_random)


BHI <- map(all_clusters, ~calc_BHI(.x, genes=ortholog_info))


clusterings <- c("louvain_66", "louvain_88", "louvain_104", "louvain_119", "louvain_131", "louvain_140", "random")

df_enrichment_louvain <- 
  data.frame(clusterings = clusterings, 
             significance_overall = unlist(significance_overall), 
             significance_filtered = unlist(significance_filtered),
             significance_overall_perclust = unlist(significance_overall_perclust),
             significance_filtered_perclust = unlist(significance_filtered_perclust),
             num_sig_terms_total = unlist(num_sig_terms_total), 
             num_sig_terms_per_enriched_cluster = unlist(num_sig_terms_per_enriched_cluster), 
             avg_minq_cluster = unlist(avg_minq_cluster),
             avg_min_logq_cluster = unlist(avg_min_logq_cluster),
             perc_enrichmemt = unlist(perc_enrichmemt)#,
             #BHI = unlist(BHI)
  ) 

df_enrichment_louvain <- 
  df_enrichment_louvain %>%
  gather(key = measure, value = value, significance_overall:perc_enrichmemt) 

df_enrichment_louvain$clusterings <- factor(df_enrichment_louvain$clusterings,
                                            levels = unique(df_enrichment_louvain$clusterings))

df_enrichment_louvain$measure <- factor(df_enrichment_louvain$measure,
                                        levels = unique(df_enrichment_louvain$measure))



df_enrichment_louvain %>%
  ggplot(aes(x= as.factor(clusterings), y = value, fill = clusterings)) +
  geom_bar(stat = "Identity") +
  theme_minimal() +
  facet_wrap(~measure, scales="free_y") +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=10)) + 
  labs(y = "Value", x = "Clustering") +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 


df_enrichment_louvain %>%
  spread(key = measure, value = value) %>%
  ggplot(aes(x=avg_min_logq_cluster, y = perc_enrichmemt, color = num_sig_terms_per_enriched_cluster)) +
  geom_point() +
  theme_minimal()

save(enrichments_louvain,
     df_enrichment_louvain, file ="data/processed/enrichment_louvain.Rdata")


```

```{r Evaluation based on Mutual Information}

#### Find GO_filtered
load("data/processed/GO_filtered.Rdata")

GOterms <- read_tsv("data/meta/E92_all_GOterms.txt")

gene_ids <- ortholog_info %>% filter(enssscg_id %in% gene_names) %>% pull(ensg_id)
names(gene_ids) <- c()
GO_terms <- 
  GOterms %>%
  filter(`GO domain` == "biological_process") %>%
  select(ID = `Gene stable ID`, GOID = `GO term accession`) %>%
  na.omit() %>%
  # filter(ID %in% gene_ids)  %>%
  distinct()


unique(GO_terms$GOID) #11949

terms_to_keep <- 
  GO_terms %>%
  group_by(GOID) %>%
  count() %>%
  filter(n < 500) %>%
  filter(n > 2) %>%
  pull(GOID)

GO_filtered <- 
  GO_terms %>%
  filter(GOID %in% terms_to_keep)

save(GO_filtered, file = "data/processed/GO_filtered.Rdata")

MI_df <- 
  all_partitions %>%
  group_by(id) %>%
  do ({
    data <- .
    id <- unique(data$id)
    c <- data %>%
      left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
      select(ensg_id,value) %>%
      na.omit()
    
    clusters_pig <- as.integer(c$value)
    names(clusters_pig) <- c$ensg_id

    dat <- 
      data.frame(ID = names(clusters_pig),
                 cluster = as.numeric(clusters_pig)) %>% 
      left_join(GO_filtered)
    
    randoms <- c()
    for (i in c(1:50)) {
      dat_random <- 
        dat %>%
        mutate(cluster = sample(cluster))
      MI_pig_random <- mutinformation(dat_random$cluster,dat_random$GOID)
      randoms <- c(randoms, MI_pig_random)
    }

    MI_pig <- mutinformation(dat$cluster,dat$GOID)
    MI_pig_random <- mean(randoms)
    zscore <- (MI_pig - MI_pig_random)/var(randoms)
    
    data.frame(id = id, MI = MI_pig, MI_random = MI_pig_random, 
               MI_random_min = min(randoms), MI_random_max = max(randoms),
               MI_random_var = sd(randoms), MI_zscore = zscore)
  })
  

MI_to_plot <- 
  MI_df %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))



MI <- 
  MI_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    #m <- unique(data$method)
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  })


MI <- MI %>% 
  ungroup %>%
  mutate(MI_zscore = (MI - MI_random)/sqrt(MI_random_var)) 

MI %>%
  ggplot(aes(x=id, y = MI, fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 6)) +
  facet_wrap(~(clustering), scale="free_x") +
 scale_fill_viridis(discrete = F)+
  labs(y = "Value", x = "Clustering strategy")




save(MI, file = "data/processed/MI.Rdata")

```

```{r}



```

```{r Merge evaluations}
#save(all_eval, file = "data/processed/all_eval.Rdata")

load("data/processed/all_eval.Rdata")
load("data/processed/MI.Rdata")
load("data/processed/all_bio.Rdata")

# All measures
all <- 
  all_eval %>% 
  rename(id = clustering_id) %>%
  select(id, connectivity_index, avg_silhouette_width,time) %>%
  left_join(MI %>% ungroup() %>% select(id,MI_zscore)) %>%
  left_join(all_bio %>% select(-min_significance)) %>%
  na.omit() %>%
  gather(key = measure, value = value, connectivity_index:annotability) 

all$value <- as.numeric(all$value)

# Annotation & color palettes
annotation <-
  data.frame(id = all$id %>% unique()) %>%
  group_by(id) %>%
  mutate(scale_dist = paste(strsplit(id, " ")[[1]][1], strsplit(id, " ")[[1]][2]),
         clust = strsplit(id, " ")[[1]][3],
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  column_to_rownames("id")

p <- pal_npg("nrc")(6)
clust <- c(p[3],p[1],p[4],p[6],p[2],p[5]) 
names(clust) = c("fasthclust", "fastkmeans", "leiden", "louvain", "medoids", "SOM")
scale_dist = c("#D15A64", "#8AD264","#7C2B64","#EEA564")
names(scale_dist) = c("max euclidean", "max pearson", "zscore euclidean", "zscore pearson")
ann_colors = list(clust = clust, scale_dist = scale_dist)
```

```{r Ranking 1 - all -> best k}

measures <- c("connectivity_index", "avg_silhouette_width", "MI_zscore","annotability")

all_ranked <- 
  all %>%
  ungroup() %>%
  scaled_rank(measures=measures)

# Sorted by average
all_ranked %>% 
 select(-value) %>% 
  spread(key=measure, value=scaled_value) %>% 
  select(id,connectivity_index, avg_silhouette_width, MI_zscore, annotability,avg_value) %>% 
  arrange(-avg_value) %>% 
  column_to_rownames("id") %>%
  pheatmap(cluster_cols = F, cluster_rows = F, fontsize = 8, border_color = NA, 
           color = colorRampPalette(c("white","#00846b"))(100),
           annotation_row  = annotation, show_rownames = F,
           annotation_colors = ann_colors)
a <- all_ranked %>% 
 select(-value) %>% 
  spread(key=measure, value=scaled_value)

# Clustered
all_ranked %>% 
 select(-value) %>% 
  spread(key=measure, value=scaled_value) %>% 
  select(id,connectivity_index, avg_silhouette_width, MI_zscore, annotability,avg_value) %>% 
  arrange(avg_value) %>%
  column_to_rownames("id") %>%
  pheatmap(cluster_cols = F, fontsize = 8, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("white","#00846b"))(100),
           annotation_row  = annotation,
           show_rownames = F,
           annotation_colors = ann_colors)


rank_k <- 
  all_ranked %>%
  group_by(id) %>%
  mutate(scale_dist_clust = paste(strsplit(id, " ")[[1]][1], strsplit(id, " ")[[1]][2], strsplit(id, " ")[[1]][3]),
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  ungroup() %>% 
  group_by(scale_dist_clust) %>%
  top_n(n = 1, wt = avg_value)
  
  
  
best_k <- 
  rank_k %>% 
  group_by(scale_dist_clust) %>% 
  scaled_rank(measures=measures) %>% 
  group_by(scale_dist_clust) %>% 
  top_n(n=1, wt=avg_value) 

best_k %>%
  select(-value) %>%
  spread(key=measure, value=scaled_value) %>% 
  ungroup() %>% 
  select(-scale_dist_clust,-k) %>% 
  column_to_rownames("id") %>%
  pheatmap(cluster_cols = F, fontsize = 8, border_color = NA, 
           cutree_rows = 3, clustering_method = "ward.D2", 
           color = colorRampPalette(c("white","#00846b"))(100),
           annotation_row  = annotation,
           show_rownames = F,
           annotation_colors = ann_colors)

 
best_clust <- 
best_k %>%
  mutate(clust = strsplit(id, " ")[[1]][3]) %>%
  ungroup() %>% 
  group_by(clust) %>% 
  scaled_rank(measures=measures) %>% 
  group_by(clust) %>% 
  top_n(n=1, wt=avg_value) 

best_clust %>%
  select(-value) %>% 
  spread(key= measure, value = scaled_value) %>% 
  arrange(-avg_value)

final_rank <- 
  best_clust %>%
  ungroup() %>% 
scaled_rank(measures=measures) %>%
  select(-value) %>% 
  spread(key= measure, value = scaled_value) %>% 
  arrange(-avg_value)

best_ids_clust <- final_rank %>% pull(id)

# # best clustering -> best
# all %>% 
#   ungroup() %>% 
#   group_by(id) %>%
#   mutate(clust = strsplit(id, " ")[[1]][3]) %>%
#   ungroup() %>% 
#   filter(measure %in% measures) %>%
#   
#     group_by(clust) %>% 
#     do ({
#       data <- .
#       if (unique(data$measure) == "connectivity_index") {
#         data %>%
#           mutate(scaled_value=  scales::rescale(-value,c(0,1)))
#       }
#       else {
#         data %>%
#           mutate(scaled_value=  scales::rescale(value,c(0,1)))
#       }
#     }) %>%
#     group_by(id) %>%
#     mutate(avg_value = mean(scaled_value)) %>%
#     group_by(id) %>%
#     arrange(-avg_value)  %>% 
#   select(-value) %>% 
#   spread(key=measure, value=scaled_value) %>% 
#   select(id,connectivity_index, avg_silhouette_width, MI_zscore, annotability,avg_value) %>% 
#   arrange(avg_value)
# 
# 
# 
# 
# all %>%
#   group_by(id) %>% 
#    mutate(clust = strsplit(id, " ")[[1]][3]) %>%
#   ungroup() %>% 
#   group_by(clust) %>% 
#   scaled_rank(measures=measures) %>% 
#   group_by(clust) %>% 
#   top_n(n=1, wt=avg_value) %>% 
#   select(-value) %>%
#   spread(key=measure, value=scaled_value) %>% 
#   arrange(-avg_)

```

```{r Radar plots}
library(fmsb)

data_radar <- 
  all_ranked %>%
  group_by(id) %>%
  mutate(clust = strsplit(id, " ")[[1]][3]) %>% 
  ungroup() %>% 
  group_by(clust) %>% 
  top_n(n=1,wt=avg_value) %>% 
  ungroup() %>% 
  select(id,measure,scaled_value) %>% 
  spread(key=measure, value=scaled_value) %>% 
  left_join(all_ranked %>%  select(id,avg_value) %>% unique()) 

data_radar$id <- factor(data_radar$id, levels = c("max pearson fasthclust 50","zscore pearson fastkmeans 70",
                                                  "zscore euclidean leiden 73","zscore pearson louvain 68",
                                                  "zscore pearson medoids 70", "zscore pearson SOM 50"))


data_radar <- data_radar %>% arrange(id)

ann_transp <- unlist(map(ann_colors$clust, ~adjust_transparency(.x, alpha = 0.007)))

c <- ann_colors$clust

names(c) <-c("max pearson fasthclust 50","zscore pearson fastkmeans 70",
                                                  "zscore euclidean leiden 73","zscore pearson louvain 68",
                                                  "zscore pearson medoids 70", "zscore pearson SOM 50")

c2 <- ann_transp
names(c2) <-c("max pearson fasthclust 50","zscore pearson fastkmeans 70",
                                                  "zscore euclidean leiden 73","zscore pearson louvain 68",
                                                  "zscore pearson medoids 70", "zscore pearson SOM 50")

library(ggiraphExtra)
library(ggradar)
library(colorspace)


ggRadar(data_radar,scales = "free", rescale = F, aes(color=id,group=id), alpha = 0.12) +
  geom_line(size=0.8) +
  theme_minimal() +
  scale_color_manual(values = c) +
 scale_fill_manual(values = c2) +
  ylim(0.3, 1)




ggRadar(data_radar,scales = "fixed",rescale=F, aes(facet = id,color=id), alpha = 0.3) +
  theme_minimal() +
    geom_line(size=0.8) +
  scale_color_manual(values = c) +
  scale_fill_manual(values = c2) +
  ylim(0, 1)



d2 <-
  data_radar %>% 
  gather(key=measure,value=value,-id) %>% 
  spread(key=id, value=value) %>% 
  filter(measure %in% c("annotability","avg_silhouette_width","connectivity_index","MI_zscore"))
  


ggRadar(d2,rescale=F,aes(facet = measure)) +
      geom_line(size=0.8) +
  theme_minimal() +
  ylim(0.2, 1)

#+
#  scale_color_manual(values = c) +
 # scale_fill_manual(values = c2)














max_min <- data.frame(
  annotability  = c(1, 0), avg_silhouette_width  = c(1, 0),
  connectivity_index  = c(1, 0), MI_zscore  = c(1, 0)#,
 # avg_value  = c(1, 0)
)
rownames(max_min) <- c("Max", "Min")

# Bind the variable ranges to the data
data_radar <- rbind(max_min, data_radar%>%  arrange(id) %>%   column_to_rownames("id")) 
   
library(colorspace)



radarchart(data_radar, pcol = ann_colors$clust,
           axistype = 10,
           pfcol = ann_transp,
           plwd = 2,
           plty =1,
           cglcol="grey", cglty=1, axislabcol="grey", 
           caxislabels=seq(0,20,5), cglwd=0.8,
           vlcex=1)

legend(x=0.7, y=1, legend = rownames(data_radar[-c(1,2),]), bty = "n", pch=20 , col=ann_colors$clust , text.col = "black", cex=1.2, pt.cex=3)

# PER CLUSTERING

data_r %>% 
radarchart(data_r, pcol = ann_colors$clust,
           pfcol = ann_transp,
           plwd = 2,
           plty =1,
           cglcol="grey", cglty=1, axislabcol="grey", 
           caxislabels=seq(0,20,5), cglwd=0.8,
           vlcex=1)

legend(x=0.7, y=1, legend = rownames(data_r[-c(1,2),]), bty = "n", pch=20 , col=ann_colors$clust , text.col = "black", cex=1.2, pt.cex=3)


# PER MEASURE



```

```{r Correlation metrics}


```




```{r All UMAPs & comparison}
#1. zscore pearson louvain 68
partition <-
  all_partitions %>%
  filter(id == "zscore pearson louvain 68") %>%
  mutate(value = value + 1)

# All umaps and save
umap_zscore_euclidean <- 
  pcas[[1]]$scores %>% 
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  rownames_to_column("enssscg_id") %>%
  select((1:30)) %>%
  umap_calc(row_names= "enssscg_id", n_neigh = 15, met = "euclidean")

umap_zscore_correlation <- 
  pcas[[1]]$scores %>% 
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  rownames_to_column("enssscg_id") %>%
  select((1:30)) %>%
  umap_calc(row_names= "enssscg_id", n_neigh = 15, met = "correlation")

umap_max_euclidean <- 
  pcas[[2]]$scores %>% 
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  rownames_to_column("enssscg_id") %>%
  select((1:30)) %>%
  umap_calc(row_names= "enssscg_id", n_neigh = 15, met = "euclidean")

umap_max_correlation <- 
  pcas[[2]]$scores %>% 
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  rownames_to_column("enssscg_id") %>%
  select((1:30)) %>%
  umap_calc(row_names= "enssscg_id", n_neigh = 15, met = "correlation")


# 
# umap_euc %>% umap_plot(color_by = "cluster", color_groups = partition)
# umap_cor %>% umap_plot(color_by = "cluster", color_groups = partition)
# 
# umap_zscore_euclidean %>% umap_plot()
# umap_zscore_correlation %>% umap_plot()
# umap_max_euclidean %>% umap_plot()
# umap_max_correlation %>% umap_plot()

umaps <- list(umap_zscore_euclidean, umap_zscore_correlation, 
     umap_max_euclidean, umap_max_correlation)

names(umaps) <- c("zscore euclidean", "zscore correlation", 
                "max euclidean", "max correlation")

save(umaps, file= "data/processed/umaps_euc_corr.Rdata")

to_plot <- 
  umap_zscore_correlation %>%
  #left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  left_join(partition, by = c("features" = "gene")) #%>%
#  mutate (value = value+1)

t <-
  to_plot %>%
  select(-value)

# Tissues for each cluster

#tissue_colors[["NA"]] <- "grey30"

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(69)


to_plot %>%
  ggplot(aes(V1,V2,color=as.factor(value))) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5) +
             theme_minimal() +
             coord_fixed() +
             scale_color_manual(values = p) 


  plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5, color = p[cluster]) +
             theme_minimal() +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) 
         })

pdf("t_cor.pdf")
plots
dev.off()



p1 <- umaps[[1]] %>%umap_plot()
p2 <- umaps[[2]] %>% umap_plot()
  p3 <- umaps[[3]] %>%  umap_plot()
  p4 <- umaps[[4]] %>%  umap_plot()

p1 + p2 + p3 + p4
```

```{r Clustering UMAPs: visual evaluation}
load("data/processed/umaps_euc_corr.Rdata")
# 
# umap_res %>%
#       mutate(enssscg_id = gene_names) %>%
#       left_join(color_groups, by = c("enssscg_id" = "gene")) %>%
#       ggplot(aes(V1,V2, color = as.factor(value))) +
#       geom_point(show.legend = F) +
#       theme_minimal() +
#       scale_color_manual(values = pal) +
#       coord_fixed()
# 
# umaps <- map(
#   pcas,
#   ~.x$scores %>%
#     set_rownames(gene_names) %>%
#     as.data.frame() %>%
#     rownames_to_column("enssscg_id") %>%
#     select((1:30)) %>%
#     umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2, n_ep = 1000) 
# )
# 
# save(umaps, file = "data/processed/umaps_zscore_max.Rdata")
# 
# load("data/processed/umaps_zscore_max.Rdata")



clustering_umaps <- function(name, umaps, partitions) { #, 
  partition <-
    partitions %>% 
    filter(id == name)
  
  if(grepl("zscore euclidean",name)) {
    umap <- umaps$`zscore euclidean`
  }
  
  else if(grepl("zscore pearson",name)) {
    umap <- umaps$`zscore correlation`
  }
  else if(grepl("max euclidean",name)) {
    umap <- umaps$`max euclidean`
  }
  else if(grepl("max pearson",name)) {
    umap <- umaps$`max correlation`
  }
  else if(grepl("random",name)) {
    umap <- umaps$`zscore correlation`
  }
  
    
  umap_plot(umap_res = umap, color_by = "cluster", color_groups = partition) +
    ggtitle(name)
  
  #return(p)
}


clustering_umaps(name = "zscore pearson fastkmeans 50",
                 umaps = umaps,
                 partitions = all_partitions)

best_ids

b <- best_clust %>% pull(id) %>% unique()
plots <- map(b,
             ~ clustering_umaps(name = .x, umaps = umaps, partitions = all_partitions) +
               ggtitle(.x))


umaps$`zscore euclidean` %>% 
  ggplot(aes(V1, V2)) +
  geom_point()

pdf("umaps_best_clust.pdf", useDingbats = F)

# map(best_ids,
#     ~ clustering_umaps(id = .x, umaps = umaps, partitions = all_partitions) +
#       ggtitle(.x))
plots
dev.off()







load("data/processed/umaps_euc_corr.Rdata")
plot_data <-
 umaps$`zscore correlation` %>%
              #select(barcode, orig.ident)) %>%
  mutate(r = scales::rescale(V1, c(0,1)),
         g = scales::rescale(V2, c(0,1)),
         b = 0.5,
         color = rgb(r, g, b)) %>%
  left_join(umaps$`max correlation`,
            by = "features",
            suffix = c("", "_int"))


plot_data %>% {
  # ggplot(., aes(V1, V2, color = features)) +
  #   geom_point(size = 0.1) +
  #   theme_bw() +
  #   coord_fixed() +
  #  
  #   ggplot(., aes(V1_int, V2_int, color = features)) +
  #   geom_point(size = 0.1) +
  #   theme_bw() +
  #   coord_fixed() +
  #  
    ggplot(., aes(V1, V2, color = color)) +
    geom_point(size = 0.1) +
    theme_bw() +
    coord_fixed() +
    scale_color_identity() +
   
    ggplot(., aes(V1_int, V2_int, color = color)) +
    geom_point(size = 0.1) +
    theme_bw() +
    coord_fixed() +
    scale_color_identity()
}







plot_data <-
  SB_SC_filt@reductions$umap@cell.embeddings %>%
  as_tibble(rownames = "barcode") %>%
  left_join(SB_SC_filt@meta.data %>%
              as_tibble(rownames = "barcode") %>%
              select(barcode, orig.ident)) %>%
  mutate(r = scales::rescale(UMAP_1, c(0,1)),
         g = scales::rescale(UMAP_2, c(0,1)),
         b = 0.5,
         color = rgb(r, g, b)) %>%
  left_join(SB_SC_int@reductions$umap@cell.embeddings %>%
              as_tibble(rownames = "barcode"),
            by = "barcode",
            suffix = c("", "_int"))


plot_data %>% {
  ggplot(., aes(UMAP_1, UMAP_2, color = orig.ident)) +
    geom_point(size = 0.1) +
    theme_bw() +
    coord_fixed() +
   
    ggplot(., aes(UMAP_1_int, UMAP_2_int, color = orig.ident)) +
    geom_point(size = 0.1) +
    theme_bw() +
    coord_fixed() +
   
    ggplot(., aes(UMAP_1, UMAP_2, color = color)) +
    geom_point(size = 0.1) +
    theme_bw() +
    coord_fixed() +
    scale_color_identity() +
   
    ggplot(., aes(UMAP_1_int, UMAP_2_int, color = color)) +
    geom_point(size = 0.1) +
    theme_bw() +
    coord_fixed() +
    scale_color_identity()
}
```

```{r Retrieve partitions for ARI}

# load("data/processed/kmeans_multk.Rdata")
# load("data/processed/medoids_multk.Rdata")
# load("data/processed/hclust_multk.Rdata")
# load("data/processed/louvain_multk.Rdata")
# load("data/processed/leiden_multk.Rdata")
# load("data/processed/som_multk.Rdata")

all_dfs <- map(list(kmeans_multk, medoids_multk, hclust_multk, 
                    louvain_multk, leidenn_multk, som_multk),
               ~to_df(.x))


all_partitions <- merge_all_df(all_dfs)

# save(all_partitions, file ="data/processed/all_partitions.Rdata")
load("data/processed/all_partitions.Rdata")
# For all
clusterings <- unique(all_partitions$id)

#36 options
res_ari <- matrix(, nrow = 168, ncol = 168)
rownames(res_ari) <- c(clusterings)
colnames(res_ari) <- c(clusterings)

# Calculate ARI matrix
for (i in clusterings) {
  for (j in clusterings) {
    a <- ARI(all_partitions %>% filter(id == i) %>% pull(value), 
             all_partitions %>% filter(id == j) %>% pull(value))
    res_ari[i,j] <- a
    }
  }

save(res_ari, file="data/processed/res_ari.Rdata")
load("data/processed/res_ari.Rdata")

annotation <-
  data.frame(id = all$id %>% unique()) %>%
  group_by(id) %>%# rename(id = all.id) %>%
  mutate(scale_dist = paste(strsplit(id, " ")[[1]][1], strsplit(id, " ")[[1]][2]),
         clust = strsplit(id, " ")[[1]][3],
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  column_to_rownames("id")

# Exclude random clustering
ids <- rownames(res_ari)
ids <- ids[!(grepl("random", ids))]
res_ari [ids,ids] %>%
  pheatmap(fontsize = 8, 
           border_color = NA, 
           clustering_method = "ward.D2", 
           annotation_row  = annotation,
           show_rownames = F,
           show_colnames = F, 
           annotation_colors = ann_colors)


## Filtered

best_partitions <-best_k_of_each$id %>% unique()

filtered <-
  all_partitions %>%
  filter(id %in% best_partitions)

res_ari_filtered <- matrix(, nrow = 24, ncol = 24)
rownames(res_ari_filtered) <- c(best_partitions)
colnames(res_ari_filtered) <- c(best_partitions)

# Calculate ARI matrix
for (i in best_partitions) {
  for (j in best_partitions) {
    a <- ARI(filtered %>% filter(id == i) %>% pull(value), 
             filtered %>% filter(id == j) %>% pull(value))
    res_ari_filtered[i,j] <- a
    }
  }

save(res_ari_filtered, file="data/processed/res_ari_filtered.Rdata")

load("data/processed/res_ari_filtered.Rdata")
annotation_filtered <-
  data.frame(id = best_partitions) %>%
  group_by(id) %>%# rename(id = all.id) %>%
  mutate(scale_dist = paste(strsplit(id, " ")[[1]][1], strsplit(id, " ")[[1]][2]),
         clust = strsplit(id, " ")[[1]][3],
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  column_to_rownames("id")

# Exclude random clustering
res_ari_filtered %>%
  pheatmap(fontsize = 8, 
           border_color = NA, 
           clustering_method = "ward.D2", 
           annotation_row  = annotation,
           show_rownames = F,
           show_colnames = F,
           annotation_colors = ann_colors)

```

```{r Linear model}
load("data/processed/all_eval.Rdata")
# all %>% 
#   ggplot(aes(x= value)) +
#   geom_density() +
#   facet_wrap(~measure, scales = "free")

to_model <- 
  all %>% 
  group_by(id) %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
        distance = strsplit(id, " ")[[1]][2],
        clustering = strsplit(id, " ")[[1]][3],
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  group_by(measure) %>%
  mutate(mean = mean(value),
         sd = sd(value)) %>%
      ungroup() %>%
      mutate (value_scaled = (value - mean)/sd) %>%
  filter(measure %in% measures) %>%
    group_by(measure) %>%
    do ({
      data <- .
      if (unique(data$measure) == "connectivity_index") {
        data %>%
          mutate(rank= rank(value,ties.method = "average"))
      }
      else {
        data %>%
          mutate(rank = rank(-(value),ties.method = "average"))
      }
    }) %>%
    group_by(id) %>%
    mutate(avg_rank = mean(rank)) %>%
    arrange(avg_rank) %>% 
  select(-value,-mean,-sd)

to_model_all <-
  bind_rows(to_model %>% 
  select(-measure,-value_scaled,-rank) %>%
  unique() %>%
  mutate(measure = "avg_rank") %>%
  rename(value_scaled=avg_rank),
  to_model %>%
  select(-rank, -avg_rank))

measures <- c("connectivity_index", "avg_silhouette_width", "MI_zscore" ,"annotability", "avg_rank")


models <- 
  to_model_all %>% 
  filter(measure %in% measures) %>% 
  group_by(measure) %>%
  do ({
    data <- . 

        model <- lm(value_scaled ~ scaling + distance + clustering + k, data = data) 
    anova_model <- aov(model) 
    
    anova_model %>%  
      tidy() %>% 
      left_join(anova_model %>% 
                  omega_sq() %>% 
                  enframe("term", "omega_sq") %>% 
                  mutate(term = trimws(term)),
                by = "term")
  })


### Does not work for k!
models_estimates <- 
  to_model %>% 
  filter(measure %in% measures) %>% 
  group_by(measure) %>%
  do ({
    data <- .
    model <- lm(value_scaled ~ scaling + distance + clustering + k, data = data) 
    anova_model <- aov(model) 
    anova_model %>% TukeyHSD() %>% 
      tidy() %>% 
      arrange(adj.p.value)
  })
  
models$term <- factor(models$term, levels = c("Residuals","k","clustering","distance","scaling"))
models$measure <- factor(models$measure, levels = c("connectivity_index","avg_silhouette_width","MI_zscore","annotability","avg_rank"))

models %>%
 # mutate(measure = as.factor(measure)) %>% 
  ggplot(aes(x = measure, y= omega_sq, fill = term)) +
  geom_bar(stat="identity",width = 0.5) +
 # scale_fill_npg() +
  #scale_fill_viridis(discrete=TRUE, name="") +
    theme_bw() +
    ylab("Omega Sq") +
  scale_fill_manual(values = pal_models) +
  theme(plot.background = element_blank(),
        panel.grid.major = element_blank(),
    panel.grid.minor = element_blank())



to_model_all %>%
  group_by(id) %>%
  mutate(scale_dist = paste(scaling,distance)) %>% 
  ungroup() %>% 
  filter(measure=="avg_rank") %>%
  ggplot(aes(x=scale_dist, y = value_scaled, fill=scale_dist)) +
  geom_boxplot() +
  scale_y_continuous(trans = "reverse")+
  #facet_wrap(~measure, scales="free_y") +
  scale_fill_manual(values = ann_colors$scale_dist)+
  theme_bw() +
  theme(plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) 


to_model_all %>%
  #filter(measure=="avg_rank") %>%
  ggplot(aes(x=clustering, y = value_scaled, fill=clustering)) +
  geom_boxplot() +
  #scale_y_continuous(trans = "reverse")+
  facet_wrap(~measure, scales="free_y") +
  scale_fill_manual(values = pal_clustering)+
  theme_bw() +
  theme(plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) #+
      theme(element_text(size = 3)) 
      
      
P1 <-
  to_model_all%>%
  filter(measure=="avg_rank") %>%
  ggplot(aes(x=clustering, y = value_scaled, fill=clustering)) +
  geom_boxplot() +
  #scale_y_continuous(trans = "reverse")+
  scale_fill_manual(values = pal_clustering)+
  theme_bw() +
  theme(plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) 

P2 <-
 to_model_all%>%
  filter(measure=="avg_rank") %>%
  group_by(id) %>%
  mutate(scale_dist = paste(scaling,distance)) %>% 
  ggplot(aes(x=scale_dist, y = value_scaled, fill=scale_dist)) +
  geom_boxplot() +
  scale_fill_manual(values = ann_colors$scale_dist)+
  theme_bw() +
  theme(plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

P3 <-
  to_model_all%>%
  group_by(id) %>%
  group_by(k) %>%
  filter(measure=="avg_rank") %>%
  mutate(k = ifelse((65<k & k<75), 60))
  
  ggplot(aes(x= as.factor(k), y = value_scaled, fill=k, group = k)) +
  geom_boxplot() +
  #scale_y_continuous(trans = "reverse")+
 # scale_fill_manual(values = pal_clustering)+
  theme_bw() +
  theme(plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) 


```

```{r Evaluation based on intra cluster correlations}

# Calcualte correlations for one clustering strategy
load("data/processed/all_partitions.Rdata")

clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value)

random_clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value) %>%
  mutate(value = sample(value))

correlations <-
  tmm_data %>%
  mutate(log_tmm = log(tmm + 1))%>%
  select(-tmm) %>%
  left_join(random_clustering, by = c("enssscg_id" = "gene")) %>%
  group_by(value) %>%
  do({
    cluster <- unique(.$value)
    data <- .
    data <- data %>%
      select(-value) %>%
      spread(key = enssscg_id, value= log_tmm) %>%
      select(-sample_ID)

    cor_matrix <- cor(data)  %>%
      as_tibble(rownames = "gene1") %>%
      gather(gene2, cor, -gene1) %>%
      filter (gene1 < gene2) %>%
      mutate(cluster = cluster)
  })

p2 <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(50)

p_random <-
  correlations %>%
  ggplot(aes(x = as.factor(cluster), y = cor, fill = as.factor(cluster))) +
    geom_hline(aes(yintercept = mean(cor)), color="grey") +
  geom_violin(draw_quantiles = 0.5, show.legend = F) +
  theme_bw() +
  scale_fill_manual(values = p2) +
  ylab("Intra cluster correlation") +
  xlab("Cluster") + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


p / p_random


correlations %>% 
  group_by(cluster) %>%
  summarize(mean(cor)) %>%
  summarize(mean(`mean(cor)`)) 




# Calculate intra-cluster correlation for all partitions

intra_cluster_cor <- function (df) {
  df %>%
    group_by(value) %>%
    do({
      cluster <- unique(.$value)
      data <- .
      data <- data %>%
        select(-value) %>%
        spread(key = enssscg_id, value= log_tmm) %>%
        select(-sample_ID)
      
      cor_matrix <- cor(data)  %>%
        as_tibble(rownames = "gene1") %>%
        gather(gene2, cor, -gene1) %>%
        filter (gene1 < gene2) %>%
        mutate(cluster = cluster)
    })
}


all_correlations <-
  all_partitions %>%
  group_by(id) %>%
  do ({
    clustering <- . 
    id <- unique(clustering$id)
    correlations <-
      tmm_data %>%
      mutate(log_tmm = log(tmm + 1))%>%
      select(-tmm) %>%
      left_join(clustering %>% select(-id), by = c("enssscg_id" = "gene")) %>%
      intra_cluster_cor()
      
    average_correlation <- correlations %>%
      group_by(cluster) %>%
      summarize(mean(cor)) %>%
      summarize(mean(`mean(cor)`))
    data.frame(id = id, avg_cor = average_correlation)
  })

save(all_correlations, file = "data/processed/all_correlations.Rdata")

load( "data/processed/all_correlations.Rdata")
cor_to_plot <- 
  all_correlations %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))

l <- unique(sort(cor_to_plot$k))
cor_to_plot$k <- factor(cor_to_plot$k, levels = l)

test <- 
  cor_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    #m <- unique(data$method)
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  })


test %>%
  ggplot(aes(x=id, y = mean..mean.cor..., fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~(clustering), scale="free") +
 # scale_fill_viridis(discrete = F)+
  labs(y = "Value", x = "Clustering strategy") +


best_ids_clust



all_correlations %>% 
  filter(id == "max euclidean fasthclust 50")

all_correlations %>% 
  filter(id %in% best_ids_clust) %>%
  arrange(-mean..mean.cor...)
```



#6- Visualization

```{r Visualize clusters in UMAP}
load("data/processed/umaps_euc_corr.Rdata")
#umap_zscores <- umaps[[4]]
umap_zscores <- umaps[[2]]

#zscore pearson louvain 68
#max pearson medoids 50

clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value) #%>%
#  mutate(value = value +1)

to_plot <- 
  umap_zscores %>%
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  left_join(clustering, by = c("features" = "gene")) #%>%
#  mutate (value = value+1)

t <-
  to_plot %>%
  select(-value,-enhanced_tissues)

# Tissues for each cluster

tissue_colors[["NA"]] <- "grey30"

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(70)
plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5, color = p[cluster]) +
             theme_minimal() +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) + 
             scale_color_manual(values = tissue_colors) 
         })


plots[[30]]
pdf("zscore_pearson_medoids_70.pdf")
plots

dev.off()

umap_plot(umap_zscores, color_by = "cluster", color_groups = clustering)

plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2,color=enhanced_tissues)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5) +
             theme_minimal() +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) + 
             scale_color_manual(values = tissue_colors) 
         })
```

```{r Network}

```


#7- Annotation

```{r Tissue expression per cluster (pheatmaps)}
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


#clustering <- louvain_multk[[1]][[1]]$cluster
#lustering <- max_clustering

data_tissue <-
  tmm_data %>%
  mutate(tissue = gsub('.{2}$', '', sample_ID)) %>%
  group_by(tissue,enssscg_id) %>%
  mutate(tissue_exp = mean(tmm)) %>%
  select(enssscg_id, tissue, tissue_exp) %>%
unique()

# clustering <-
#   clustering %>%
#   mutate(value = value+1)

clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value) 

pdf("heatmaps_zscore_person_medoids_70.pdf")
heatmaps <- lapply(sort(unique(clustering$value)),
         function(i) {
           plot <- data_tissue %>%
             left_join(clustering, by = c("enssscg_id" = "gene")) %>%
            # mutate(value = value+1) %>%
             filter(value == i) %>%
             group_by(enssscg_id) %>%
             mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
             ungroup() %>%
             select(enssscg_id,tissue,scaled_value) %>%
             spread(key=tissue, value = scaled_value) %>%
             column_to_rownames("enssscg_id") %>% 
             pheatmap(clustering_method = "ward.D2", 
                      color = viridis(20, direction = -1, option = "B"),
                      show_rownames = F,border_color = NA, fontsize = 8)
         })
dev.off()

save(heatmaps[[1]], file="h.png")

pdf("heatmaps_zscore_person_medoids_70.pdf")
heatmaps

save(heatmaps,file = "data/processed/pheatmaps.Rdata")
## scale before

scaled_data <-
  data_tissue %>%
             group_by(enssscg_id) %>%
             mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
             ungroup()



             left_join(clustering, by = c("enssscg_id" = "gene"))
pdf("heatmaps_zscore_person_medoids_70.pdf")
lapply(sort(unique(clustering$value)),
         function(i) {
           plot <-  %>%
            # mutate(value = value+1) %>%
             filter(value == i)  %>%
             select(enssscg_id,tissue,scaled_value) %>%
             spread(key=tissue, value = scaled_value) %>%
             column_to_rownames("enssscg_id") %>% 
             pheatmap(clustering_method = "ward.D2", 
                      color = viridis(20, direction = -1, option = "B"),
                      show_rownames = F,border_color = NA, fontsize = 8)
         })
dev.off()





data_tissue %>%
             left_join(clustering, by = c("enssscg_id" = "gene")) %>%
            # mutate(value = value+1) %>%
             filter(value == 67) %>%
             group_by(enssscg_id) %>%
             mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
             ungroup() %>%
             select(enssscg_id,tissue,scaled_value) %>%
             spread(key=tissue, value = scaled_value) %>%
             column_to_rownames("enssscg_id") %>% 
             pheatmap(clustering_method = "ward.D2", 
                      color = viridis(20, direction = -1, option = "B"),
                      show_rownames = F,border_color = NA, fontsize = 8)
         





load("data/processed/enrichments_medoids.Rdata")

A <-
  enrichments_medoids[[3]] %>% 
  filter(id == "zscore pearson medoids 70") %>% 
  arrange(qvalue) %>% 
   group_by(Cluster) %>% 
 arrange(qvalue) %>% 
 top_n(n=10,wt=(-qvalue))

```

```{r Agreement with previous classification}


gene_info

clus <- 
  all_partitions %>%
  filter(id == "zscore euclidean fasthclust 50") %>%
  select(gene, value)

clas <- 
  gene_info %>%
  select(gene = ensg_id, specificity_category)

# UMAP by classification
umaps[[1]] %>%
  rename(gene = features)%>%
  left_join(clus) %>%
  left_join(clas) %>%
      ggplot(aes(V1,V2, color = specificity_category )) +
      geom_point() +
      theme_minimal() +
   #   scale_color_manual(values = tissue_colors) +
      coord_fixed()


# UMAP by tissue
umaps[[1]] %>%
  rename(gene = features)%>%
  left_join(clus) %>%
  left_join(tissue_info, by = c("gene" = "ensg_id")) %>%
      ggplot(aes(V1,V2, color = enhanced_tissues)) +
      geom_point(show.legend = F) +
      theme_minimal() +
      scale_color_manual(values = tissue_colors) +
      coord_fixed()



# Tissue contributaion UMAPs
saveRDS(clust_louvain, "results/Human Clustering.RDS")

# load("data/processed/data_scaled.Rdata")
# load("data/processed/pcas.Rdata")
## UMAP

clus_umap <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2)

clus_umap %>% umap_plot(color_by = "density")

ggsave("results/human UMAP density.pdf", width = 5, height = 5)
                        
clus_umap %>%  
  left_join(clust_louvain$cluster, 
            by = c("features" = "gene")) %>%
  ggplot(aes(V1,V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  # scale_color_manual(values = pal) +
  coord_fixed()

ggsave("results/human UMAP clusters.pdf", width = 5, height = 5)


color_data <- 
  tmm_data %>%
  left_join(colors,
            by = c("celltype" = "tissue")) %>%
  group_by(ensg_id, color = consensus_tissue_color) %>% 
  summarise(tpm = max(tpm)) %>% 
  
  group_by(ensg_id) %>%
  mutate(tpm = tpm / sum(tpm, na.rm = T)) %>%
  ungroup()

ensg_exp_palette <-
  color_data %>%
  bind_cols(col2rgb(.$color) %>% 
              t() %>% 
              as_tibble()) %>%
  mutate(red = red * tpm,
         green = green * tpm,
         blue = blue * tpm) %>%
  group_by(ensg_id) %>% 
  summarise(red = sum(red),
            green = sum(green),
            blue = sum(blue)) %>%
  mutate(color = rgb(red, green, blue, maxColorValue = 255))



clus_umap %>%
  left_join(ensg_exp_palette,
            by = c("features" = "ensg_id")) %>%
  ggplot(aes(V1,V2, color = color)) +
  geom_point(show.legend = F,
             size = 0.2) +
  theme_minimal() +
  scale_color_identity() +
  coord_fixed()

ggsave("results/human UMAP tissue contribution.pdf", width = 5, height = 5)
```

```{r Fischer test 1}

## HT to tissue

tissue_info
clus %>%
  
# HT to category (enriched / enhanced)

clas
clus

# ----- Hypergeometric test ------

cluster_ht <- 
  clus %>%
  rename(cluster=value) %>%
  left_join(gene_info, by = c("gene" = "ensg_id")) %>% 
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues), 
                                   "not enriched", enhanced_tissues)) %>%
  separate_rows(enhanced_tissues, sep = ",") %>%
  
  {
    # q is the number of successes
    q <- 
      group_by(., enhanced_tissues, cluster, .drop = F) %>%
      summarise(q = n_distinct(gene))
    # k is the number of tries - i.e. the cluster size
    k <- 
      group_by(., cluster, .drop = F) %>%
      summarise(k = n_distinct(gene))
    
    # m is the number of possible successes
    m <- 
      group_by(., enhanced_tissues, .drop = F) %>%
      summarise(m = n_distinct(gene))
    
    # n is the population size - i.e. the number of genes
    n <- n_distinct(.$gene)
    
    q %>% 
      left_join(k) %>% 
      left_join(m) %>%
      # n is the population size - i.e. the number of genes
      mutate(n = n - m) 
    
  } 

cluster_ht_res <- 
  cluster_ht %>%
  group_by(enhanced_tissues, cluster) %>%
  # Testing the chance of getting the observed number or higher per cluster and tissue type
  summarise(p_value = phyper(q - 1, m, n, k, lower.tail = F)) %>%
  ungroup() %>%
  mutate(p_value = ifelse(p_value == 0, .Machine$double.xmin, p_value),
         adj_pval = p.adjust(p_value, method = "BH"),
         sign_level = case_when(adj_pval < 0.00000001 ~ 8,
                                adj_pval < 0.0000001 ~ 7,
                                adj_pval < 0.000001 ~ 6,
                                adj_pval < 0.00001 ~ 5,
                                adj_pval < 0.0001 ~ 4,
                                adj_pval < 0.001 ~ 3,
                                adj_pval < 0.01 ~ 2,
                                adj_pval < 0.05 ~ 1,
                                T ~ 0),
         log_p = -log10(adj_pval))

# a <- 
#   pig_gene_cluster_enrich_hyper %>% 
#   select(1:2, log_p) %>%
#   spread(enhanced_tissues, log_p) %>% 
#   column_to_rownames("cluster") %>% 
#   umap_calc(npcs = 2)
# 
# a %>%
#   as_tibble(rownames = "cluster") %>% 
#   ggplot(aes(V1, V2, label = cluster)) +
#   geom_text()
# 

cluster_ht_res %>% 
  select(enhanced_tissues, cluster, adj_pval, log_p) %>%
  group_by(enhanced_tissues, cluster) %>%
  mutate(log_p = min(log_p, 20)) %>%
  ungroup() %>% 
  mutate(log_p = ifelse(adj_pval < 0.05, 1, 0)) %>%
  select(-adj_pval) %>%
  spread(cluster, log_p, fill = 0) %>% 
  column_to_rownames("enhanced_tissues") %>%
  pheatmap(clustering_method = "ward.D2",
           #cutree_cols = n_distinct(similar_clusters$similar_cluster),
          # annotation_col = similar_clusters %>% 
          #   column_to_rownames("cluster"), 
           #filename = savepath("UMAP similar clusters heatmap.pdf"),
           width = 15, 
           height = 6)

## Dendogram add
cluster_ht_res %>% 
  select(enhanced_tissues, cluster, log_p) %>%
  ggplot(aes(x=cluster,y=enhanced_tissues,size = log_p)) +
  geom_point() +
  theme_bw()
  
  
  ###
  group_by(enhanced_tissues, cluster) %>%
  mutate(log_p = min(log_p, 20)) %>%
  ungroup() %>%
  mutate(log_p = ifelse(log_p < 3, 0, 1)) %>%
  spread(cluster, log_p) %>%
  column_to_rownames("enhanced_tissues") %>%
  pheatmap(clustering_method = "ward.D2")

```

```{r Fischer test 2}
# Cluster data?
# pca_red <-
#   tmm_data %>%
#   #filter(gene_id %in% expressed_genes) %>%
#   spread(key = "sample_ID",  value = "tmm") %>%
#   column_to_rownames("enssscg_id") %>%
#   t() %>%
#   scale() %>%
#   t() %>%
#   pca_calc(npcs = 10)
# 
# n_genes <-
#   pca_red$scores %>%
#   nrow()
# 
# gene_clusters <-
#   pca_red$scores %>%
#   head(n_genes) %>%
#   dist() %>%
#   fastcluster::hclust(method = "ward.D2") %>%
#   cutree(50) %>%
#   enframe("gene_id", "cluster") %>% 
#   left_join(meta_data_gene) %>%
#   left_join(gene_classification)


gene_clusters <- 
  all_partitions %>%
  filter(id == "max pearson medoids 50") %>%
  select(gene, value) %>%
  rename(gene_id = gene, cluster = value) %>%
  left_join(tissue_info, by = c("gene_id" = "ensg_id")) 

n_genes <- length(unique(gene_clusters$gene_id))

fisher_data <- 
  gene_clusters %>%
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues),
                                   "not enriched",
                                   enhanced_tissues),
         enhanced_tissues = factor(enhanced_tissues),
         cluster = factor(cluster)) %>%
  group_by(cluster, enhanced_tissues, .drop = F) %>%
  count() %>%
  group_by(enhanced_tissues) %>%
  mutate(total_specific = sum(n),
         total_not_specific = n_genes - total_specific) %>%
  group_by(cluster) %>%
  mutate(total_in_cluster = sum(n),
         total_not_in_cluster = n_genes - total_in_cluster ) %>%
  ungroup()
 


# Compare to classification (fisher test)
# 
# fisher_data <-
#   gene_clusters %>%
#   mutate(enhanced_tissues = gsub(", ", "_", enhanced_tissues)) %>%
#   separate_rows(enhanced_tissues, sep = ",") %>%
#   mutate(enhanced_tissues = gsub("_", ", ", enhanced_tissues),
#          enhanced_tissues = ifelse(is.na(enhanced_tissues),
#                                    "not enriched",
#                                    enhanced_tissues),
#          enhanced_tissues = factor(enhanced_tissues),
#          cluster = factor(cluster)) %>%
#   group_by(cluster, enhanced_tissues, .drop = F) %>%
#   count() %>%
#   group_by(enhanced_tissues) %>%
#   mutate(total_specific = sum(n),
#          total_not_specific = n_genes - total_specific) %>%
#   group_by(cluster) %>%
#   mutate(total_in_cluster = sum(n),
#          total_not_in_cluster = n_genes - total_in_cluster ) %>%
#   ungroup()
 

fisher_res <-
  fisher_data %>%
  group_by(cluster, enhanced_tissues) %>%
  do({
    g_data <<- .
   
    contingency_matrix <-
      matrix(c(g_data$n,
               g_data$total_in_cluster - g_data$n,
               g_data$total_specific - g_data$n,  
               g_data$total_not_specific - (g_data$total_specific - g_data$n)),
             nrow = 2,
             dimnames =
               list(c("Specific", "Not specific"),
                    c("In cluster", "Not in cluster")))
   
    contingency_matrix %>%
      fisher.test(alternative = "greater") %>%
      broom::tidy()
  }) %>%
  ungroup() %>%
  mutate(p.value = ifelse(p.value == 0, .Machine$double.xmin, p.value),
                          adj_p = p.adjust(p.value, method = "BH")) %>%
  rename(odds_ratio = estimate)


fisher_data %>%
  filter(cluster == 8) %>%
  arrange(-n)

fisher_res %>%
  select(1, 2, adj_p) %>%
  filter(adj_p < 0.05) %>%
  spread(enhanced_tissues, adj_p, fill = 1) %>%
  column_to_rownames("cluster") %>%
  {-log10(.)}  %>%
  pheatmap(color = viridis::magma(100, direction = -1),
           border_color = NA,
           clustering_distance_rows = "binary",
           clustering_distance_cols = "binary",
           clustering_method = "ward.D2")

fisher_res %>%
  filter(adj_p < 0.05) %>%
  select(1, 2, odds_ratio) %>%
  spread(enhanced_tissues, odds_ratio, fill = 0) %>%
  column_to_rownames("cluster") %>%
  pheatmap(color = viridis::magma(100, direction = -1),
           border_color = NA,
           clustering_distance_rows = "binary",
           clustering_distance_cols = "binary",
           clustering_method = "ward.D2")


fisher_res %>%
  filter(adj_p < 0.05) %>%
  select(1, 2, odds_ratio) %>%
  #spread(enhanced_tissues, odds_ratio, fill = 0) %>% 
  ggplot(aes(x=cluster,y=enhanced_tissues,size = odds_ratio, color = enhanced_tissues)) +
  geom_point() +
  theme_bw() +
  scale_color_manual(values = tissue_colors)




cluster_query <-
  function(query_data, clustering, distance_method = "euclidean") {
   
    stopifnot(clustering %in% c("ward.D", "ward.D2", "single", "complete",
                                "average", "mcquitty", "median", "centroid"))
   
    query_data %>%
      select(gene_name, tissue, exp) %>%
      mutate(exp = ifelse(!is.finite(exp), 0, exp)) %>%
      spread(tissue, exp) %>%
      column_to_rownames("gene_name") %>%
     
      {list(gene = dist(., distance_method) %>%
              stats::hclust(method = clustering),
           sample = t(.) %>%
             dist(distance_method) %>%
             stats::hclust(method = clustering))}
    
  }


library(ggdendro)

query_data <-
  
  fisher_res %>% 
  mutate(odds_ratio = ifelse(adj_p < 0.05,
                             odds_ratio, 
                             0)) %>%
  select(gene_name = cluster, tissue = enhanced_tissues, exp = odds_ratio)


query_clustering <-
  query_data %>%
  cluster_query(clustering = "ward.D2", distance_method = "binary")


plot_data <-
  query_data %>%
  mutate(gene_name = factor(gene_name, query_clustering$gene$labels[query_clustering$gene$order]),
         tissue = factor(tissue, query_clustering$sample$labels[query_clustering$sample$order]))

gene_segments <-
  dendro_data(query_clustering$gene)$segments %>%
  as_tibble() %>%
  rename(x = y,
         y = x,
         xend = yend,
         yend = xend) %>%
  mutate(segment_id = row_number()) %>%
  gather(x_type, x_coord, x, xend) %>%
  mutate(x_coord = - scales::rescale(x_coord,
                                     c(0, 5)) + 0.5) %>%
  spread(x_type, x_coord)

sample_segments <-
  dendro_data(query_clustering$sample)$segments %>%
  as_tibble() %>%
  mutate(segment_id = row_number()) %>%
  gather(y_type, y_coord, y, yend) %>%
  mutate(y_coord = scales::rescale(y_coord,
                                   c(0, 5)) +
           n_distinct(query_data$gene_name) + 0.5) %>%
  spread(y_type, y_coord)


plot_data %>%
  left_join(fisher_res, by = c("gene_name"="cluster", "tissue"="enhanced_tissues")) %>%
  filter(exp > 0) %>%
  ggplot(aes(tissue, gene_name, size = exp, fill=tissue)) +
  geom_point(show.legend = T, shape = 21) +
  geom_segment(data = gene_segments,
               aes(x, y, xend = xend, yend = yend),
               inherit.aes = F) +
  geom_segment(data = sample_segments,
               aes(x, y, xend = xend, yend = yend),
               inherit.aes = F) +
  coord_fixed() +
  scale_y_discrete(expand = expansion(0),
                   position = "right") +
  theme(panel.background = element_blank(),
        panel.spacing = unit(0, "mm"),
         plot.margin = unit(c(0,0,0,0), "mm"),
        panel.grid =element_line(color="grey85"),
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        axis.text.x = element_text(angle = 90, vjust = 0.3, hjust = 1),
        legend.position = "right") +
  scale_fill_manual(values = tissue_colors)
p
```

```{r New Fischer function}

### Fisher score
clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value) %>%
  left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
  select(gene,value) %>%
  na.omit() 

clust_list <- set_names(clustering$value,
                        clustering$gene)

tissue_list <- 
  tissue_info %$%
  split(ensg_id, enhanced_tissues)


gene_fischer_test <-
  function(cluster_membership, term_membership, alternative = "greater") {
   
    # cluster_membership <- clust_list 
    # term_membership <- tissue_list 
    n_genes <-
      term_membership %>%
      unlist() %>%
      n_distinct()
   
    fisher_data <-
      cluster_membership %>%
      enframe("gene", "cluster") %>%
      full_join(term_membership %>%
                  map(enframe, "row", "gene") %>%
                  bind_rows(.id = "term")) %>%
      mutate(term = factor(term),
             cluster = factor(cluster)) %>%
      group_by(cluster, term, .drop = F) %>%
      count() %>%
      group_by(term) %>%
      mutate(total_term = sum(n),
             total_nonterm = n_genes - total_term) %>%
      filter(!is.na(cluster) &
               !is.na(term)) %>%
      group_by(cluster) %>%
      mutate(total_in_cluster = sum(n),
             total_not_in_cluster = n_genes - total_in_cluster ) %>%
      ungroup()
       
   
   
    fisher_data %>%
      group_by(cluster, term) %>%
      do({
        g_data <<- .
       
        contingency_matrix <-
          matrix(c(g_data$n,
                   g_data$total_in_cluster - g_data$n,
                   g_data$total_term - g_data$n,  
                   g_data$total_nonterm - (g_data$total_term - g_data$n)),
                 nrow = 2,
                 dimnames =
                   list(c("Term", "Nonterm"),
                        c("In cluster", "Not in cluster")))
       
        contingency_matrix %>%
          fisher.test(alternative = alternative) %>%
          tidy() %>%
          mutate(term_overlap = paste(contingency_matrix[1],
                                      contingency_matrix[1] + contingency_matrix[3],
                                      sep = "/"),
                 cluster_membership = paste(contingency_matrix[1],
                                            contingency_matrix[1] + contingency_matrix[2],
                                            sep = "/"))
      }) %>%
      ungroup() %>%
      mutate(p.value = ifelse(p.value == 0, .Machine$double.xmin, p.value),
             adj_p = p.adjust(p.value, method = "BH")) %>%
      rename(odds_ratio = estimate) %>%
      arrange(adj_p)
  }

FT_medoids_zscore_pearson_70 <- 
  gene_fischer_test(cluster_membership = clust_list, term_membership = tissue_list)

# very slow? compared to the previous




fisher_res <- FT_medoids_zscore_pearson_70 %>% rename(enhanced_tissues = term)


fisher_res %>%
  select(1, 2, adj_p) %>%
  filter(adj_p < 0.05) %>%
  spread(enhanced_tissues, adj_p, fill = 1) %>%
  column_to_rownames("cluster") %>%
  {-log10(.)}  %>%
  pheatmap(color = viridis::magma(100, direction = -1),
           border_color = NA,
           clustering_distance_rows = "binary",
           clustering_distance_cols = "binary",
           clustering_method = "ward.D2")

fisher_res %>%
  filter(adj_p < 0.05) %>%
  select(1, 2, odds_ratio) %>%
  spread(enhanced_tissues, odds_ratio, fill = 0) %>%
  column_to_rownames("cluster") %>%
  pheatmap(color = viridis::magma(100, direction = -1),
           border_color = NA,
           clustering_distance_rows = "binary",
           clustering_distance_cols = "binary",
           clustering_method = "ward.D2")


fisher_res %>%
  filter(adj_p < 0.05) %>%
  select(1, 2, odds_ratio) %>%
  #spread(enhanced_tissues, odds_ratio, fill = 0) %>% 
  ggplot(aes(x=cluster,y=enhanced_tissues,size = odds_ratio, color = enhanced_tissues)) +
  geom_point() +
  theme_bw() +
  scale_color_manual(values = tissue_colors)




cluster_query <-
  function(query_data, clustering, distance_method = "euclidean") {
   
    stopifnot(clustering %in% c("ward.D", "ward.D2", "single", "complete",
                                "average", "mcquitty", "median", "centroid"))
   
    query_data %>%
      select(gene_name, tissue, exp) %>%
      mutate(exp = ifelse(!is.finite(exp), 0, exp)) %>%
      spread(tissue, exp) %>%
      column_to_rownames("gene_name") %>%
     
      {list(gene = dist(., distance_method) %>%
              stats::hclust(method = clustering),
           sample = t(.) %>%
             dist(distance_method) %>%
             stats::hclust(method = clustering))}
    
  }


library(ggdendro)

query_data <-
  
  fisher_res %>% 
  mutate(odds_ratio = ifelse(adj_p < 0.05,
                             odds_ratio, 
                             0)) %>%
  select(gene_name = cluster, tissue = enhanced_tissues, exp = odds_ratio)


query_clustering <-
  query_data %>%
  cluster_query(clustering = "ward.D2", distance_method = "binary")


plot_data <-
  query_data %>%
  mutate(gene_name = factor(gene_name, query_clustering$gene$labels[query_clustering$gene$order]),
         tissue = factor(tissue, query_clustering$sample$labels[query_clustering$sample$order]))

gene_segments <-
  dendro_data(query_clustering$gene)$segments %>%
  as_tibble() %>%
  rename(x = y,
         y = x,
         xend = yend,
         yend = xend) %>%
  mutate(segment_id = row_number()) %>%
  gather(x_type, x_coord, x, xend) %>%
  mutate(x_coord = - scales::rescale(x_coord,
                                     c(0, 5)) + 0.5) %>%
  spread(x_type, x_coord)

sample_segments <-
  dendro_data(query_clustering$sample)$segments %>%
  as_tibble() %>%
  mutate(segment_id = row_number()) %>%
  gather(y_type, y_coord, y, yend) %>%
  mutate(y_coord = scales::rescale(y_coord,
                                   c(0, 5)) +
           n_distinct(query_data$gene_name) + 0.5) %>%
  spread(y_type, y_coord)


plot_data %>%
  left_join(fisher_res, by = c("gene_name"="cluster", "tissue"="enhanced_tissues")) %>%
  filter(exp > 0) %>%
  ggplot(aes(tissue, gene_name, size = exp, fill=tissue)) +
  geom_point(show.legend = T, shape = 21) +
  geom_segment(data = gene_segments,
               aes(x, y, xend = xend, yend = yend),
               inherit.aes = F) +
  geom_segment(data = sample_segments,
               aes(x, y, xend = xend, yend = yend),
               inherit.aes = F) +
  coord_fixed() +
  scale_y_discrete(expand = expansion(0),
                   position = "right") +
  theme(panel.background = element_blank(),
        panel.spacing = unit(0, "mm"),
         plot.margin = unit(c(0,0,0,0), "mm"),
        panel.grid =element_line(color="grey85"),
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        axis.text.x = element_text(angle = 90, vjust = 0.3, hjust = 1),
        legend.position = "right") +
  scale_fill_manual(values = tissue_colors)

```

```{r GOSemSim}
library(GOSemSim)

GO_data <- godata("org.Hs.eg.db", ont = "BP", computeIC=FALSE)

# sim < - mgeneSim(eg, semData = hsGO, measure = "Wang", drop =
# "IEA", combine = "BMA")
# > rownames(sim) < - genes[rownames(sim)]
# > colnames(sim) < - genes[colnames(sim)]

# Docu
# cluster1 <- c("835", "5261","241", "994")
# cluster2 <- c("307", "308", "317", "321", "506", "540", "378", "388", "396")
# clusterSim(cluster1, cluster2, semData=GO_data, measure="Wang")

#12342 original GO terms
# 6156 fitlered

GOs <- GO_data@geneAnno%>% group_by(GO) %>% count() %>% filter(n<500) %>% filter (n>3) %>% pull(GO)

# GO_data@geneAnno %>% pull(ENTREZID) %>% unique()  %>% length() #-> 18866 -> 16940

GO_data@geneAnno <-
  GO_data@geneAnno %>% 
  filter(GO %in% GOs)



# 159740   -> 134047 
clustering <- 
  all_partitions %>%
  filter(id == "max pearson medoids 50") %>%
  select(gene, value) %>%
  left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
  select(entrez,value) %>%
  na.omit() 
#%>%
 # mutate(value = value+1)



num_clusters <- length(unique(clustering$value)) 


cluster_list <- c()
num_clus <- length(unique(clustering$value))
num_genes <- length(clustering$value)
for (i in c(0:num_clus)) {
  genes <- clustering %>%
    filter(value==i) %>%
    pull(entrez) %>%
    as.character()
  cluster_list[[as.character(i)]] <- genes
}



Sys.time()
sim_matrix <- mclusterSim(clusters= cluster_list, semData = GO_data, measure = "Wang") #1.30 h 
Sys.time()

sim_matrix_max_pearson_medoids_50 <- sim_matrix
save(sim_matrix_max_pearson_medoids_50, file = "data/processed/sim_matrix_max_pearson_medoids_50.Rdata")

sim_matrix_zscore_pearson_louvain_68 <- sim_matrix
save(sim_matrix_zscore_pearson_louvain_68, file = "data/processed/sim_matrix_zscore_pearson_louvain_68.Rdata")



sim_matrix_max_pearson_medoids_50 %>% 
  pheatmap()


# Manual matrix
# sim_df <- data.frame()
# 
# Sys.time()
# for (i in (1:num_clusters)) {
#   for(j in (1:num_clusters)) {
#     cluster_i <- clustering %>% filter(value == i) %>% pull(entrez) %>% as.character()
#     cluster_j <-  clustering %>% filter(value == j) %>% pull(entrez)  %>% as.character()
#     sim_value <- clusterSim(cluster_i, cluster_j, semData=GO_data, measure="Wang")
#     sim_df <- bind_rows(sim_df,data.frame(cluster1 = i, cluster2 = j, sim = sim_value)) 
#   }
# }
# 
# Sys.time()

  #11,05
    
# sim_df%>%
#   as_tibble(rownames = "cluster1") %>%    
#   gather(cluster2, sim, -cluster1) %>%
#   filter (cluster1 < cluster2) #%>%
 # mutate(cluster = cluster)



# mclsterSim

cluster_list <- c()
num_clus <- length(unique(clustering$value))
num_genes <- length(clustering$value)
for (i in c(0:num_clus)) {
  genes <- clustering %>%
    filter(value==i) %>%
    pull(entrez) %>%
    as.character()
  cluster_list[[as.character(i)]] <- genes
}

# Repeat with 
#max(GO_filtered %>% group_by(GOID) %>% count() %>% pull(n))

library(GOSemSim)
Sys.time()
sim_matrix <- mclusterSim(clusters= cluster_list, semData = GO_filtered %>% as.data.frame(), measure = "Wang") #1.30 h 
Sys.time()

d <- godata(GO_filtered %>% as.data.frame(), keytype = "ENTREZID", ont="BP", computeIC=FALSE)
class(d)

godata(OrgDb = GO_filtered %>% as.data.frame(), ont="BP")



class(GO[["0.01"]])
sim_matrix %>%
  pheatmap(border_color = NA, 
            clustering_method = "ward.D2", 
           color = colorRampPalette(c("white","#00846b"))(50)) #cutree_rows = 5,
save(sim_matrix, file="data/processed/sim_matrix_zscore_euclidean_fasthclust_50.Rdata")
```

```{r GO network}

```






```{r UMAP visualization - tests with Max}
load("data/processed/data_scaled.Rdata")
load("data/processed/pcas.Rdata")
load("data/processed/pcas.Rdata")
load("data/processed/louvain_multk.Rdata")
load("data/processed/medoids_multk.Rdata")

clustering <- louvain_multk[[1]][[1]]$cluster


## UMAP

clus_umap <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:10)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2) 

clus_umap %>% umap_plot()
                        
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, fill = cluster, color = cluster)) +
  geom_point(size = 0.1) +
  stat_density2d(bins = 2, h = 2) +
  theme_bw()


geom_density_2d(bins = 2, )
  geom_density2d_filled(bins = 20)


## Hexagonal
  getmode <- function(v) {
  uniqv <- unique(v)
  as.character(uniqv[which.max(tabulate(match(v, uniqv)))])
}
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, z = cluster, group = 1)) +
  stat_summary_hex(fun = getmode,
                   bins = 100,
                   show.legend = F)
## Density
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, group=cluster)) +
  geom_density2d_filled(h = 10) +
  facet_wrap(~cluster)


#######

density_settings <-
  expand.grid(n = c(100,200,300,400),
              h = c(0.02, 0.05, 0.1, 0.2)) %>% 
  as_tibble()


clus_umap_density_all <- 
  density_settings %>% 
  group_by_all() %>% 
  do({
    h <- .$h
    n <- .$n
    
    clus_umap %>% 
      left_join(clustering %>% 
                  dplyr::rename(features = 1, 
                                cluster = 2)) %>% 
      mutate(cluster = factor(cluster)) %>%
      group_by(cluster) %>%
      do ({
        data <- .
        data %$%
          MASS::kde2d(V1, V2, n = n, h = h, lims = c(range(clus_umap$V1), range(clus_umap$V2))) %>%
          {.$z} %>%
          as_tibble(rownames = "x") %>%
          gather(y, density, -1) %>%
          mutate(y = gsub("V", "", y),
                 x = as.numeric(x),
                 y = as.numeric(y))
      })
  })
clus_umap_density <-
  clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  group_by(cluster) %>%
  do ({
    data <- .
    data %$%
      MASS::kde2d(V1, V2, n = 300, h = 0.1, lims = c(range(clus_umap$V1), range(clus_umap$V2))) %>%
      {.$z} %>%
      as_tibble(rownames = "x") %>%
      gather(y, density, -1) %>%
      mutate(y = gsub("V", "", y),
             x = as.numeric(x),
             y = as.numeric(y))
  })

clus_umap_density%>%
 # filter(n== 400, h == 0.05) %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.01) %>%
  ggplot(aes(x,y,fill = cluster, alpha = density)) +
 # geom_encircle() +
  geom_point(data = clus_umap %>% 
               left_join(clustering, by = c("features" = "gene")), #%>% filter(value ==2),
             aes(V1, V2),
             size = 0.07,
             inherit.aes = F) + 
  geom_tile() +
  theme_bw() +
  scale_alpha_continuous(limits = c(0, 0.3), range = c(0.05, 0.6))
  #facet_wrap(~cluster)#+
 # facet_grid(h ~ n)


clus_umap_density %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.01) %>%
  ggplot(aes(x,y,fill = cluster, alpha = density)) 
  
clus_umap %>%
  ggplot() +
  geom_point(aes(V1, V2),
             size = 0.1,
             inherit.aes = F) 

clus_umap_density %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.005) %>%
  ggplot(aes(x,y,fill = cluster)) +
  geom_tile()+
  theme_bw() +
  scale_fill_manual(values = pal)

clus_umap %>%
  left_join(clustering, by = c("features" = "gene")) %>%
  ggplot(aes(V1,V2,color=as.factor(value))) +
  geom_point(size=0.6) +
  theme_bw() +
  scale_color_manual(values = pal) #+
  oose#facet_wrap(~value)

library(pals)
pal <- colorRampPalette(ocean.curl(60))(60)


clus_umap_density %>%
  filter(density > 0.1) %>% 
  group_by(x, y) %>%
  count %>% 
  group_by(n) %>% 
  count


# clus_umap_density %>%
#   filter(density>0.01) 
#   ggplot(aes(density)) +
#   geom_density() 

```

```{Visual evaluation - test UMAP distance metrics (Louvain / Leiden)}

umap <- scaled[[1]]$scaled %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2)

umap$umap %>%
  left_join(res_kmeansfast[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() 

umap_corr <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "correlation")

umap_manhattan <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "manhattan")

umap_euclidean <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "euclidean")

umaps <- list(umap_euclidean,umap_manhattan,umap_corr)

save(umaps,
     file = "data/processed/umaps.Rdata")


load("data/processed/results_leiden.Rdata")
load("data/processed/results_louvain.Rdata")

# 2D clustering
d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)
c <- clust(dist = d$distance, k = 100, m = "fastkmeans", genes = gene_names, 
           id = d$id)

umap$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 


d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)

c <- clust(dist = d$distance, k = 100, m = "louvain", genes = gene_names, 
           id = d$id)

umaps[[3]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 



load("data/processed/umaps.Rdata")


df2 <- data.frame(value = as.numeric(medoids$cluster), gene = gene_names)
umaps[[1]]$umap %>%
  left_join(df2, by = c("features" = "gene")) %>%
  ggplot(aes(V1, V2)) +
   geom_hex(aes(fill = 1), fill = "gray90") +
  theme_minimal() +
  geom_encircle(aes(fill = value,
                     color = value),
                 alpha = 0.1,
                 s_shape=1,
                 expand=0,
                 # color = "white",
                 show.legend = F) +
   geom_encircle(aes(color = value),
                 fill = NA,
                 alpha = 0.7,
                 s_shape=1,
                 expand=0,
                 # color = "white",
                 show.legend = F)# +
#   geom_text(data = . %>%
#               select(merged_cluster, mean_V1, mean_V2, n) %>%
#               distinct(),
#             aes(mean_V1, mean_V2, label = merged_cluster),
#             show.legend = F)
  
  
load("data/processed/results_louvain.Rdata")

df2 <- data.frame(value = as.numeric(medoids$cluster), gene = gene_names)

umaps[[1]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  head(200) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal)  + 
    scale_fill_manual(values = pal)  + 

  #geom_density_2d()
#  stat_density_2d(aes(fill = value), geom="polygon")+ 
# stat_ellipse()
   geom_encircle(aes(group = value, fill = as.factor(value)),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                #fill = "gray",
               # color = "black",
                show.legend = F)  +
    geom_point(show.legend = F, size = 1) 
  #   ggplot(aes(V1, V2)) +
#   geom_hex(aes(fill = 1),
#            data = plot_data,
#            fill = "gray90",
#            bins = plot_bins) +
#   geom_encircle(aes(fill = merged_cluster,
#                     color = merged_cluster),
#                 alpha = 0.1,
#                 s_shape=1,
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   geom_encircle(aes(color = merged_cluster),
#                 fill = NA,
#                 alpha = 0.7,
#                 s_shape=1,
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   geom_text(data = . %>%
#               select(merged_cluster, mean_V1, mean_V2, n) %>%
#               distinct(),
#             aes(mean_V1, mean_V2, label = merged_cluster),
#             show.legend = F)
  
  

pal <- colorRampPalette(ocean.curl(50))(50)

  geom_encircle(aes(group = value),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                fill = "gray",
                color = "black",
                show.legend = F) 
#+ scale_color_manual(values = pal) 
```

```{r Cluster visualziation - UMAP}


umap_louv %>% 
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  filter(enhanced_tissues == "retina") %>%
  ggplot(aes(V1,V2)) +
  geom_point(data = t, aes(V1,V2), color = "grey", inherit.aes = F) +
  geom_point(show.legend = F, color = tissue_colors[["retina"]]) +
  
  theme_minimal() +
  #scale_color_manual(values = tissue_colors) +
  coord_fixed()

# For all tissues

tissues <- unique(tissue_info$enhanced_tissues)
t

plots <- 
  lapply(unique(tissue_info$enhanced_tissues),
         function(tissue) {
           umap_louv %>%
             left_join(tissue_info, by= c("features" = "ensg_id")) %>%
             filter(enhanced_tissues == tissue) %>%
            ggplot(aes(V1,V2)) +
              geom_point(data = t, aes(V1,V2), color = "grey", inherit.aes = F) +
              geom_point(show.legend = F, color = tissue_colors[[tissue]]) +
              theme_minimal() +
              coord_fixed() +
             ggtitle(tissue)
         })

pdf("umaps_tissues.pdf")
plots
dev.off()


# Tissues for each cluster

to_plot <- 
  umap_louv %>%
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  left_join(clustering, by = c("features" = "gene")) %>%
  mutate (value = value+1)

t <-
  to_plot %>%
  select(-value,-enhanced_tissues)

tissue_colors[["NA"]] <- "grey30"

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(67)
plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2,color=enhanced_tissues)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5) +
             theme_minimal() +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) + 
             scale_color_manual(values = tissue_colors) 
         })

pdf("us.pdf")
plots
dev.off()


```
