---
title: "HPA-gene-clustering"
author: "Maria Bueno Alvez"
date: "2/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r load packages and functions, include=FALSE}

# Packages
library(tidyverse)
library(magrittr)
library(broom)
library(xlsx)
library(readxl)
library(writexl)
library(readr)
library(NOISeq)
library(uwot)
library(pcaMethods)
library(MASS)
library(factoextra)
library(cluster)
library(kohonen)
library(ClusterR)
library(flashClust)
library(Seurat)
library(dbscan)
library(kmed)
library(aricode)
library(pheatmap)
library(clValid)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(annotate)
library(GO.db)
library(clusterProfiler)
library(ClusterJudge)
library(mixOmics)
library(ggthemes)
library(ggrepel)
library(ggsci)
library(scales)
library(treemapify)
library(rrvgo)
library(patchwork)
library(ggplotify)
library(ggdendro)
library(viridis)
library(concaveman)
library(sf)
library(ggalt)
library(ggforce)
library(ggiraphExtra)
library(ggradar)
library(colorspace)


# Wd, scripts
setwd("/Users/maria.bueno/Desktop/HPA-gene-clustering")
source("scripts/functions_preprocessing.R")
source("scripts/functions_dimensionalityreduction.R")
source("scripts/functions_clustering.R")
source("scripts/functions_evaluation.R")
source("scripts/functions_annotation.R")

map <- purrr::map
select <- dplyr::select
pca <- pcaMethods::pca
rank <- base::rank
filter <- dplyr::filter

```

```{r load data, include=FALSE}

# Data in tpm
data <- 
  read_delim("data/pig_filtered_sample_data.tab", delim = "\t")

# Specificity and distribution gene annotation 
gene_info <-
  read_delim("data/meta/pig_consensus_category_piggenome_92.tsv", delim = "\t")

# Genes annotated as "not_detected"
genes_to_filter <-
  gene_info %>%
  filter(distribution_category == "not detected") %>%
  pull(ensg_id) 

# Enhanced tissues for detected genes
tissue_info <-
  gene_info %>%
  filter(!ensg_id %in% genes_to_filter) %>%
  separate_rows("enhanced_tissues", sep = ',') %>%
  select(ensg_id,enhanced_tissues) %>%
  remove_missing()

# Gene description table
geninfo <- 
  read_delim("data/meta/geninfo_92.tsv", delim = "\t")

# Tissue description table
tissues <- data  %>% 
  mutate(tissue = gsub('.{2}$', '', sample_ID)) %$%
  tissue %>% 
  unique()

tissue_mapping <-
  read_delim("data/meta/tissue_mapping.tsv", delim = "\t") %>% 
  select(tissue,tissue_name,consensus_tissue_name,organ_name) %>% 
  left_join(colors %>% select(tissue_name,tissue_color)) %>% 
  filter(tissue %in% tissues)

# Information of orthologs
ortholog_info <- 
  read_delim("data/meta/gene_orthologs.csv", delim = ",") %>%
  select(enssscg_id, ensg_id) %>%
  dplyr::mutate(entrez = mapIds(org.Hs.eg.db,
                  keys= ensg_id, 
                  column="ENTREZID",
                  keytype="ENSEMBL"))

# GO annotation
GOterms <- read_tsv("data/meta/E92_all_GOterms.txt")

```

```{r plot theme & color palettes}
theme_simple <- 
  theme_bw() + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

theme_simple_2 <-
  theme(panel.background = element_rect(fill = NA, colour = NA),
        plot.background = element_rect(fill = NA, color = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        axis.line = element_line(colour="black",size=0.5))


# Color palettes
colors <-
  read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors <- colors  %>%
  select(organ_color, consensus_tissue_name) %>%
  distinct() %$%
  set_names(organ_color, consensus_tissue_name)

gene_category_pal <-
  c("tissue enriched" = "#e41a1c",
    "group enriched" = "#FF9D00",
    "tissue enhanced" = "#984ea3",
    "low tissue specificity" = "grey40",
   
    "detected in all" = "#253494",
    "detected in many" = "#2c7fb8",
    "detected in some" = "#41b6c4",
    "detected in single" = "#a1dab4",
   
    "not detected" = "grey",
    "not detected " = "grey",
   
    "Tissue enriched" = "#e41a1c",
    "Group enriched" = "#FF9D00",
    "Tissue enhanced" = "#984ea3",
    "Low tissue specificity" = "grey40",
   
    "Detected in all" = "#253494",
    "Detected in many" = "#2c7fb8",
    "Detected in some" = "#41b6c4",
    "Detected in single" = "#a1dab4",
   
    "Not detected" = "grey",
    "Not detected " = "grey")

cluster_palette <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(70)
names(cluster_palette) <- c(1:70)

```


#1- Data pre-processing

```{r pre-procesing}
# Filtering genes
filtered_data <-
  data %>% 
  filter(!enssscg_id %in% genes_to_filter)

# Save gene names
gene_names <- 
  filtered_data %>%
  select(enssscg_id, sample_ID, tpm) %>%
  spread(key = sample_ID , value = tpm) %$%
  enssscg_id

# Normalize data
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)

# Scale data
scaled <- list("zscore","max") %>% 
  map(~data_scaling(df = tmm_data, col_value = "tmm",
                    col_gene = "enssscg_id", col_sample = "sample_ID",
                    log_transform = F, m = .x))

```

#2- Dimensionality reduction

```{r dimensionality reduction}

# Dimensionality reduction of the two scaled datasets (PCA, 200 components)
pcas <- scaled %>%
  map(~pca_calc(data = .x$scaled, npcs = 200, m = .x$method))

```

#3- Distance calculation

```{r distance calculation}

# Distance calculation for the two datasets (Euclidean & Pearson)
distances <- lapply(list("euclidean","pearson"), function(i) 
  map(pcas, ~dist_calc(df = .x$scores, comp = 30, m = i, 
                       id = .x$method)))

```

#4- Clustering

```{r Clustering analysis}

# Run hierarchical clustering with k = 50, 70, 90, 110, 130 and 150
hclust_multk <- 
  lapply(list(50,70,90,110,130,150), function(i)
    map(d, ~clust(dist = .x$distance, 
                  k = as.numeric(i), 
                  m = "fasthclust", 
                  genes = gene_names, 
                  id = .x$id, 
                  mult_k = T)))

# Run k-means clustering with k = 50, 70, 90, 110, 130 and 150
kmeans_multk <- 
  lapply(list(50,70,90,110,130,150), function(i)
    map(d, ~clust(dist = .x$distance, 
                  k = as.numeric(i), 
                  m = "fastkmeans", 
                  genes = gene_names, 
                  id = .x$id, 
                  mult_k = T)))

# Run k-medoids clustering with k = 50, 70, 90, 110, 130 and 150
medoids_multk <- 
  lapply(list(50,70,90,110,130,150), function(i)
    map(d, ~clust(dist = .x$distance, 
                  k = as.numeric(i), 
                  m = "medoids", 
                  genes = gene_names, 
                  id = .x$id, 
                  mult_k = T)))

# Run louvain clustering with resolution parameter = 4, 6, 8, 10, 12 and 14
louvain_multk <- 
  lapply(list(4,6,8,10,12,14), function(i)
    map(d, ~clust(dist = .x$distance, 
                  r = as.numeric(i), 
                  m = "louvain", 
                  genes = gene_names, 
                  id = .x$id)))

# Run leiden clustering with resolution parameter = 4, 6, 8, 10, 12 and 14
leiden_multk <- 
  lapply(list(4,6,8,10,12,14), function(i)
    map(d, ~clust(dist = .x$distance, 
                  r = as.numeric(i), 
                  m = "leiden", 
                  genes = gene_names, 
                  id = .x$id)))

# Run SOM clustering with k = 50, 70, 90, 110, 130 and 150
medoids_som <- 
  lapply(list(50,70,90,110,130,150), function(i)
    map(d, ~clust(dist = .x$distance, 
                  k = as.numeric(i), 
                  m = "SOM", 
                  genes = gene_names, 
                  id = .x$id, 
                  mult_k = T)))
```


#5- Evaluation
```{r Retrieve partitions for ARI}

# load("data/processed/kmeans_multk.Rdata")
# load("data/processed/medoids_multk.Rdata")
# load("data/processed/hclust_multk.Rdata")
# load("data/processed/louvain_multk.Rdata")
# load("data/processed/leiden_multk.Rdata")
# load("data/processed/som_multk.Rdata")

all_dfs <- map(list(kmeans_multk, medoids_multk, hclust_multk, 
                    louvain_multk, leidenn_multk, som_multk),
               ~to_df(.x))


all_partitions <- merge_all_df(all_dfs)

# save(all_partitions, file ="data/processed/all_partitions.Rdata")

# For all
clusterings <- unique(all_partitions$id)

res_ari <- matrix(, nrow = 168, ncol = 168)
rownames(res_ari) <- c(clusterings)
colnames(res_ari) <- c(clusterings)

# Calculate ARI matrix
for (i in clusterings) {
  for (j in clusterings) {
    a <- ARI(all_partitions %>% filter(id == i) %>% pull(value), 
             all_partitions %>% filter(id == j) %>% pull(value))
    res_ari[i,j] <- a
    }
  }

# save(res_ari, file="data/processed/res_ari.Rdata")
# load("data/processed/res_ari.Rdata")

annotation <-
  data.frame(id = all$id %>% unique()) %>%
  group_by(id) %>%# rename(id = all.id) %>%
  mutate(scale_dist = paste(strsplit(id, " ")[[1]][1], strsplit(id, " ")[[1]][2]),
         clust = strsplit(id, " ")[[1]][3],
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  column_to_rownames("id")

p <- pal_npg("nrc")(6)
clust <- c(p[3],p[1],p[4],p[6],p[2],p[5]) 
names(clust) = c("fasthclust", "fastkmeans", "leiden", "louvain", "medoids", "SOM")

dist = c("#613659", "#81599F")
names(dist) = c("euclidean", "pearson")

scale = c("#2F7604", "#9DCC5F")
names(scale) = c("zscore","max")

k = colorRampPalette(c("white","#FF9663"))(100)

ann_colors = list(clust = clust, scale = scale, dist = dist, k=k)


# Exclude random clustering
ids <- rownames(res_ari)
ids <- ids[!(grepl("random", ids))]
res_ari [ids,ids] %>%
  pheatmap(fontsize = 8, 
           border_color = NA, 
           clustering_method = "ward.D2", 
           annotation_row  = annotation,
           show_rownames = F,
           show_colnames = F, 
           annotation_colors = ann_colors)


# Filtered
best_partitions <-best_k_of_each$id %>% unique()

filtered <-
  all_partitions %>%
  filter(id %in% best_partitions)

res_ari_filtered <- matrix(, nrow = 24, ncol = 24)
rownames(res_ari_filtered) <- c(best_partitions)
colnames(res_ari_filtered) <- c(best_partitions)

# Calculate ARI matrix
for (i in best_partitions) {
  for (j in best_partitions) {
    a <- ARI(filtered %>% filter(id == i) %>% pull(value), 
             filtered %>% filter(id == j) %>% pull(value))
    res_ari_filtered[i,j] <- a
    }
  }

# save(res_ari_filtered, file="data/processed/res_ari_filtered.Rdata")
# load("data/processed/res_ari_filtered.Rdata")

annotation_filtered <-
  data.frame(id = best_partitions) %>%
  group_by(id) %>%# rename(id = all.id) %>%
  mutate(scale_dist = paste(strsplit(id, " ")[[1]][1], strsplit(id, " ")[[1]][2]),
         clust = strsplit(id, " ")[[1]][3],
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  column_to_rownames("id")

# Exclude random clustering
res_ari_filtered %>%
  pheatmap(fontsize = 8, 
           border_color = NA, 
           clustering_method = "ward.D2", 
           annotation_row  = annotation,
           show_rownames = F,
           show_colnames = F,
           annotation_colors = ann_colors)

```

```{r Evaluation (connectivity, dunn, asw, BHI, BP_enriched, time)}

# Evaluation (repeat for each clustering method)
eval_som_multk <- lapply(list(1,2,3,4,5,6), function(i) 
  map(som_multk[[i]], 
      ~eval(clustering = .x, 
            genes= ortholog_info,
            dis = distances
      )))

eval_som_df <- eval_to_df(eval_som_multk)

all_som <- add_clusterings(df = eval_som_df$res, 
                        dist = d, 
                        pca = pcas[[1]],
                        ref_clust = som_multk[[1]][[1]]$cluster,
                        gene_names = gene_names, 
                        clust_m = "SOM",
                        orthologs = ortholog_info)

# save(eval_som_multk,eval_som_df,all_som, file = "data/processed/evaluation_leiden.Rdata")


# Merge evaluation results
all_eval <- 
  bind_rows(all_kmeans,
            all_medoids,
            all_hclust,
            all_louvain,
            all_leiden,
            all_som)


# Plot evaluation results 
eval_to_plot <- 
  all_eval %>%
  filter(!grepl("2D",clustering_id)) %>%
  group_by(clustering_id) %>%
  rename(id = clustering_id) %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))


eval_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  }) %>%
  select(id,clustering,k, BHI_index) %>%
  ggplot(aes(x=id, y = BHI_index, fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~(clustering), scale="free") +
  scale_fill_viridis(discrete = F)+
  labs(y = "Value", x = "Clustering strategy")

```

```{r Evaluations based on enrichments (clusterProfiler)}

# Calculate enrichments (repeat for each clustering)

#load("data/processed/som_multk.Rdata")

df_som <- 
  to_df(kmeans_som, add_random = T)

enrichments_kmeans <- df_som %>%
  map(~enrichments_df(df = .x, grouping = "id", orthologs = ortholog_info))

#
enrichment_scores_kmeans <-
  map(enrichments_kmeans,
  ~enrichment_scores(df = .x, grouping = "id")) 
 

# Create dataframe with enrichment scores for all clusterings

all_bio <- bind_rows(c(enrichment_scores_kmeans, enrichment_scores_medoids, enrichment_scores_hclust, 
                      enrichment_scores_louvain, enrichment_scores_leiden, enrichment_scores_som))

#save(all_bio, file="data/processed/all_bio.Rdata")

# Plot results
bio_to_plot <- 
  all_bio %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))


test <- 
  bio_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    #m <- unique(data$method)
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  })

test %>%
  select(-annotability) %>%
  ggplot(aes(x=id, y = min_significance, fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~(clustering), scale="free") +
 scale_fill_viridis(discrete = F)+
  labs(y = "Value", x = "Clustering strategy")


```

```{r Evaluation based on Mutual Information}

gene_ids <- ortholog_info %>% filter(enssscg_id %in% gene_names) %>% pull(ensg_id)
names(gene_ids) <- c()

GO_terms <- 
  GOterms %>%
  filter(`GO domain` == "biological_process") %>%
  select(ID = `Gene stable ID`, GOID = `GO term accession`) %>%
  na.omit() %>%
  distinct()

terms_to_keep <- 
  GO_terms %>%
  group_by(GOID) %>%
  count() %>%
  filter(n < 500) %>%
  filter(n > 2) %>%
  pull(GOID)

GO_filtered <- 
  GO_terms %>%
  filter(GOID %in% terms_to_keep)

#save(GO_filtered, file = "data/processed/GO_filtered.Rdata")

# Calculate mutual information z-score for each of the 144 clustering strategies
MI_df <- 
  all_partitions %>%
  group_by(id) %>%
  do ({
    data <- .
    id <- unique(data$id)
    c <- data %>%
      left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
      select(ensg_id,value) %>%
      na.omit()
    
    clusters_pig <- as.integer(c$value)
    names(clusters_pig) <- c$ensg_id

    dat <- 
      data.frame(ID = names(clusters_pig),
                 cluster = as.numeric(clusters_pig)) %>% 
      left_join(GO_filtered)
    
    randoms <- c()
    for (i in c(1:50)) {
      dat_random <- 
        dat %>%
        mutate(cluster = sample(cluster))
      MI_pig_random <- mutinformation(dat_random$cluster,dat_random$GOID)
      randoms <- c(randoms, MI_pig_random)
    }

    MI_pig <- mutinformation(dat$cluster,dat$GOID)
    MI_pig_random <- mean(randoms)
    zscore <- (MI_pig - MI_pig_random)/var(randoms)
    
    data.frame(id = id, MI = MI_pig, MI_random = MI_pig_random, 
               MI_random_min = min(randoms), MI_random_max = max(randoms),
               MI_random_var = sd(randoms), MI_zscore = zscore)
  })
  

MI_to_plot <- 
  MI_df %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))


MI <- 
  MI_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    #m <- unique(data$method)
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  })


MI <- MI %>% 
  ungroup %>%
  mutate(MI_zscore = (MI - MI_random)/sqrt(MI_random_var)) 

MI %>%
  ggplot(aes(x=id, y = MI, fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 6)) +
  facet_wrap(~(clustering), scale="free_x") +
 scale_fill_viridis(discrete = F)+
  labs(y = "Value", x = "Clustering strategy")

#save(MI, file = "data/processed/MI.Rdata")

```

```{r Merge evaluations}

#load("data/processed/all_eval.Rdata")
#load("data/processed/MI.Rdata")
#load("data/processed/all_bio.Rdata")

# Merge all measures
all <- 
  all_eval %>% 
  rename(id = clustering_id) %>%
  select(id, connectivity_index, avg_silhouette_width,time) %>%
  left_join(MI %>% ungroup() %>% select(id,MI_zscore)) %>%
  left_join(all_bio %>% select(-min_significance)) %>%
  na.omit() %>%
  gather(key = measure, value = value, connectivity_index:annotability) 

all$value <- as.numeric(all$value)

# Annotation & color palettes
annotation <-
  data.frame(id = all$id %>% unique()) %>%
  group_by(id) %>%
  mutate(scale =strsplit(id, " ")[[1]][1],
         dist = strsplit(id, " ")[[1]][2],
         clust = strsplit(id, " ")[[1]][3],
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  column_to_rownames("id")


p <- pal_npg("nrc")(6)
clust <- c(p[3],p[1],p[4],p[6],p[2],p[5]) 
names(clust) = c("fasthclust", "fastkmeans", "leiden", "louvain", "medoids", "SOM")

dist = c("#180F34", "#944B80")
names(dist) = c("euclidean", "pearson")

scale = c("#D35327", "#F99E35")
names(scale) = c("zscore","max")

k = colorRampPalette(c("white","#95C971"))(100)

ann_colors = list(clust = clust, scale = scale, dist = dist, k=k)


# Time plot
all_eval %>% 
  ggplot(aes(time/60,clustering, fill = clustering)) +
  geom_boxplot(show.legend = F)+
  theme_simple +
  xlab("Time (min)") +
  ylab("Clustering algorithm")+
  scale_fill_manual(values=ann_colors$clust)


time_df <-
  all_eval %>% 
  filter(!(str_detect(clustering_id, pattern = "random"))) %>% 
    filter(!(str_detect(clustering_id, pattern = "2D"))) 

time_df$clustering <- factor(time_df$clustering, levels = names(ann_colors$clust))
time_df$clustering_id <- factor(time_df$clustering_id, levels = unique(time_df$clustering_id))

time_df %>% 
  ggplot(aes(x =  time/60, y = clustering_id, fill = clustering)) +
  geom_bar(stat = "identity")+ # show.legend = F
  theme_simple +
  xlab("Time (s)") +
  ylab("Clustering algorithm")+
  scale_fill_manual(values=clust) +
   theme(axis.text.y = element_text( size = 5)) +
scale_y_discrete(limits = rev(levels(time_df$clustering_id))) 


# Evaluation plots (repeat for each clustering)
all %>% 
  group_by(id) %>%
  mutate(scale_dist = paste(strsplit(id, " ")[[1]][1], strsplit(id, " ")[[1]][2]),
         clust = strsplit(id, " ")[[1]][3],
         k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  ungroup() %>% 
  filter(clust == "louvain") %>% 
  filter(!measure == "time") %>% 
  ggplot(aes(x=as.numeric(k), y=value, group =1, color = as.factor(measure))) +
  geom_point(show.legend = F) +
  geom_line(show.legend = F) +
  #theme_minimal() +
  theme_simple + 
  #ggtitle("Hierarchical clustering") +
  facet_wrap(scale_dist~measure, scales="free_y") +
  theme(text = element_text(size=10)) + 
  labs(y = "Value", x = "Number of clusters") +
  scale_color_viridis(discrete = T)

```

```{r Ranking 1 - all -> best k}

measures <- c("connectivity_index", "avg_silhouette_width", "MI_zscore","annotability")

all_ranked <- 
  all %>%
  ungroup() %>%
  scaled_rank(measures=measures)

# Sorted by average
all_ranked %>% 
 select(-value) %>% 
  spread(key=measure, value=scaled_value) %>% 
  select(id,connectivity_index, avg_silhouette_width, MI_zscore, annotability,avg_value) %>% 
  arrange(-avg_value) %>% 
  column_to_rownames("id") %>%
  pheatmap(cluster_cols = F, cluster_rows = F, fontsize = 8, border_color = NA, 
           color = colorRampPalette(c("white","#426764"))(100),
           annotation_row  = annotation, show_rownames = F,
           annotation_colors = ann_colors)

a <- all_ranked %>% 
 select(-value) %>% 
  spread(key=measure, value=scaled_value)

# Clustered
all_ranked %>% 
 select(-value) %>% 
  spread(key=measure, value=scaled_value) %>% 
  select(id,connectivity_index, avg_silhouette_width, MI_zscore, annotability,avg_value) %>% 
  arrange(avg_value) %>%
  column_to_rownames("id") %>%
  pheatmap(cluster_cols = F, fontsize = 8, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("white","#00846b"))(100),
           annotation_row  = annotation,
           show_rownames = F,
           annotation_colors = ann_colors)


rank_k <- 
  all_ranked %>%
  group_by(id) %>%
  mutate(scale_dist_clust = paste(strsplit(id, " ")[[1]][1], strsplit(id, " ")[[1]][2], strsplit(id, " ")[[1]][3]),
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  ungroup() %>% 
  group_by(scale_dist_clust) %>%
  top_n(n = 1, wt = avg_value)
  
best_k <- 
  rank_k %>% 
  group_by(scale_dist_clust) %>% 
  scaled_rank(measures=measures) %>% 
  group_by(scale_dist_clust) %>% 
  top_n(n=1, wt=avg_value) 

best_k %>%
  select(-value) %>%
  spread(key=measure, value=scaled_value) %>% 
  ungroup() %>% 
  select(-scale_dist_clust,-k) %>% 
  column_to_rownames("id") %>%
  pheatmap(cluster_cols = F, fontsize = 8, border_color = NA, 
           cutree_rows = 3, clustering_method = "ward.D2", 
           color = colorRampPalette(c("white","#00846b"))(100),
           annotation_row  = annotation,
           show_rownames = F,
           annotation_colors = ann_colors)

 
best_clust <- 
best_k %>%
  mutate(clust = strsplit(id, " ")[[1]][3]) %>%
  ungroup() %>% 
  group_by(clust) %>% 
  scaled_rank(measures=measures) %>% 
  group_by(clust) %>% 
  top_n(n=1, wt=avg_value) 

best_clust %>%
  select(-value) %>% 
  spread(key= measure, value = scaled_value) %>% 
  arrange(-avg_value)

final_rank <- 
  best_clust %>%
  ungroup() %>% 
scaled_rank(measures=measures) %>%
  select(-value) %>% 
  spread(key= measure, value = scaled_value) %>% 
  arrange(-avg_value)

best_ids_clust <- final_rank %>% pull(id)

```

```{r Radar plots}

data_radar <- 
  all_ranked %>%
  group_by(id) %>%
  mutate(clust = strsplit(id, " ")[[1]][3]) %>% 
  ungroup() %>% 
  group_by(clust) %>% 
  top_n(n=1,wt=avg_value) %>% 
  ungroup() %>% 
  select(id,measure,scaled_value) %>% 
  spread(key=measure, value=scaled_value) %>% 
  left_join(all_ranked %>%  select(id,avg_value) %>% unique()) 

data_radar$id <- factor(data_radar$id, levels = c("max pearson fasthclust 50","zscore pearson fastkmeans 70",
                                                  "zscore euclidean leiden 73","zscore pearson louvain 68",
                                                  "zscore pearson medoids 70", "zscore pearson SOM 50"))
data_radar <- data_radar %>% arrange(id)

ann_transp <- unlist(map(ann_colors$clust, ~adjust_transparency(.x, alpha = 0.007)))


c <- ann_colors$clust
names(c) <-c("max pearson fasthclust 50","zscore pearson fastkmeans 70",
             "zscore euclidean leiden 73","zscore pearson louvain 68",
             "zscore pearson medoids 70", "zscore pearson SOM 50")

c2 <- ann_transp
names(c2) <-c("max pearson fasthclust 50","zscore pearson fastkmeans 70",
              "zscore euclidean leiden 73","zscore pearson louvain 68",
              "zscore pearson medoids 70", "zscore pearson SOM 50")




# Radar plot for the best strategy from each algorithm 
ggRadar(data_radar,scales = "free", rescale = F, aes(color=id,group=id), alpha = 0.12) +
  geom_line(size=0.8, show.legend = F) +
  theme_minimal() +
  scale_color_manual(values = c) +
  scale_fill_manual(values = c2) +
  ylim(0.3, 1)


# Radar plot per clustering
ggRadar(data_radar,scales = "fixed",rescale=F, aes(facet = id,color=id), alpha = 0.3) +
  theme_minimal() +
  geom_line(size=0.8) +
  scale_color_manual(values = c) +
  scale_fill_manual(values = c2) +
  ylim(0, 1)

#  Radar plot per measure
ggRadar(data_radar %>% column_to_rownames("id") %>%  t() %>% as.data.frame() %>%  rownames_to_column("measure"),
        scales = "fixed",rescale=F, aes(facet = measure,color=measure), alpha = 0.3) +
  theme_minimal() +
  geom_line(size=0.8) +
  ylim(0, 1)

```

```{r Correlation of evaluation metrics}

# Correlation heatmap
cor_matrix <- 
  all_ranked %>%
  select(id,measure,scaled_value) %>% 
  spread(measure,scaled_value) %>%
  column_to_rownames("id") %>% 
  cor()

cor_matrix %>% 
  pheatmap(viridis(20, direction = -1, option = "B"),
           border_color = NA)

# Pairwise correlation matrix (with R2)
d <- 
  all_ranked %>%
  select(id,measure,scaled_value) %>% 
  spread(measure,scaled_value)

lower.panel<-function(x, y){
  points(x,y, pch=19, col = "grey")
  r <- round(cor(x, y), digits=2)
  txt <- paste0("R = ", r)
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  text(0.5, 0.9, txt)
}

pairs(d[,2:5], upper.panel = NULL, 
      lower.panel = lower.panel)
```

```{r UMAP clustering results}
clusterings_umap <- 
  all_ranked %>% 
  select(-value, -avg_value) %>% 
  spread(key = measure, value = scaled_value) %>% 
  umap_calc(n_neigh = 15, row_names = "id")

ann_transp <- unlist(map(ann_colors$scale, ~adjust_transparency(.x, alpha = 0.02)))
              
clusterings_umap %>% 
  group_by(features) %>% 
  mutate(scale = strsplit(features, " ")[[1]][1], 
        dist = strsplit(features, " ")[[1]][2],
         clust = strsplit(features, " ")[[1]][3],
         k = as.numeric(strsplit(features, " ")[[1]][4])) %>% 
  ggplot(aes(x=V1, y=V2, color = scale, fill = scale)) +
    geom_encircle() +
  geom_point( size=3) +
  scale_color_manual(values = ann_colors$scale) +
  scale_fill_manual(values = ann_transp) +
  theme_simple
              
```

```{r Linear models}
load("data/processed/all_eval.Rdata")

to_model <- 
  all_ranked %>% 
  group_by(id) %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
        distance = strsplit(id, " ")[[1]][2],
        clustering = strsplit(id, " ")[[1]][3],
        k = as.numeric(strsplit(id, " ")[[1]][4])) %>%
  select(-value)

to_model_all <-
  bind_rows(to_model %>% 
  select(-measure,-scaled_value) %>%
  distinct() %>%
  mutate(measure = "avg_value") %>%
  rename(scaled_value= avg_value),
  to_model %>%
  select(-avg_value))

measures <- c("connectivity_index", "avg_silhouette_width", "MI_zscore" ,"annotability", "avg_value")

models <- 
  to_model_all %>% 
  filter(measure %in% measures) %>% 
  group_by(measure) %>%
  do ({
    data <- . 

        model <- lm(scaled_value ~ scaling + distance + clustering + k, data = data) 
    anova_model <- aov(model) 
    
    anova_model %>%  
      tidy() %>% 
      left_join(anova_model %>% 
                  omega_sq() %>% 
                  enframe("term", "omega_sq") %>% 
                  mutate(term = trimws(term)),
                by = "term")
  })


### Does not work for k!
models_estimates <- 
  to_model_all %>% 
  filter(measure %in% measures) %>%
  group_by(measure) %>%
  do ({
    data <- .
    model <- lm(scaled_value ~ scaling + distance + clustering + k, data = data) 
    anova_model <- aov(model) 
    anova_model %>% TukeyHSD() %>% 
      tidy() %>% 
      arrange(adj.p.value)
  })

#write_xlsx(models, path = "anova.xlsx")
#write_xlsx(models_estimates, path = "tukey.xlsx")


    
    
# Palette
aovmodels$term <- factor(models$term, levels = c("Residuals","k","clustering","distance","scaling"))
models$measure <- factor(models$measure, levels = c("connectivity_index","avg_silhouette_width","MI_zscore","annotability","avg_rank"))

pal_models <- c("Residuals" = "#80818099",
                "k" = "#9DCC5F", #BB002199
                "clustering" = "#7E9DC2", #00828099
                "distance" = "#6E3562",
                "scaling"= "#FD8740") #008B4599

# Plot explained variance
models %>%
  ggplot(aes(x = measure, y= omega_sq, fill = term)) +
  geom_bar(stat="identity",width = 0.5) +
  theme_bw() +
  ylab("Omega Sq") +
  scale_fill_manual(values = pal_models) +
  theme(plot.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))+
  xlab("Evaluation measure") +
  ylab("% of expained variance")


# Plot results
dat_scale_dist <-
  to_model_all %>%
  filter(measure!="avg_value") %>%
  select(-clustering, -k) %>% 
  gather(key="factor", value = "value",-id,-scaled_value,-measure) 

dat_scale_dist$factor <-  factor(dat_scale_dist$factor,levels = c("scaling","distance"))
dat_scale_dist$value <-  factor(dat_scale_dist$value,levels = c("zscore","max","euclidean","pearson"))

P1 <- 
  dat_scale_dist %>% 
  ggplot(aes(x=factor, y = scaled_value, fill=value)) +
  geom_boxplot() +
  #scale_y_continuous(trans = "reverse")+
  #facet_wrap(~measure, scales="free_y") +
  scale_fill_manual(values = c(ann_colors$scale,ann_colors$dist))+
      facet_wrap(~measure, scales="free_y", ncol = 4) +

  theme_simple +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

P2 <-
  to_model_all %>%
  filter(measure!="avg_value") %>%
  ggplot(aes(x=clustering, y = scaled_value, fill=clustering)) +
  geom_boxplot() +
  scale_fill_manual(values = ann_colors$clust)+
    facet_wrap(~measure, scales="free_y", ncol = 4) +
  theme_simple +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

P3 <- 
  to_model_all %>%   
  filter(measure!="avg_value") %>%
  ggplot(aes(x=k, y=scaled_value, color = clustering, fill = clustering)) +
  #geom_smooth(aes(x=k, y=scaled_value), inherit.aes = F)+
  geom_smooth(alpha = 0.1) +
  geom_point() +
  scale_color_manual(values = ann_colors$clust) +
  scale_fill_manual(values =  ann_colors$clust) +
      facet_wrap(~measure, scales="free_y", ncol = 4) +
  theme_simple +
  xlab("Number of clusters (k)")



P1/ P2 / P3
```

```{r Evaluation based on intra cluster correlations}

# Calcualte correlations for one clustering strategy
#load("data/processed/all_partitions.Rdata")

clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value)

random_clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value) %>%
  mutate(value = sample(value))

# Calculate for clustering and for random_clustering
correlations <-
  tmm_data %>%
  mutate(log_tmm = log(tmm + 1))%>%
  select(-tmm) %>%
  left_join(random_clustering, by = c("enssscg_id" = "gene")) %>%
  group_by(value) %>%
  do({
    cluster <- unique(.$value)
    data <- .
    data <- data %>%
      select(-value) %>%
      spread(key = enssscg_id, value= log_tmm) %>%
      select(-sample_ID)

    cor_matrix <- cor(data)  %>%
      as_tibble(rownames = "gene1") %>%
      gather(gene2, cor, -gene1) %>%
      filter (gene1 < gene2) %>%
      mutate(cluster = cluster)
  })


# Plot results
p2 <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))750)

p_random <-
  correlations %>%
  ggplot(aes(x = as.factor(cluster), y = cor, fill = as.factor(cluster))) +
    geom_hline(aes(yintercept = mean(cor)), color="grey") +
  geom_violin(draw_quantiles = 0.5, show.legend = F) +
  theme_bw() +
  scale_fill_manual(values = p2) +
  ylab("Intra cluster correlation") +
  xlab("Cluster") + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

p / p_random


correlations %>% 
  group_by(cluster) %>%
  summarize(mean(cor)) %>%
  summarize(mean(`mean(cor)`)) 


# Calculate intra-cluster correlation for all partitions

intra_cluster_cor <- function (df) {
  df %>%
    group_by(value) %>%
    do({
      cluster <- unique(.$value)
      data <- .
      data <- data %>%
        select(-value) %>%
        spread(key = enssscg_id, value= log_tmm) %>%
        select(-sample_ID)
      
      cor_matrix <- cor(data)  %>%
        as_tibble(rownames = "gene1") %>%
        gather(gene2, cor, -gene1) %>%
        filter (gene1 < gene2) %>%
        mutate(cluster = cluster)
    })
}


all_correlations <-
  all_partitions %>%
  group_by(id) %>%
  do ({
    clustering <- . 
    id <- unique(clustering$id)
    correlations <-
      tmm_data %>%
      mutate(log_tmm = log(tmm + 1))%>%
      select(-tmm) %>%
      left_join(clustering %>% select(-id), by = c("enssscg_id" = "gene")) %>%
      intra_cluster_cor()
      
    average_correlation <- correlations %>%
      group_by(cluster) %>%
      summarize(mean(cor)) %>%
      summarize(mean(`mean(cor)`))
    data.frame(id = id, avg_cor = average_correlation)
  })

#save(all_correlations, file = "data/processed/all_correlations.Rdata")

cor_to_plot <- 
  all_correlations %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))

l <- unique(sort(cor_to_plot$k))
cor_to_plot$k <- factor(cor_to_plot$k, levels = l)

test <- 
  cor_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  })


test %>%
  ggplot(aes(x=id, y = mean..mean.cor..., fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~(clustering), scale="free") +
  labs(y = "Value", x = "Clustering strategy") +


all_correlations %>% 
  filter(id == "max euclidean fasthclust 50")

all_correlations %>% 
  filter(id %in% best_ids_clust) %>%
  arrange(-mean..mean.cor...)
```



#6- Visualization

```{r Generate all UMAPs}

# Generate UMAPs and save

umap_zscore_euclidean <- 
  pcas[[1]]$scores %>% 
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  rownames_to_column("enssscg_id") %>%
  select((1:30)) %>%
  umap_calc(row_names= "enssscg_id", n_neigh = 15, met = "euclidean")

umap_zscore_correlation <- 
  pcas[[1]]$scores %>% 
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  rownames_to_column("enssscg_id") %>%
  select((1:30)) %>%
  umap_calc(row_names= "enssscg_id", n_neigh = 15, met = "correlation")

umap_max_euclidean <- 
  pcas[[2]]$scores %>% 
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  rownames_to_column("enssscg_id") %>%
  select((1:30)) %>%
  umap_calc(row_names= "enssscg_id", n_neigh = 15, met = "euclidean")

umap_max_correlation <- 
  pcas[[2]]$scores %>% 
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  rownames_to_column("enssscg_id") %>%
  select((1:30)) %>%
  umap_calc(row_names= "enssscg_id", n_neigh = 15, met = "correlation")


umaps <- list(umap_zscore_euclidean, umap_zscore_correlation, 
     umap_max_euclidean, umap_max_correlation)

names(umaps) <- c("zscore euclidean", "zscore correlation", 
                "max euclidean", "max correlation")

#save(umaps, file= "data/processed/umaps_euc_corr.Rdata")


```

```{r Clustering UMAPs: visual evaluation}
#load("data/processed/umaps_euc_corr.Rdata")

# One clustering 
clustering_umaps(name = "zscore pearson fastkmeans 50",
                 umaps = umaps,
                 partitions = all_partitions)

#For many clusterings (in best_ids)
b <- best_clust %>% pull(id) %>% unique()
plots <- map(b,
             ~ clustering_umaps(name = .x, umaps = umaps, partitions = all_partitions) +
               ggtitle(.x))


pdf("umaps_best_clust.pdf", useDingbats = F)
plots
dev.off()


# Coordinate comparison

#load("data/processed/umaps_euc_corr.Rdata")
plot_data <-
 umaps$`zscore correlation` %>%
  mutate(r = scales::rescale(V1, c(0,1)),
         g = scales::rescale(V2, c(0,1)),
         b = 0.5,
         color = rgb(r, g, b)) %>%
  left_join(umaps$`max correlation`,
            by = "features",
            suffix = c("", "_int"))


plot_data %>% {
    ggplot(., aes(V1, V2, color = color)) +
    geom_point(size = 0.1) +
    theme_bw() +
    coord_fixed() +
    scale_color_identity() +
   
    ggplot(., aes(V1_int, V2_int, color = color)) +
    geom_point(size = 0.1) +
    theme_bw() +
    coord_fixed() +
    scale_color_identity()
}

```

```{r Visualize clusters in UMAP - optimal clustering}
#load("data/processed/umaps_euc_corr.Rdata")
#load("data/processed/all_partitions.Rdata")

umap_zscores_correlation <- umaps[["zscore correlation"]]

final_clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value)

# Clusterin UMAP
umap_plot(umap_zscores_correlation, color_by = "cluster", color_groups = final_clustering) + theme_void()


# Plot number of genes per cluster 
clustering %>% 
  group_by(value) %>% 
  count() %>% 
  ggplot(aes(x=as.factor(value),y=n,fill=as.factor(value))) +
  geom_bar(stat="identity", show.legend = F)+
  theme_simple_2 +
  scale_fill_manual(values = p) +
  xlab("Cluster number")+
  ylab("Number of genes")


# UMAP for each cluster (1 highlighted at a time)
to_plot <- 
  umap_zscores %>%
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  left_join(clustering, by = c("features" = "gene")) 

t <-
  to_plot %>%
  select(-value,-enhanced_tissues)

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(70)
plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5, color = p[cluster]) +
             theme_simple +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) + 
             scale_color_manual(values = tissue_colors) 
         })

pdf("zscore_pearson_medoids_70_all.pdf")
plots
dev.off()


# UMAP for each cluster (colored by tissue)
tissue_colors[["NA"]] <- "grey30"

plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2,color=enhanced_tissues)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5) +
             theme_minimal() +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) + 
             scale_color_manual(values = tissue_colors) 
         })
```

```{r Network}
# Cluster size
genes_per_cluster <- clustering %>% 
  group_by(value) %>% 
  count()

# Clustering network
to_network <- 
  fisher_res %>% 
  mutate(cluster = as.numeric(cluster)) %>% 
  filter(adj_p < 0.01) %>% 
  select(cluster,enhanced_tissues,odds_ratio) %>% 
  left_join(genes_per_cluster, by= c("cluster" = "value"))

network <- to_network %>%
  left_join(tissue_colors %>%  enframe(), by = c("enhanced_tissues" = "name")) %>% 
  rename(tissue_color = value,
         n_genes = n,
         tissue = enhanced_tissues)


# Add annotation
cluster_annotation <- read_excel(path = "20210426_annotation.xlsx", sheet = "cluster_info")

network_annotated <- network %>% 
  left_join(annotation %>% select(cluster,Max)) %>% 
  rename(annotation = Max) %>% 
  mutate (n_genes = sqrt(n_genes)) %>% 
  filter(odds_ratio > 2)

node_annotation <-
  network_annotated %>% 
  select(cluster,tissue,annotation) %>% 
  gather(type, name, cluster, tissue) %>% 
  mutate(label = ifelse(type == "cluster",
                        annotation, 
                        name)) %>% 
  select(shared_name=name,name=label)

#write_csv(network_annotated, file = "network_annotated_sqrtgenes.csv")
#write_csv(node_annotation, file = "node_annotation.csv")

```


#7- Annotation

```{r Tissue expression per cluster (pheatmaps)}

data_tissue <-
  tmm_data %>%
  mutate(tissue = gsub('.{2}$', '', sample_ID)) %>%
  group_by(tissue,enssscg_id) %>%
  mutate(tissue_exp = mean(tmm)) %>%
  select(enssscg_id, tissue, tissue_exp) %>%
  distinct()

clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value) 

annotation <-
  clustering %>%
  as_tibble() %>% 
  column_to_rownames("gene")

ann_colors <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(70)
names(ann_colors) = c(1:70)
ann_colors = list(value = ann_colors)
annotation$value <- as.factor(annotation$value)

dat <- data_tissue %>%
  group_by(enssscg_id) %>%
  mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
  ungroup() %>%
  select(enssscg_id,tissue,scaled_value) %>%
  spread(key=tissue, value = scaled_value) %>%
  left_join(clustering, by = c("enssscg_id" = "gene")) %>% 
  arrange(value) %>% 
  mutate(value = as.factor(value))

annot <- data.frame(value = dat$value)          
rownames(annot) <- dat$enssscg_id
gap_cols <- match(unique(dat$value), dat$value)
cols <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(70)

dat %>% 
  group_by(value) %>% 
  select(-enssscg_id) %>% 
  summarise(across(adiB:vp, list(mean))) %>% 
  as_tibble() %>% 
  column_to_rownames("value") %>% 
  pheatmap(clustering_method = "ward.D2", 
           color = viridis(20, direction = -1, option = "B"),
           #show_rownames = F,
           border_color = NA, 
           fontsize = 8) 

mutate(c(adiB:vp) = mean(c(adiB:vp)))

dat %>% 
  select(-value) %>% 
  column_to_rownames("enssscg_id") %>% 
  pheatmap(clustering_method = "ward.D2", 
           color = viridis(20, direction = -1, option = "B"),
           show_rownames = F,
           border_color = NA, 
           fontsize = 8, 
           cluster_rows = F,
           annotation_row = annot,
           annotation_colors = ann_colors)


pdf("heatmaps_zscore_person_medoids_70.pdf")
heatmaps <- lapply(sort(unique(clustering$value)),
                   function(i) {
                     plot <- data_tissue %>%
                       left_join(clustering, by = c("enssscg_id" = "gene")) %>%
                       # mutate(value = value+1) %>%
                       filter(value == i) %>%
                       group_by(enssscg_id) %>%
                       mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
                       ungroup() %>%
                       select(enssscg_id,tissue,scaled_value) %>%
                       spread(key=tissue, value = scaled_value) %>%
                       column_to_rownames("enssscg_id") %>% 
                       pheatmap(clustering_method = "ward.D2", 
                                color = viridis(20, direction = -1, option = "B"),
                                show_rownames = F,border_color = NA, fontsize = 8)
                   })
dev.off()

save(heatmaps[[1]], file="h.png")

pdf("heatmaps_zscore_person_medoids_70.pdf")
heatmaps

save(heatmaps,file = "data/processed/pheatmaps.Rdata")
## scale before

scaled_data <-
  data_tissue %>%
  group_by(enssscg_id) %>%
  mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
  ungroup()



left_join(clustering, by = c("enssscg_id" = "gene"))
pdf("heatmaps_zscore_person_medoids_70.pdf")
lapply(sort(unique(clustering$value)),
       function(i) {
         plot <-  %>%
           # mutate(value = value+1) %>%
           filter(value == i)  %>%
           select(enssscg_id,tissue,scaled_value) %>%
           spread(key=tissue, value = scaled_value) %>%
           column_to_rownames("enssscg_id") %>% 
           pheatmap(clustering_method = "ward.D2", 
                    color = viridis(20, direction = -1, option = "B"),
                    show_rownames = F,border_color = NA, fontsize = 8)
       })
dev.off()


data_tissue %>%
  left_join(clustering, by = c("enssscg_id" = "gene")) %>%
  # mutate(value = value+1) %>%
  filter(value == 67) %>%
  group_by(enssscg_id) %>%
  mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
  ungroup() %>%
  select(enssscg_id,tissue,scaled_value) %>%
  spread(key=tissue, value = scaled_value) %>%
  column_to_rownames("enssscg_id") %>% 
  pheatmap(clustering_method = "ward.D2", 
           color = viridis(20, direction = -1, option = "B"),
           show_rownames = F,border_color = NA, fontsize = 8)


```

```{r Cluster annotation DBSCAN}
cluster_centers<- 
  umap_zscores %>%
  left_join(clustering, by = c("features" = "gene")) %>% 
  group_by(value) %>%
  mutate(V1_mean = mean(V1),
         V2_mean = mean(V2)) %>% 
  select(value,V1_mean,V2_mean) %>% 
  left_join(annotation %>% select(cluster,Max), by = c("value" = "cluster")) %>% 
  distinct()


p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(70)

# Cluster centers by dbscan
plot_data <- 
  umap_zscores %>% 
  rename(gene_name = features) %>%
  left_join(clustering,
            by = c("gene_name" = "gene")) %>% 
  mutate(cluster = factor(value)) %>% 
  select(-value)

plot_data_sub <-
  plot_data %>%
  group_by(cluster) %>%
  mutate(sub_cluster = data.frame(V1, V2) %>%
           fpc::dbscan(eps = 0.2) %$%
           cluster) %>%
  ungroup()

subcluster_count <- 
  plot_data_sub %>% 
  group_by(cluster, sub_cluster) %>% 
  count() %>% 
  group_by(cluster) %>% 
  mutate(frac = n / sum(n))

keep_subcluster_fraction <- 
  subcluster_count %>% 
  filter(frac > 0.1) %>%
  filter(sub_cluster != 0) 

keep_subclusters <- 
  subcluster_count %>%
  filter(!sub_cluster == 0) %>% 
  top_n(1, n) %>% 
  slice(1) %>% 
  ungroup()

plot_data_sub %>% 
  ggplot(aes(V1, V2, color = factor(cluster))) +
  geom_point(show.legend = F, size =0.5) +
  scale_color_manual(values = p) +
  coord_fixed() +
  theme_bw() +
  ggtitle("All points") +
  
  plot_data_sub %>% 
  filter(sub_cluster != 0) %>%
  ggplot(aes(V1, V2, color = factor(cluster))) +
  geom_point(show.legend = F, size =0.5) +
  scale_color_manual(values = p) +
  coord_fixed() +
  theme_bw() +
  ggtitle("No outliers") +
  
  plot_data_sub %>% 
  inner_join(keep_subcluster_fraction) %>%
  ggplot(aes(V1, V2, color = factor(cluster))) +
  geom_point(show.legend = F, size =0.5) +
  scale_color_manual(values = p) +
  coord_fixed() +
  theme_bw() +
  ggtitle("Only subcluster of 10% total") +
  
  plot_data_sub %>% 
  inner_join(keep_subclusters) %>%
  ggplot(aes(V1, V2, color = factor(cluster))) +
  geom_point(show.legend = F, size =0.5) +
  scale_color_manual(values = p) +
  coord_fixed() +
  theme_bw() +
  ggtitle("Only main cluster")


#ggsave("results/UMAP/subcluster removal.pdf", width = 8, height = 8)  

plot_data_sub %>% 
  ggplot(aes(V1, V2, color = factor(cluster))) +
  geom_point(show.legend = F, size =0.5) +
  scale_color_manual(values = p) +
  coord_fixed() +
  theme_simple 

plot_data_sub %>% 
  inner_join(keep_subclusters) %>%
  ggplot(aes(V1, V2, color = factor(cluster))) +
  geom_point(show.legend = F, size =0.5) +
  scale_color_manual(values = p) +
  coord_fixed() +
  theme_simple

new_centers <-
  plot_data_sub %>% 
  inner_join(keep_subclusters) %>% 
  group_by(cluster) %>%
  mutate(V1_mean = mean(V1),
         V2_mean = mean(V2),
         cluster = as.numeric(cluster)) %>% 
  select(cluster,V1_mean,V2_mean) %>% 
  left_join(cluster_annotation %>% select(cluster, Max)) %>% 
  distinct() %>% 
  mutate(Max = paste(cluster,Max))

plot_data_hulls <-
  plot_data_sub %>% 
  inner_join(keep_subclusters) %>%
  filter(sub_cluster != 0) %>% 
  select(cluster, V1, V2) %>%
  group_by(cluster) %>%
  do({
    st_as_sf(., coords=c('V1','V2')) %>%
      concaveman(concavity = 2.5, length_threshold = .1) %$%
      st_coordinates(polygons) %>%
      as_tibble()
  })

plot_data_sub %>% 
  inner_join(keep_subclusters)  %>%  
  filter(sub_cluster != 0) %>% 
  ggplot(aes(V1, V2, color = factor(cluster))) +
  geom_polygon(data = plot_data_hulls,
               aes(X, Y,
                   color = cluster,
                   fill = cluster),
               alpha = 0.2,
               size = 0.4,
               inherit.aes = F,
               show.legend = F) +
  geom_point(show.legend = F, size =0.5, alpha = 0.2) +
  scale_color_manual(values = p) +
  coord_fixed() +
  theme_bw() +
  geom_point(data = new_centers, aes(x=V1_mean, y=V2_mean, fill=as.factor(cluster)),
             size = 4,
             colour = "gray50",
             pch=21,
             inherit.aes = F,
             show.legend = F) +
  scale_fill_manual(values = p) +
  #geom_text(data = new_centers,aes(V1_mean,V2_mean, label=Max, color = as.factor(cluster)), 
  #         size = 3,
  #         show.legend = F) +
  theme_void() 


# Cluster centers by dbscan
plot_data <- 
  umap_zscores %>% 
  rename(gene_name = features) %>%
  left_join(clustering,
            by = c("gene_name" = "gene")) %>% 
  mutate(cluster = factor(value)) %>% 
  select(-value)

plot_data_sub <-
  plot_data %>%
  group_by(cluster) %>%
  mutate(sub_cluster = data.frame(V1, V2) %>%
           fpc::dbscan(eps = 0.2) %$%
           cluster) %>%
  ungroup()

subcluster_count <- 
  plot_data_sub %>% 
  group_by(cluster, sub_cluster) %>% 
  count() %>% 
  group_by(cluster) %>% 
  mutate(frac = n / sum(n))

keep_subcluster_fraction <- 
  subcluster_count %>% 
  filter(sub_cluster != 0) %>% 
  filter(frac > 0.1) 


# Select region and highlight clusters
selected_clusters <- cluster_annotation %>%  filter(Region =="Brain") %>% pull(cluster)

keep_subclusters <- 
  subcluster_count %>%
  filter(sub_cluster != 0) %>% 
  top_n(1, n) %>% 
  slice(1) %>% 
  ungroup() 

selected_subclusters <- 
  keep_subclusters %>% 
  filter(cluster %in% selected_clusters)

plot_data_hulls <-
  plot_data_sub %>% 
  inner_join(selected_subclusters) %>%
  filter(sub_cluster != 0) %>% 
  select(cluster, V1, V2) %>%
  group_by(cluster) %>%
  do({
    st_as_sf(., coords=c('V1','V2')) %>%
      concaveman(concavity = 2.5, length_threshold = .1) %$%
      st_coordinates(polygons) %>%
      as_tibble()
  })



plot_data_sub %>% 
  inner_join(keep_subclusters)  %>%  
  filter(sub_cluster != 0) %>% 
  ggplot(aes(V1, V2)) +
  geom_point(show.legend = F, size =0.5, alpha = 0.2, color = "gray70") +
  
  geom_polygon(data = plot_data_hulls,
               aes(X, Y, group = cluster),
               color = "#FFDD00",
               fill = "#FFDD00",
               alpha = 0.2,
               size = 0.4,
               inherit.aes = F,
               show.legend = F) +
  #scale_color_manual(values = p) +
  coord_fixed() +
  theme_void() 

# Color palette
#95D4F5 -> male tissues
# red -> E34234
#green -> 016367
#yellow -> FFDD00

#write_tsv(cluster_annotation %>%  filter(Region =="Brain") %>% select(cluster,Max), "Brain.tsv")

```

```{r Agreement with previous classification}

clus <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value)

clas <- 
  gene_info %>%
  select(gene = ensg_id, specificity_category)

# UMAP by classification
umaps[[1]] %>%
  rename(gene = features)%>%
  left_join(clus) %>%
  left_join(clas) %>%
  ggplot(aes(V1,V2, color = specificity_category )) +
  geom_point() +
  theme_minimal() +
  scale_color_manual(values = gene_category_pal) +
  coord_fixed()


# UMAP by tissue
umaps[[1]] %>%
  rename(gene = features)%>%
  left_join(clus) %>%
  left_join(tissue_info, by = c("gene" = "ensg_id")) %>%
  ggplot(aes(V1,V2, color = enhanced_tissues)) +
  geom_point(show.legend = F) +
  theme_minimal() +
  scale_color_manual(values = tissue_colors) +
  coord_fixed()



# Tissue contributaion UMAPs

#saveRDS(clust_louvain, "results/Human Clustering.RDS")
# load("data/processed/data_scaled.Rdata")
# load("data/processed/pcas.Rdata")

clus_umap <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2)

clus_umap %>% umap_plot(color_by = "density")

#ggsave("results/human UMAP density.pdf", width = 5, height = 5)

clus_umap %>%  
  left_join(clust_louvain$cluster, 
            by = c("features" = "gene")) %>%
  ggplot(aes(V1,V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  # scale_color_manual(values = pal) +
  coord_fixed()

ggsave("results/human UMAP clusters.pdf", width = 5, height = 5)


color_data <- 
  tmm_data %>%
  left_join(colors,
            by = c("celltype" = "tissue")) %>%
  group_by(ensg_id, color = consensus_tissue_color) %>% 
  summarise(tpm = max(tpm)) %>% 
  
  group_by(ensg_id) %>%
  mutate(tpm = tpm / sum(tpm, na.rm = T)) %>%
  ungroup()

ensg_exp_palette <-
  color_data %>%
  bind_cols(col2rgb(.$color) %>% 
              t() %>% 
              as_tibble()) %>%
  mutate(red = red * tpm,
         green = green * tpm,
         blue = blue * tpm) %>%
  group_by(ensg_id) %>% 
  summarise(red = sum(red),
            green = sum(green),
            blue = sum(blue)) %>%
  mutate(color = rgb(red, green, blue, maxColorValue = 255))


clus_umap %>%
  left_join(ensg_exp_palette,
            by = c("features" = "ensg_id")) %>%
  ggplot(aes(V1,V2, color = color)) +
  geom_point(show.legend = F,
             size = 0.2) +
  theme_minimal() +
  scale_color_identity() +
  coord_fixed()

#ggsave("results/human UMAP tissue contribution.pdf", width = 5, height = 5)
```

```{r HT - Fischer function}
load("data/processed/all_partitions.Rdata")
### Fisher score
clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value) %>%
  left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
  select(gene,value) %>%
  na.omit() 

clust_list <- set_names(clustering$value,
                        clustering$gene)

tissue_list <- 
  tissue_info %$%
  split(ensg_id, enhanced_tissues)


FT_medoids_zscore_pearson_70 <- 
  gene_fischer_test(cluster_membership = clust_list, term_membership = tissue_list)


fisher_res <- FT_medoids_zscore_pearson_70 %>% rename(enhanced_tissues = term)


query_data <-
  
  fisher_res %>% 
  mutate(odds_ratio = ifelse(adj_p < 0.05,
                             odds_ratio, 
                             0)) %>%
  select(gene_name = cluster, tissue = enhanced_tissues, exp = odds_ratio)


query_clustering <-
  query_data %>%
  cluster_query(clustering = "ward.D2", distance_method = "binary")


plot_data <-
  query_data %>%
  mutate(gene_name = factor(gene_name, query_clustering$gene$labels[query_clustering$gene$order]),
         tissue = factor(tissue, query_clustering$sample$labels[query_clustering$sample$order]))

gene_segments <-
  dendro_data(query_clustering$gene)$segments %>%
  as_tibble() %>%
  rename(x = y,
         y = x,
         xend = yend,
         yend = xend) %>%
  mutate(segment_id = row_number()) %>%
  gather(x_type, x_coord, x, xend) %>%
  mutate(x_coord = - scales::rescale(x_coord,
                                     c(0, 5)) + 0.5) %>%
  spread(x_type, x_coord)

sample_segments <-
  dendro_data(query_clustering$sample)$segments %>%
  as_tibble() %>%
  mutate(segment_id = row_number()) %>%
  gather(y_type, y_coord, y, yend) %>%
  mutate(y_coord = scales::rescale(y_coord,
                                   c(0, 5)) +
           n_distinct(query_data$gene_name) + 0.5) %>%
  spread(y_type, y_coord)


plot_data %>%
  left_join(fisher_res, by = c("gene_name"="cluster", "tissue"="enhanced_tissues")) %>%
  filter(exp > 0) %>%
  bind_rows(data.frame(gene_name = as.factor(35))) %>% 
  ggplot(aes(tissue, gene_name,size = exp, fill=tissue)) +
  geom_point( shape = 21) +
  geom_segment(data = gene_segments,
               aes(x, y, xend = xend, yend = yend),
               inherit.aes = F) +
  geom_segment(data = sample_segments,
               aes(x, y, xend = xend, yend = yend),
               inherit.aes = F) +
  coord_fixed() +
  scale_y_discrete(expand = expansion(0),
                   position = "right") +
  theme(panel.background = element_blank(),
        panel.spacing = unit(0, "mm"),
        plot.margin = unit(c(0,0,0,0), "mm"),
        panel.grid =element_line(color="grey85"),
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        axis.text.x = element_text(angle = 90, vjust = 0.3, hjust = 1),
        legend.position = "right") +
  scale_fill_manual(values = tissue_colors)



```

```{r Annotation reports}
# Clustering information

load("data/processed/all_partitions.Rdata")

final_clustering <-
  all_partitions %>%
  filter(id=="zscore pearson medoids 70")

n_clusters <- final_clustering %>% pull(value) %>% unique() %>% length()

# Summarized tissue expression data
data_tissue <-
  tmm_data %>%
  mutate(tissue = gsub('.{2}$', '', sample_ID)) %>%
  group_by(tissue,enssscg_id) %>%
  mutate(tissue_exp = mean(tmm)) %>%
  select(enssscg_id, tissue, tissue_exp) %>%
  unique()

scaled_data <-
  data_tissue %>%
  group_by(enssscg_id) %>%
  mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
  ungroup()


# Run GO enrichments
GO_enrichments <- map(c("BP","CC","MF"),
                      ~enrich(clustering = final_clustering, ontology_type = .x, orthologs = ortholog_info))

#save(GO_enrichments, file = "data/processed/GO_enrichments.Rdata")
#load("data/processed/GO_enrichments.Rdata")


# Annotations
tissues <- scaled_data %>%  pull(tissue) %>%  unique()

tissue_ann <-
  colors %>%  select(tissue,consensus_tissue_name) %>% 
  unique() %>% 
  filter(tissue %in% tissues) %>% 
  column_to_rownames("tissue")

e <- tissue_ann %>% 
  rownames_to_column("tissue") %>% 
  left_join(enframe(tissue_colors), by = c("consensus_tissue_name" = "name")) %>% 
  select(-tissue) %>% 
  unique()

ann_column <- e %>%  pull(value)
names(ann_column) <- e %>% pull(consensus_tissue_name)

row_annotation<-
  gene_info %>% 
  select(ensg_id, specificity_category) %>% 
  filter(ensg_id %in% gene_names)

categories <- row_annotation %>% pull(specificity_category) %>% unique()
ann_row <- gene_category_pal[categories]


all_ann <- list(consensus_tissue_name = ann_column, specificity_category = ann_row)


# Indidicual annotation reports -> selected clusters: 22,43,61,65,67
r22 <- annotation_report(43)
pdf("22.pdf", width = 7, height = 9)
r22
dev.off()

# All annotation reports

reports_v5 <- map(c(1:n_clusters), ~annotation_report(.x))

pdf("reports_v5.pdf", width = 7, height = 9)
reports_v5
dev.off()



# Specificity classifiction legend
load("data/processed/umaps_euc_corr.Rdata")
gene_info %>% 
  select(ensg_id,specificity_category) %>% 
  left_join(umaps[[1]], by = c("ensg_id"="features")) %>% 
  ggplot(aes(x = V1, y = V2, color = specificity_category)) +
  geom_point() +
  scale_color_manual(values = gene_category_pal) +
  theme_simple

# Create file for annotation
cluster_info <-
  final_clustering %>% 
  group_by(value) %>% 
  count %>% 
  rename(cluster = value,
         nr_genes = n) %>% 
  left_join(fisher_res %>% 
              filter(adj_p<0.05) %>% 
              select(cluster,enhanced_tissues) %>% 
              ungroup() %>% 
              group_by(cluster) %>%
              summarize(enhanced_tissues = toString(enhanced_tissues)) %>%
              ungroup %>% 
              mutate(cluster = as.numeric(cluster)))

GO_info <- bind_rows(GO_terms_BP,GO_terms_CC,GO_terms_MF)


gene_description <- final_clustering %>% 
  select(gene, cluster=value) %>% 
  left_join(ortholog_info %>% select(gene=enssscg_id,human_id = ensg_id)) %>% 
  left_join(geninfo %>% 
              select(ensg_id,gene_name,gene_description), by = c("human_id" = "ensg_id")) %>% 
  left_join(gene_info %>% 
              select(ensg_id,specificity_category,distribution_category,enhanced_tissues), by= c("gene" = "ensg_id")) %>% 
  arrange(cluster) 




all <- list(cluster_info, GO_info, gene_description)
names(all) <- c("cluster_info","GO_info","gene_info")
write_xlsx(all, path = "annotation.xlsx")

write.xlsx(cluster_info, file = "annotation.xlsx",
           sheetName = "cluster_info", append = FALSE)
write.xlsx(GO_info, file = "annotation.xlsx", 
           sheetName="GO_info", append=TRUE)
write.xlsx(gene_description, file = "annotation.xlsx",
           sheetName="gene_info", append=TRUE)
```

```{r GOSemSim}
library(GOSemSim)

# Save GO data from orgs.Hs.eg.db
GO_data <- godata("org.Hs.eg.db", ont = "BP", computeIC=FALSE)

# Filtering too broad and too specific terms
GOs <- GO_data@geneAnno%>% group_by(GO) %>% count() %>% filter(n<500) %>% filter (n>3) %>% pull(GO) 

GO_data@geneAnno <-
  GO_data@geneAnno %>% 
  filter(GO %in% GOs)

clustering <- 
  all_partitions %>%
  filter(id == "zscore pearson medoids 70") %>%
  select(gene, value) %>%
  left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
  select(entrez,value) %>%
  na.omit() 

num_clusters <- length(unique(clustering$value)) 


cluster_list <- c()
num_clus <- length(unique(clustering$value))
num_genes <- length(clustering$value)
for (i in c(0:num_clus)) {
  genes <- clustering %>%
    filter(value==i) %>%
    pull(entrez) %>%
    as.character()
  cluster_list[[as.character(i)]] <- genes
}


# Calculate semantic similarity matrix
sim_matrix <- mclusterSim(clusters= cluster_list, semData = GO_data, measure = "Wang")

sim_matrix_max_pearson_medoids_50 %>% 
  pheatmap(border_color = NA, 
           clustering_method = "ward.D2", 
           color = colorRampPalette(c("white","#00846b"))(50))


# Filter uncharacterized clusters from the matrix (no or few GO terms present)
unchar <- cluster_annotation %>% 
  filter(Region == "Uncharacterized") %>% 
  pull(cluster)

char <- setdiff(c(1:70),unchar)

sim_matrix_zscore_pearson_medoids_70[char,char]%>% 
  pheatmap(color= viridis(20))
```

```{r GO Semantic similarity UMAP}
load("data/processed/GO_enrichments.Rdata")
all_go_terms <- 
  bind_rows(GO_enrichments[[1]]$enriched_terms,
            GO_enrichments[[2]]$enriched_terms,
            GO_enrichments[[3]]$enriched_terms)
GO <- 
  GO_enrichments[[1]]$enriched_terms %>% # Only BP
  filter(qvalue < 0.05) %>% 
  select(cluster = Cluster,
         GOID = ID)

GO_sparse <-
  GO %>%
  mutate(value = 1) %>% 
  pivot_wider(names_from = GOID, 
              values_from = value, values_fill = 0)

GO_pca <- 
  GO_sparse %>% 
  pca_calc(npcs = 200)

set.seed(42)

GO_umap <- 
  GO_pca$scores %>%
  as_tibble() %>%
  mutate(cluster = GO_sparse_2$cluster) %>% 
  select(c(1:50,"cluster")) %>% 
  umap_calc(n_neigh = 5, row_names = "cluster")

toplot <- cluster_annotation %>%
  mutate(cluster = as.character(cluster)) %>% 
  select(cluster, annotation = Max) %>% 
  left_join(GO_umap, by = c("cluster" = "features"))


toplot %>% 
  ggplot(aes(V1,V2,color=cluster)) +
  geom_point(show.legend = F, size =4) + 
  theme_minimal() +
  geom_text_repel(label = toplot$annotation, show.legend = F, size = 5) +
  scale_color_manual(values = cluster_palette)+
  theme_simple

```

```{r UMAP tissue contribution}

color_data <- 
  data_tissue %>% 
  left_join(colors) %>%
  group_by(enssscg_id, organ_name, color = organ_color) %>% 
  summarise(tissue_exp = max(tissue_exp)) %>% 
  group_by(enssscg_id) %>%
  mutate(tissue_exp = tissue_exp / max(tissue_exp, na.rm = T)) %>%
  ungroup() %>% 
  filter(tissue_exp > 0.1) %>% 
  group_by(enssscg_id) %>%
  mutate(tissue_exp = tissue_exp / sum(tissue_exp, na.rm = T)) %>%
  ungroup() 

ensg_exp_palette <-
  color_data %>%
  bind_cols(col2rgb(.$color) %>% 
              t() %>% 
              as_tibble()) %>%
  mutate(red = red * tissue_exp,
         green = green * tissue_exp,
         blue = blue * tissue_exp) %>%
  group_by(enssscg_id) %>% 
  summarise(red = sum(red),
            green = sum(green),
            blue = sum(blue)) %>%
  mutate(color = rgb(red, green, blue, maxColorValue = 255))

# Tissue contribution plot
umap_zscores_correlation %>%
  left_join(ensg_exp_palette,
            by = c("features" = "enssscg_id")) %>%
  ggplot(aes(V1,V2, color = color)) +
  geom_point(show.legend = F,
             size = 0.2) +
  theme_minimal() +
  scale_color_identity() +
  coord_fixed()

# organ_name legend
color_groups <- tissue_info %>% 
  left_join(colors %>%  select(consensus_tissue_name, organ_name,organ_color),
            by = c("enhanced_tissues" = "consensus_tissue_name"))

organ_colors_df <- color_groups %>%  select(organ_name, organ_color) %>% 
  distinct()

organ_colors <- organ_colors_df$organ_color  
names(organ_colors) <- organ_colors_df$organ_name

umap_zscores_correlation %>% 
  mutate(enssscg_id = gene_names) %>%
  left_join(color_groups, by = c("enssscg_id" = "ensg_id")) %>%
  ggplot(aes(V1,V2, color = organ_name)) +
  geom_point(show.legend = , size=3) +    
  theme_minimal() +
  scale_color_manual(values = organ_colors) +
  coord_fixed()

```

```{r UMAP summarized organ group expression heatmap}

dat <-
  color_data %>% 
  left_join(final_clustering, by = c("enssscg_id" = "gene")) %>% 
  group_by(value, organ_name) %>%
  summarise(tissue_exp = mean(tissue_exp)) %>%
  group_by(value) %>%
  mutate(tissue_exp = tissue_exp / sum(tissue_exp)) %>% 
  mutate(value = as.factor(value))

# Annotation
dat$value <- factor(dat$value, levels = levels(plot_data$gene_name))
annotation_cluster <- data.frame(value = factor(dat$value, levels = levels(plot_data$gene_name)))
annotation_cluster$value <- as.factor(annotation_cluster$value)
ann_colors <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(70)
names(ann_colors) = as.factor(c(1:70))
organs <- color_data %>%  select(organ_name, color) %>%  distinct() 
ann_organs <- organs$color
names(ann_organs) <- organs$organ_name
annotation_organ <- data.frame(organ_name = organs$organ_name)
rownames(annotation_organ) <- annotation_organ$organ_name
ann = list(value = ann_colors, organ_name = ann_organs)

# Expression heatmap
dat %>% 
  spread(organ_name, tissue_exp) %>% 
  column_to_rownames("value") %>% 
  t() %>% 
  
  pheatmap(color = viridis(20, direction = -1, option = "B"),
           clustering_method = "ward.D2",
           border_color = NA,
           cluster_cols = F,
           cluster_rows = T,
           annotation_col = annotation_cluster,
           annotation_row = annotation_organ,
           annotation_colors = ann)

plot_data %>% 
  left_join(dat, by = c("gene_name" = "value"))%>% 
  select(-tissue,-exp) %>% 
  group_by(gene_name) %>% 
  spread(organ_name, tissue_exp) %>% 
  column_to_rownames("gene_name") %>% 
  t() %>% 
  pheatmap(color = viridis(20, direction = -1, option = "B"),
           clustering_method = "ward.D2",
           border_color = NA)


color_data %>% 
  left_join(final_clustering, by = c("enssscg_id" = "gene")) %>% 
  group_by(value, organ_name) %>%
  summarise(tissue_exp = mean(tissue_exp)) %>%
  group_by(value) %>%
  mutate(tissue_exp = tissue_exp / sum(tissue_exp)) %>% 
  spread(organ_name, tissue_exp) %>% 
  column_to_rownames("value") %>% 
  t() %>% 
  
  pheatmap(color = viridis(20, direction = -1, option = "B"),
           clustering_method = "ward.D2",
           border_color = NA,
           cluster_cols = F,
           annotation_col = annotation_cluster,
           annotation_colors = ann)

```

```{r UMAP with annotation}

# Cluster centers based on average coordinates
cluster_centers<- 
  umap_zscores_correlation %>%
  left_join(final_clustering, by = c("features" = "gene")) %>% 
  group_by(value) %>%
  mutate(V1_mean = mean(V1),
         V2_mean = mean(V2)) %>% 
  select(value,V1_mean,V2_mean) %>% 
  left_join(cluster_annotation %>% select(cluster,Max), by = c("value" = "cluster")) %>% 
  distinct()


# UMAP colored by tissue
umaps[[4]] %>%
  left_join(ensg_exp_palette,
            by = c("features" = "enssscg_id")) %>%
  ggplot(aes(V1,V2, color = color)) +
  geom_point(show.legend = T,
             size = 1, alpha=0.5,  stroke = 0, shape=16) +
  theme_simple_2 +
  scale_color_identity() +
  coord_fixed()


# UMAP colored by cluster
cluster_centers %>% 
  ggplot(aes(V1_mean,V2_mean, label=Max, color = as.factor(value))) + # color=as.factor(value),
  geom_point(size = 4, show.legend = F)+ #, color = "grey29"
  geom_point(data = cluster_centers, 
             inherit.aes = F, aes(V1_mean,V2_mean, label=Max, color = as.factor(value))) + #, color = "gray90"
  scale_color_manual(values = pal) +
  scale_fill_manual(values = pal) +
  geom_text() +
  simple_theme 


```

