---
title: "HPA-gene-clustering"
author: "Maria Bueno Alvez"
date: "2/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r load packages and functions, include=FALSE}

# Packages
library(tidyverse)
library(magrittr)
library(readxl)
library(NOISeq)
library(uwot)
library(MASS)
library(viridis)
library(pcaMethods)
library(factoextra)
library(cluster)
library(kohonen)
library(ClusterR)
library(fclust)
library(flashClust)
library(Seurat)
library(dbscan)
library(igraph)
library(kmed)
library(aricode)
library(pheatmap)
library(clValid)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(annotate)
library(GO.db)
library(ggthemes)
library(ggrepel)
library(readr)
library(clusterProfiler)
library(mixOmics)

map <- purrr::map
pca <- pcaMethods::pca
#concaveman, ggalt, sf

# Wd, scripts
setwd("/Users/maria.bueno/Desktop/HPA-gene-clustering")
source("scripts/functions_preprocessing.R")
source("scripts/functions_dimensionalityreduction.R")
source("scripts/functions_clustering.R")
source("scripts/functions_evaluation.R")

map <- purrr::map
select <- dplyr::select
pca <- pcaMethods::pca
rank <- base::rank

```

```{r load data, include=FALSE}
# Data
data <- 
  read_delim("data/pig_filtered_sample_data.tab", delim = "\t")

gene_info <-
  read_delim("data/meta/pig_consensus_category_piggenome_92.tsv", delim = "\t")

genes_to_filter <-
  gene_info %>%
  filter(distribution_category == "not detected") %>%
  pull(ensg_id) 

tissue_info <-
  gene_info %>%
  filter(!ensg_id %in% genes_to_filter) %>%
  separate_rows("enhanced_tissues", sep = ',') %>%
  select(ensg_id,enhanced_tissues) %>%
  remove_missing()
  
# Color palette
colors <-
  read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors <- colors  %>%
  select(organ_color, consensus_tissue_name) %>%
  distinct() %$%
  set_names(organ_color, consensus_tissue_name)

```

#1- Data pre-processing

```{r pre-procesing}
filter <- dplyr::filter
# Filtering genes
filtered_data <-
  data %>% 
  filter(!enssscg_id %in% genes_to_filter)

# Save gene names in order
gene_names <- 
  filtered_data %>%
  select(enssscg_id, sample_ID, tpm) %>%
  spread(key = sample_ID , value = tpm) %$%
  enssscg_id

#save(gene_names,
 #    file = "data/processed/gene_names.Rdata")

# Normalize
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


# Scale data
scaled <- list("zscore","max") %>% #"min-max",
  map(~data_scaling(df = tmm_data, col_value = "tmm",
                    col_gene = "enssscg_id", col_sample = "sample_ID",
                    log_transform = F, m = .x))

```

#2- Dimensionality reduction

```{r dimensionality reduction}

#load("data/processed/scaled_data.Rdata")

pcas <- scaled %>%
  map(~pca_calc(data = .x$scaled, npcs = 200, m = .x$method))

# save(pcas,
#      file = "data/processed/pcas.Rdata")

```

#3- Distance calculation

```{r distance calculation}

#All distances
distances <- lapply(list("euclidean","pearson"), function(i) #"manhattan"
  map(pcas, ~dist_calc(df = .x$scores, comp = 30, m = i, 
                       id = .x$method)))

#save(distances,
 #    file = "data/processed/all_distances.Rdata")
```

#4- Clustering

```{r Clustering analysis}
#load("data/processed/distances_4.Rdata")
#load("data/processed/gene_names.Rdata")

#d <- list(distances[[1]][[1]],distances[[1]][[2]],distances[[2]][[1]],distances[[2]][[2]])

# Run with 1 k 
res_kmeans <- 
  distances %>%
  map(~clust(dist = .x$distance, 
             k = 100, 
             m = "fastkmeans", 
             genes = gene_names, 
             id = .x$id))

# Run multi k
hclust_multk <- 
  lapply(list(50,70,90,110,130,150), function(i)
    map(d, ~clust(dist = .x$distance, 
                  k = as.numeric(i), 
                  m = "fasthclust", 
                  genes = gene_names, 
                  id = .x$id, 
                  mult_k = T)))

# Run with 1 r 
res_louvain <- 
  distances %>%
  map(~clust(dist = .x$distance, 
             r = 4, 
             m = "louvain", 
             genes = gene_names, 
             id = .x$id))

# Run multi r
res_louvain_multk <- 
  lapply(list(4,6,8,10,12,14), function(i)
    map(d, ~clust(dist = .x$distance, 
                  r = as.numeric(i), 
                  m = "louvain", 
                  genes = gene_names, 
                  id = .x$id)))

#save(hclust_multk,
 #    file = "data/processed/hclust_multk.Rdata")

```


#5- Evaluation

```{r load data}
# load("data/processed/distances_4.Rdata")
# load("data/processed/gene_names.Rdata")
# load("data/processed/data_scaled.Rdata")

# Ortholog information
ortholog_info <- 
  read_delim("data/meta/gene_orthologs.csv", delim = ",") %>%
  select(enssscg_id, ensg_id) %>%
  dplyr::mutate(entrez = mapIds(org.Hs.eg.db,
                  keys= ensg_id, 
                  column="ENTREZID",
                  keytype="ENSEMBL"))

```

```{r Create df for evaluation}
load("data/processed/distances_4.Rdata")
load("data/processed/leiden_multk.Rdata")


eval_hclust_multk <- lapply(list(1,2,3,4,5,6), function(i) 
  map(hclust_multk[[i]], 
      ~eval(clustering = .x, 
            genes= ortholog_info,
            dis = d
      )))

#save(kmeans_multk, file = "data/processed/kmeans_multk.Rdata")
# load("data/processed/eval_som_multk.Rdata")
# load("data/processed/som_multk.Rdata")

eval_som_df <- eval_to_df(eval_somt_multk)


all_som <- add_clusterings(df = eval_som_df$res, 
                        dist = d, 
                        pca = pcas[[1]],
                        ref_clust = som_multk[[1]][[1]]$cluster,
                        gene_names = gene_names, 
                        clust_m = "SOM",
                        orthologs = ortholog_info)

bestk_leiden <- find_bestk(all_leiden)

save(eval_leiden_df,all_leiden,bestk_leiden, file = "data/processed/evaluation_leiden.Rdata")

```

```{r Plot ranking}
to_eval <- bind_rows(bestk_hclust$bestk, bestk_kmeans$bestk,
                     bestk_medoids$bestk, bestk_louvain$bestk,
                     bestk_leiden$bestk, bestk_som$bestk)

to_eval<-
  to_eval %>%
  ungroup() %>%
  mutate(rank_time2 = (base::rank(time)),
         rank_BHI2 = (base::rank(-BHI_index)),
        rank_connectivity2 = (base::rank(connectivity_index)),
          rank_asw2 = (base::rank(-avg_silhouette_width)))

to_eval %>%
  filter(!(grepl("2D",clustering_id))) %>%
  column_to_rownames("clustering_id") %>%
  select(rank_asw2,rank_connectivity2,rank_BHI2,rank_time2) %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))

e <- 
  to_eval %>%
  filter(!(grepl("2D",clustering_id))) %>%
  #select(clustering_id, scale_dist, avg_silhouette_width,connectivity_index,dunn_index, BHI_index, BP_enriched, time) %>%
  find_bestk()

```

```{r Plot metrics - multk}


# Barplot
df_louvain %>%
  as_tibble() %>%
  filter(scale_dist == 'zscore euclidean')%>%
  ggplot(aes(x=k, y=connectivity_index)) +
  geom_bar(stat = "identity")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


# Prepare data
to_plot <- df_hclust %>%
  as_tibble() %>%
  #mutate(id = gsub('[[:digit:]]+', '', clustering_id)) %>%
#  filter(id == "zscore euclidean fasthclust ") %>%
  select(scale_dist,k,connectivity_index,dunn_index, avg_silhouette_width, 
         BHI_index, time) %>% #BP_enriched
  gather(key = measure, value = value, connectivity_index:time) 

hclust_plot <- 
  to_plot %>%
  filter(measure != "time") %>%
  ggplot(aes(x=as.numeric(k), y=value, group =1, color = as.factor(measure))) +
  geom_point(show.legend = F) +
  geom_line(show.legend = F) +
  theme_minimal() +
  ggtitle("Hierarchical clustering") +
  facet_wrap(scale_dist~measure, scales="free_y"
             ) +
  scale_color_viridis(discrete = T) +
  theme(text = element_text(size=10)) + 
  labs(y = "Value", x = "Number of clusters")

ggsave("ev_hclust.pdf", plot = hclust_plot)




# Rank
df_louvain <-
  df_louvain %>%
  mutate(rank_connectivity = rank(connectivity_index),
         rank_dunn = rank(-(dunn_index)),
         rank_asw = rank(-(avg_silhouette_width)),
         rank_BHI = rank(-(BHI_index)),
         rank_BPE = rank(-(BP_enriched)),
         rank_time = rank(time)
  )

# Select best k 
louvain_bestk <- 
  df_louvain %>%
  mutate(total_rank = rank_connectivity + rank_dunn + rank_asw + rank_BPE + rank_BHI) %>%
  #mutate(id = gsub('[[:digit:]]+', '', clustering_id)) %>%
  group_by(scale_dist) %>%
  mutate(min = (min(total_rank))) %>%
  filter(total_rank == min)

louvain_bestk_ranodm_2D <-
  bind_rows(
  louvain_bestk %>% select(clustering_id,connectivity_index, dunn_index,avg_silhouette_width, BHI_index, BP_enriched, time),
  ev_louvain_extra %>% select(clustering_id,connectivity_index, dunn_index,avg_silhouette_width, BHI_index, BP_enriched, time))


louvain_bestk_ranodm_2D %>%
  as_tibble() %>%
  gather(key = measure, value = value, connectivity_index:time) %>%
  select(-scale_dist) %>%
  #filter(scale_dist == 'zscore euclidean')%>%
  ggplot(aes(x=clustering_id, y=value, fill=clustering_id)) +
  geom_bar(stat = "identity")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~measure, scales="free_y") +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=10)) + 
  theme_minimal()
  labs(y = "Value", x = "Number of clusters")

ggsave("ev_hclust.pdf", plot = hclust_plot)

```

```{r Prepare all clusterings}

load("data/processed/results_kmeansfast.Rdata")
load("data/processed/results_medoids.Rdata")
load("data/processed/results_hclust.Rdata")
load("data/processed/results_leiden.Rdata")
load("data/processed/results_louvain.Rdata")
load("data/processed/results_som.Rdata")


```

```{r Plot metrics - different clusterings}
all_best <- data.frame()

for (df in dfs) {
  b <- bestk %>%
    select(-rank_connectivity, -rank_dunn, -rank_BHI, -rank_time,-min,-total_rank, -k, -clustering_id)
  all_best <- bind_rows(all_best,b)

  
# Generate ranking - plot heatmaps

ev_res %>%
  select(clustering_id,rank_connectivity, rank_dunn, rank_BHI, rank_time) %>%
  column_to_rownames("clustering_id") %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))


df_eval_medoids$rank_connectivity <-
  rank(-(df_eval_medoids$connectivity_index))
df_eval_medoids$rank_dunn <-
  rank((df_eval_medoids$dunn_index))
df_eval_medoids$rank_BHI <-
  rank(-(df_eval_medoids$BHI_index))
df_eval_medoids$rank_time <-
  rank(df_eval_medoids$time)

df_eval_medoids %>%
  select(clustering_id,rank_connectivity, rank_dunn, rank_BHI, rank_time) %>%
  column_to_rownames("clustering_id") %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))

df_eval_medoids %>%
  as_tibble() %>%
  select(scale_dist,k,rank_connectivity) %>%
 # select(-clustering_id) %>%
  #select(clustering, scale_dist, rank_dunn) %>%
  spread(key = scale_dist, value = rank_connectivity) %>% 
  #d#rop_na(clustering) %>%
  #select(-`random NA`)%>%
  #select(k,`max euclidean`) %>%
  as.data.frame() %>%
  column_to_rownames("scale_dist") %>%
  pheatmap()
```

```{r test different scaling for evaluation pheatmap}
ev %>%
  column_to_rownames("id") %>%
  mutate(connectivity_index = scale(connectivity_index)) %>%
  pheatmap(cluster_cols = F)

ev_minmax <- 
  ev %>%
  left_join(df$cluster_time %>% bind_rows(data.frame(id = "random", time = 1))) %>%
  column_to_rownames("id") %>%
  mutate(min = min(connectivity_index),
         max = max(connectivity_index),
         connectivity = (connectivity_index - min)/(max-min)) %>%
  select(connectivity,dunn_index,BHI_index, time) %>%
  mutate(min = min(dunn_index),
         max = max(dunn_index),
         dunn = (dunn_index - min)/(max-min)) %>%
  select(connectivity,dunn,BHI_index, time) %>%
  mutate(min = min(BHI_index),
         max = max(BHI_index),
         BHI = (BHI_index - min)/(max-min)) %>%
  select(connectivity,dunn,BHI, time) %>%
   mutate(min = min(time),
         max = max(time),
         time = (time - min)/(max-min)) %>%
  select(connectivity,dunn,BHI, time) 
  #mutate(sum_connectivity = scale(connectivity_index)) %>%

ev_minmax %>% 
  pheatmap(cluster_cols = F)


### UMAP clusterings

ev_umap <- 
  ev_minmax %>%
  select(-time) %>%
  rownames_to_column("id") %>%
  umap_calc(row_names = "id", n_neigh = 15, n_comp = 2)


scaling <- c()
distance <- c()
clustering <- c()

for (row in 1:nrow(ev_umap)) {
  elements <- strsplit(ev_umap[[row, "features"]], " ")
  scaling <- c(scaling, elements[[1]][1])
  distance <- c(distance, elements[[1]][2])
  clustering <- c(clustering, elements[[1]][3])
}

ev_umap_colors <-
  ev_umap %>%
  mutate(scaling = scaling,
         distance = distance,
         clustering = clustering)


ev_umap %>% 
  ggplot(aes(V1,V2, color =scaling)) +
      geom_point() +
      theme_minimal()+
      coord_fixed()
#color = viridis(10)
```

```{r pairwise ARI}

# Prepare data 
df2 <- res %>% select(-time,-clustering_id) %>% unique()
clusterings <- c(df2$id,"random")
to_eval <- df$cluster_results %>% bind_rows(random %>% mutate(id = "random"))


################ Problem -> retrieve partitions

clusterings <- to_eval$clustering_id
#36 options
res_ari <- matrix(, nrow = 36, ncol = 36)
rownames(res_ari) <- c(clusterings)
colnames(res_ari) <- c(clusterings)

# Calculate ARI matrix
for (i in clusterings) {
  for (j in clusterings) {
    a <- ARI(to_eval %>% filter(id == i) %>% pull(cluster), 
             to_eval %>% filter(id == j) %>% pull(cluster))
    res_ari[i,j] <- a
    }
  }

# Pheatmap for all scalings, distances and clustering methods
res_ari %>%
  pheatmap(fontsize = 5)

# Pheatmap for specific scalings

d <- c()
for (id in clusterings) {
  if (grepl("max",id) & !grepl("min",id)){d <- c(d,id)}
}

d <- c()
for (id in clusterings) {
  if (grepl("pearson",id)){d <- c(d,id)}
}

res_ari %>%
  as_tibble() %>%
  mutate(id = clusterings)%>%
  filter(id %in% d) %>%
  select(d) %>%
  as.matrix() %>%
  set_rownames(d)%>%
  pheatmap(fontsize = 10)

```

```{r RankAggreg}

load("data/processed/evaluation_louvain.Rdata")

rank <- base::rank

connectivity <- 
  bestk_louvain$bestk %>%
  ungroup() %>%
  select(clustering_id, connectivity_index) %>%
  arrange(connectivity_index) %>%
  pull(clustering_id)

avg_silhouette_width <- 
  bestk_louvain$bestk %>%
  ungroup() %>%
  select(clustering_id, avg_silhouette_width) %>%
  arrange(desc(avg_silhouette_width)) %>%
  pull(clustering_id)

BHI_index <- 
  bestk_louvain$bestk %>%
  ungroup() %>%
  select(clustering_id, BHI_index) %>%
  arrange(desc(BHI_index)) %>%
  pull(clustering_id)

connectivity_weight <- 
  bestk_louvain$bestk %>%
  ungroup() %>%
  select(clustering_id, connectivity_index) %>%
  arrange(connectivity_index) %>%
  pull(connectivity_index)

avg_silhouette_width_weight <- 
  bestk_louvain$bestk %>%
  ungroup() %>%
  select(clustering_id, avg_silhouette_width) %>%
  arrange(desc(avg_silhouette_width)) %>%
  pull(avg_silhouette_width)

BHI_index_weight <- 
  bestk_louvain$bestk %>%
  ungroup() %>%
  select(clustering_id, BHI_index) %>%
  arrange(desc(BHI_index)) %>%
  pull(BHI_index)


ranks <- data.frame(connctivity = connectivity, 
           avg_silhouette_width = avg_silhouette_width, 
           BHI_index = BHI_index) %>%
  t()

weights <- data.frame(connctivity = connectivity_weight, 
           avg_silhouette_width = avg_silhouette_width_weight, 
           BHI_index = BHI_index_weight) %>%
  t()
        

library(RankAggreg)

rank <- RankAggreg(ranks, 6, weights = weights, seed=123)
rank_noweight <- RankAggreg(ranks, 6,  seed=123)




rank_manual <-
  bestk_louvain$bestk %>%
  ungroup()  %>%
  mutate(rank_connectivity = rank(connectivity_index),
           rank_asw = rank(-(avg_silhouette_width)),
           rank_BHI = rank(-(BHI_index)),
    ) %>%
  select(clustering_id, rank_connectivity, rank_asw, rank_BHI) %>%
  mutate(total_rank = rank_connectivity + rank_asw +  rank_BHI ) %>% 
  dplyr::arrange(total_rank) %>% 
  pull(clustering_id)


```


#6- Visualization

```{r UMAP visualization - tests with Max}
load("data/processed/data_scaled.Rdata")
load("data/processed/pcas.Rdata")
load("data/processed/pcas.Rdata")
load("data/processed/louvain_multk.Rdata")
load("data/processed/medoids_multk.Rdata")

clustering <- louvain_multk[[1]][[1]]$cluster


## UMAP

clus_umap <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:10)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2) 

clus_umap %>% umap_plot()
                        
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, fill = cluster, color = cluster)) +
  geom_point(size = 0.1) +
  stat_density2d(bins = 2, h = 2) +
  theme_bw()


geom_density_2d(bins = 2, )
  geom_density2d_filled(bins = 20)


## Hexagonal
  getmode <- function(v) {
  uniqv <- unique(v)
  as.character(uniqv[which.max(tabulate(match(v, uniqv)))])
}
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, z = cluster, group = 1)) +
  stat_summary_hex(fun = getmode,
                   bins = 100,
                   show.legend = F)
## Density
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, group=cluster)) +
  geom_density2d_filled(h = 10) +
  facet_wrap(~cluster)


#######

density_settings <-
  expand.grid(n = c(100,200,300,400),
              h = c(0.02, 0.05, 0.1, 0.2)) %>% 
  as_tibble()


clus_umap_density_all <- 
  density_settings %>% 
  group_by_all() %>% 
  do({
    h <- .$h
    n <- .$n
    
    clus_umap %>% 
      left_join(clustering %>% 
                  dplyr::rename(features = 1, 
                                cluster = 2)) %>% 
      mutate(cluster = factor(cluster)) %>%
      group_by(cluster) %>%
      do ({
        data <- .
        data %$%
          MASS::kde2d(V1, V2, n = n, h = h, lims = c(range(clus_umap$V1), range(clus_umap$V2))) %>%
          {.$z} %>%
          as_tibble(rownames = "x") %>%
          gather(y, density, -1) %>%
          mutate(y = gsub("V", "", y),
                 x = as.numeric(x),
                 y = as.numeric(y))
      })
  })
clus_umap_density <-
  clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  group_by(cluster) %>%
  do ({
    data <- .
    data %$%
      MASS::kde2d(V1, V2, n = 300, h = 0.1, lims = c(range(clus_umap$V1), range(clus_umap$V2))) %>%
      {.$z} %>%
      as_tibble(rownames = "x") %>%
      gather(y, density, -1) %>%
      mutate(y = gsub("V", "", y),
             x = as.numeric(x),
             y = as.numeric(y))
  })

clus_umap_density%>%
 # filter(n== 400, h == 0.05) %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.01) %>%
  ggplot(aes(x,y,fill = cluster, alpha = density)) +
 # geom_encircle() +
  geom_point(data = clus_umap %>% 
               left_join(clustering, by = c("features" = "gene")), #%>% filter(value ==2),
             aes(V1, V2),
             size = 0.07,
             inherit.aes = F) + 
  geom_tile() +
  theme_bw() +
  scale_alpha_continuous(limits = c(0, 0.3), range = c(0.05, 0.6))
  #facet_wrap(~cluster)#+
 # facet_grid(h ~ n)


clus_umap_density %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.01) %>%
  ggplot(aes(x,y,fill = cluster, alpha = density)) 
  
clus_umap %>%
  ggplot() +
  geom_point(aes(V1, V2),
             size = 0.1,
             inherit.aes = F) 

clus_umap_density %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.005) %>%
  ggplot(aes(x,y,fill = cluster)) +
  geom_tile()+
  theme_bw() +
  scale_fill_manual(values = pal)

clus_umap %>%
  left_join(clustering, by = c("features" = "gene")) %>%
  ggplot(aes(V1,V2,color=as.factor(value))) +
  geom_point(size=0.6) +
  theme_bw() +
  scale_color_manual(values = pal) #+
  oose#facet_wrap(~value)

library(pals)
pal <- colorRampPalette(ocean.curl(60))(60)


clus_umap_density %>%
  filter(density > 0.1) %>% 
  group_by(x, y) %>%
  count %>% 
  group_by(n) %>% 
  count


# clus_umap_density %>%
#   filter(density>0.01) 
#   ggplot(aes(density)) +
#   geom_density() 

```

```{Visual evaluation - test UMAP distance metrics (Louvain / Leiden)}

umap <- scaled[[1]]$scaled %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2)

umap$umap %>%
  left_join(res_kmeansfast[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() 

umap_corr <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "correlation")

umap_manhattan <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "manhattan")

umap_euclidean <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "euclidean")

umaps <- list(umap_euclidean,umap_manhattan,umap_corr)

save(umaps,
     file = "data/processed/umaps.Rdata")


load("data/processed/results_leiden.Rdata")
load("data/processed/results_louvain.Rdata")

# 2D clustering
d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)
c <- clust(dist = d$distance, k = 100, m = "fastkmeans", genes = gene_names, 
           id = d$id)

umap$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 


d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)

c <- clust(dist = d$distance, k = 100, m = "louvain", genes = gene_names, 
           id = d$id)

umaps[[3]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 



load("data/processed/umaps.Rdata")


df2 <- data.frame(value = as.numeric(medoids$cluster), gene = gene_names)
umaps[[1]]$umap %>%
  left_join(df2, by = c("features" = "gene")) %>%
  ggplot(aes(V1, V2)) +
   geom_hex(aes(fill = 1), fill = "gray90") +
  theme_minimal() +
  geom_encircle(aes(fill = value,
                     color = value),
                 alpha = 0.1,
                 s_shape=1,
                 expand=0,
                 # color = "white",
                 show.legend = F) +
   geom_encircle(aes(color = value),
                 fill = NA,
                 alpha = 0.7,
                 s_shape=1,
                 expand=0,
                 # color = "white",
                 show.legend = F)# +
#   geom_text(data = . %>%
#               select(merged_cluster, mean_V1, mean_V2, n) %>%
#               distinct(),
#             aes(mean_V1, mean_V2, label = merged_cluster),
#             show.legend = F)
  
  
load("data/processed/results_louvain.Rdata")

df2 <- data.frame(value = as.numeric(medoids$cluster), gene = gene_names)

umaps[[1]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  head(200) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal)  + 
    scale_fill_manual(values = pal)  + 

  #geom_density_2d()
#  stat_density_2d(aes(fill = value), geom="polygon")+ 
# stat_ellipse()
   geom_encircle(aes(group = value, fill = as.factor(value)),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                #fill = "gray",
               # color = "black",
                show.legend = F)  +
    geom_point(show.legend = F, size = 1) 
  #   ggplot(aes(V1, V2)) +
#   geom_hex(aes(fill = 1),
#            data = plot_data,
#            fill = "gray90",
#            bins = plot_bins) +
#   geom_encircle(aes(fill = merged_cluster,
#                     color = merged_cluster),
#                 alpha = 0.1,
#                 s_shape=1,
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   geom_encircle(aes(color = merged_cluster),
#                 fill = NA,
#                 alpha = 0.7,
#                 s_shape=1,
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   geom_text(data = . %>%
#               select(merged_cluster, mean_V1, mean_V2, n) %>%
#               distinct(),
#             aes(mean_V1, mean_V2, label = merged_cluster),
#             show.legend = F)
  
  

pal <- colorRampPalette(ocean.curl(50))(50)

  geom_encircle(aes(group = value),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                fill = "gray",
                color = "black",
                show.legend = F) 
#+ scale_color_manual(values = pal) 
```

```{r Cluster visualziationn - UMAP}


umap_louv %>% 
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  filter(enhanced_tissues == "retina") %>%
  ggplot(aes(V1,V2)) +
  geom_point(data = t, aes(V1,V2), color = "grey", inherit.aes = F) +
  geom_point(show.legend = F, color = tissue_colors[["retina"]]) +
  
  theme_minimal() +
  #scale_color_manual(values = tissue_colors) +
  coord_fixed()

# For all tissues

tissues <- unique(tissue_info$enhanced_tissues)
t

plots <- 
  lapply(unique(tissue_info$enhanced_tissues),
         function(tissue) {
           umap_louv %>%
             left_join(tissue_info, by= c("features" = "ensg_id")) %>%
             filter(enhanced_tissues == tissue) %>%
            ggplot(aes(V1,V2)) +
              geom_point(data = t, aes(V1,V2), color = "grey", inherit.aes = F) +
              geom_point(show.legend = F, color = tissue_colors[[tissue]]) +
              theme_minimal() +
              coord_fixed() +
             ggtitle(tissue)
         })

pdf("umaps_tissues.pdf")
plots
dev.off()


# Tissues for each cluster

to_plot <- 
  umap_louv %>%
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  left_join(clustering, by = c("features" = "gene")) %>%
  mutate (value = value+1)

t <-
  to_plot %>%
  select(-value,-enhanced_tissues)

tissue_colors[["NA"]] <- "grey30"

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(67)
plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2,color=enhanced_tissues)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5) +
             theme_minimal() +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) + 
             scale_color_manual(values = tissue_colors) 
         })

pdf("us.pdf")
plots
dev.off()


```

```{r repeat - final clustering}
umap_zscores <- umaps[[1]]

clustering <- 
  all_partitions %>%
  filter(id == "zscore euclidean fasthclust 50") %>%
  select(gene, value)

to_plot <- 
  umap_zscores %>%
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  left_join(clustering, by = c("features" = "gene")) #%>%
#  mutate (value = value+1)

t <-
  to_plot %>%
  select(-value,-enhanced_tissues)

# Tissues for each cluster

tissue_colors[["NA"]] <- "grey30"

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(50)
plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2,color=enhanced_tissues)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5) +
             theme_minimal() +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) + 
             scale_color_manual(values = tissue_colors) 
         })

pdf("us.pdf")
plots
dev.off()
```

#7- Annotation

```{r Run enrichments}
load("data/processed/som_multk.Rdata")


to_df <- function(multk, add_random = T) {
  lapply(c(1:4),
         function(i) 
           
           if(add_random == T) {
             random <- as.data.frame(multk[[1]][[i]]$cluster) %>%
               mutate(value = sample(value))
             df <- random %>%
               mutate(id = 
                        paste((gsub('[[:digit:]]+', '', 
                                    multk[[1]][[i]]$id)),"random", 
                              sep = ""))
             
             for (j in c(1:6)) {
               df <- bind_rows(df, as.data.frame(multk[[j]][[i]]$cluster) %>% 
                                 mutate(id = multk[[j]][[i]]$id))
             }
             return(df)
           }
         
         else {
           df <- data.frame()
           for (j in c(1:6)) {
             df <- bind_rows(df, as.data.frame(multk[[j]][[i]]$cluster) %>% 
                               mutate(id = multk[[j]][[i]]$id))
           }
           return(df)
         }
         
  )
}

enrich <- function (clustering,ontology_type="BP",orthologs, BP_index=F) {
  clustering <-
    clustering %>%
    left_join(orthologs, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(entrez,value)
  
  cluster_list <- c()
  num_clus <- tail(sort(clustering$value), n=1)
  num_genes <- length(clustering$value)
  for (i in c(0:num_clus)) {
    genes <- clustering %>%
      filter(value==i) %>%
      pull(entrez) %>%
      as.character()
    cluster_list[[as.character(i)]] <- genes
  }
  
  # Profile clusters (with universe)
  ck <- try(compareCluster(geneClusters = cluster_list, 
                           fun = "enrichGO", 
                           OrgDb = org.Hs.eg.db, 
                           ont = ontology_type, 
                           universe = as.character(clustering$entrez),
                           pAdjustMethod = "BH",
                           pvalueCutoff = 1, 
                           qvalueCutoff = 1))
  
  ck_tibble <- as_tibble(ck) 
  
  if(BP_index == T) {
    bio_index <-
      length(unique(as.numeric(ck_tibble$Cluster)))/(num_clus)
    
    return(list(enriched_terms = ck_tibble,
                enrichment_score = bio_index))
  }
  
  else {
    return(ck_tibble)
  }
  
}
enrichments_df <- function(df, grouping, orthologs) {
  
  res <- df %>%
    group_by(.data[[grouping]]) %>%
    do({
      subset <- .
      
      enrichment <-
        enrich(clustering = subset, orthologs = orthologs) %>%
        mutate(id = unique(subset$id))
    })
  
}



df_som <- 
  to_df(som_multk, add_random = T)

Sys.time() #5.37
enrichments_som <- df_som %>%
  map(~enrichments_df(df = .x, grouping = "id", orthologs = ortholog_info))

save(enrichments_som, file="data/processed/enrichments_som.Rdata")

load("data/processed/enrichments_louvain.Rdata")


enrichment_scores <- function(df,grouping) {
  
  df %>%
    group_by(.data[[grouping]]) %>%
    do ({
      data <- .
      min_significance <- data %>%
        mutate(log_qvalue = -log10(qvalue)) %>%
                                group_by(Cluster) %>% 
                                summarize(max(log_qvalue)) %>% 
                                pull(`max(log_qvalue)`) %>% mean() 
      annotability <- length(data %>%
               filter(qvalue < 0.05) %>% pull(Cluster) %>% sort() %>% unique())/
                           length(data %>% pull(Cluster) %>% sort() %>% unique())
      data.frame(id = unique(data$id), min_significance = min_significance, annotability = annotability)
        
    })
}

# to_plot_kmeans <- 
#   enrichment_scores(df = enrichments_kmeans[[1]], grouping = "id") %>%
#   gather(key = "measure", value = "value", -id) 
# 
# 
# to_plot_kmeans$id <- factor(to_plot_kmeans$id,
#                                             levels = c("zscore euclidean fastkmeans 50","zscore euclidean fastkmeans 70",
#                                                        "zscore euclidean fastkmeans 90","zscore euclidean fastkmeans 110",
#                                                        "zscore euclidean fastkmeans 130","zscore euclidean fastkmeans 150",
#                                                        "zscore euclidean fastkmeans random")
#                             )
# 
# to_plot_kmeans %>%
#   ggplot(aes(x= id, y = value, fill = id)) +
#   geom_bar(stat = "Identity") +
#   theme_minimal() +
#   facet_wrap(~measure, scales="free_y") +
#   scale_fill_viridis(discrete = T) +
#   theme(text = element_text(size=15)) + 
#   labs(y = "Value", x = "Clustering") +
#       theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
  
  
  
  plot_res <- function(x,grouping) {
  x %>%
    group_by(.data[[grouping]]) %>%
    do ({
      data <- .
      curr_id <- strsplit(data$id, " ")
      title <- paste(curr_id[[1]][1],curr_id[[1]][2],curr_id[[1]][3])

      data$new_id <- curr_id[[1]][4]
            data$group <- title

      data
    })

}

# All 
to_plot_louvain <-
  map(enrichments_louvain,
  ~enrichment_scores(.x, grouping = "id"))
  
list_to_plot <- 
  map(to_plot_louvain,
  ~.x %>% 
  plot_res(grouping = "id")) 
  

df_to_plot <- data.frame()
for(i in c(1:4)){
  df_to_plot <- bind_rows(df_to_plot, list_to_plot[[i]])
}
  
df_to_plot$new_id <-  factor(df_to_plot$new_id,levels = c("50","70","90","110","130","150","random"))
df_to_plot$new_id <-  factor(df_to_plot$new_id,levels = c("69","87","104","118","127","138","random"))

df_to_plot%>%
  ggplot(aes(x= new_id, y = value, fill = new_id)) +
  geom_bar(stat = "Identity") +
  theme_minimal() +
  facet_wrap(group~measure, scales="free_y", ncol = 2) +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=10)) + 
  labs(y = "Value", x = "Clustering") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  
}
```

```{r Enrichment: test BP function}
# Select clustering result to annotate
#clustering <- XXX$cluster

# Prepare clustering data
r <- 
   clustering %>%
    left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
    dplyr::select(-gene,-ensg_id)%>%
    na.omit()

clusters <- as.numeric(as.character(r$value))
cluster_list <- c()
num_clus <- tail(sort(r$value), n=1)
num_genes <- length(r$value)
for (i in c(0:num_clus)) {
  genes <- r %>%
    filter(value==i) %>%
    pull(entrez) %>%
    as.character()
  print(i)
  cluster_list[[as.character(i)]] <- genes
}
  
# Profile clusters (with universe)
ck_nouni <- compareCluster(geneClusters = cluster_list, 
                     fun = "enrichGO", 
                     OrgDb = org.Hs.eg.db, 
                     ont = "BP", 
                     #universe = as.character(r$entrez),
                     pAdjustMethod = "BH",
                     pvalueCutoff = 0.05, 
                    qvalueCutoff = 0.05)

cknouni_tibble <- as_tibble(ck_nouni) 

length(unique(as.numeric(ck_tibble$Cluster)))/num_clus
#0.969697 (no universe) - 0.9393939 (universe)

# Distribution of clusters
clustering %>% 
  group_by(value) %>%
  count

# Most significant term for each cluster
ck_tibble %>%
  group_by(Cluster) %>%
  summarize(min(p.adjust)) 

# Explore specific clusters
clus21 <- 
  ck_tibble %>%
  filter(Cluster == 21)

```

```{r Tissue expression per cluster (pheatmaps)}
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


#clustering <- louvain_multk[[1]][[1]]$cluster
#lustering <- max_clustering

data_tissue <-
  tmm_data %>%
  mutate(tissue = gsub('.{2}$', '', sample_ID)) %>%
  group_by(tissue,enssscg_id) %>%
  mutate(tissue_exp = mean(tmm)) %>%
  select(enssscg_id, tissue, tissue_exp) %>%
unique()

# clustering <-
#   clustering %>%
#   mutate(value = value+1)

clustering <- 
  all_partitions %>%
  filter(id == "zscore euclidean fasthclust 50") %>%
  select(gene, value) 

pdf("heatmaps_zscore_euclidean_fasthclust_50_2.pdf")
lapply(sort(unique(clustering$value)),
         function(i) {
           plot <- data_tissue %>%
             left_join(clustering, by = c("enssscg_id" = "gene")) %>%
            # mutate(value = value+1) %>%
             filter(value == i) %>%
             group_by(enssscg_id) %>%
             mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
             ungroup() %>%
             select(enssscg_id,tissue,scaled_value) %>%
             spread(key=tissue, value = scaled_value) %>%
             column_to_rownames("enssscg_id") %>% 
             pheatmap(clustering_method = "ward.D2", 
                      color = viridis(20, direction = -1, option = "B"),
                      show_rownames = F,border_color = NA, fontsize = 8)
         })
dev.off()
```

```{r Intra cluster correlations}

# load("data/processed/louvain_multk.Rdata")
# clustering <- louvain_multk[[1]][[2]]$cluster
# clustering <- max_clustering


clustering <- 
  all_partitions %>%
  filter(id == "zscore euclidean fasthclust 50") %>%
  select(gene, value) #%>%
  mutate(value = sample(value))


correlations <-
  tmm_data %>%
  mutate(log_tmm = log(tmm + 1))%>%
  select(-tmm) %>%
  left_join(clustering, by = c("enssscg_id" = "gene")) %>%
  group_by(value) %>%
  do({
    cluster <- unique(.$value)
    data <- .
    data <- data %>%
      select(-value) %>%
      spread(key = enssscg_id, value= log_tmm) %>%
      select(-sample_ID)

    cor_matrix <- cor(data)  %>%
      as_tibble(rownames = "gene1") %>%
      gather(gene2, cor, -gene1) %>%
      filter (gene1 < gene2) %>%
      mutate(cluster = cluster)
  })


correlations %>%
  ggplot(aes(x = as.factor(cluster), y = cor, fill = as.factor(cluster))) +
  geom_violin(draw_quantiles = 0.5) +
  theme_bw()

correlations %>% 
  group_by(cluster) %>%
  summarize(mean(cor)) %>%
  summarize(mean(`mean(cor)`)) 




# Calculate intra-cluster correlation for all partitions! output df (do)

intra_cluster_cor <- function (df) {
  df %>%
    group_by(value) %>%
    do({
      cluster <- unique(.$value)
      data <- .
      data <- data %>%
        select(-value) %>%
        spread(key = enssscg_id, value= log_tmm) %>%
        select(-sample_ID)
      
      cor_matrix <- cor(data)  %>%
        as_tibble(rownames = "gene1") %>%
        gather(gene2, cor, -gene1) %>%
        filter (gene1 < gene2) %>%
        mutate(cluster = cluster)
    })
}


all_correlations <-
  all_partitions %>%
  group_by(id) %>%
  do ({
    clustering <- . 
    id <- unique(clustering$id)
    correlations <-
      tmm_data %>%
      mutate(log_tmm = log(tmm + 1))%>%
      select(-tmm) %>%
      left_join(clustering %>% select(-id), by = c("enssscg_id" = "gene")) %>%
      intra_cluster_cor()
      
    average_correlation <- correlations %>%
      group_by(cluster) %>%
      summarize(mean(cor)) %>%
      summarize(mean(`mean(cor)`))
    data.frame(id = id, avg_cor = average_correlation)
  })


correlations <-
  tmm_data %>%
  mutate(log_tmm = log(tmm + 1))%>%
  select(-tmm) %>%
  left_join(clustering, by = c("enssscg_id" = "gene")) %>%
  group_by(value) %>%
  do({
    cluster <- unique(.$value)
    data <- .
    data <- data %>%
      select(-value) %>%
      spread(key = enssscg_id, value= log_tmm) %>%
      select(-sample_ID)

    cor_matrix <- cor(data)  %>%
      as_tibble(rownames = "gene1") %>%
      gather(gene2, cor, -gene1) %>%
      filter (gene1 < gene2) %>%
      mutate(cluster = cluster)
  })


correlations %>%
  ggplot(aes(x = as.factor(cluster), y = cor, fill = as.factor(cluster))) +
  geom_violin(draw_quantiles = 0.5) +
  theme_bw()

correlations %>% 
  group_by(cluster) %>%
  summarize(mean(cor)) %>%
  summarize(mean(`mean(cor)`)) 




cor_to_plot <- 
  all_correlations %>%
  mutate(scaling = strsplit(id, " ")[[1]][1],
    distance = strsplit(id, " ")[[1]][2],
    clustering = as.factor(strsplit(id, " ")[[1]][3] ),
  k = as.numeric(strsplit(id, " ")[[1]][4]),
  method = paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2],strsplit(id, " ")[[1]][3]))

l <- unique(sort(cor_to_plot$k))
cor_to_plot$k <- factor(cor_to_plot$k, levels = l)

test <- 
  cor_to_plot %>%
  group_by(method) %>%
  do({
    data <- .
    #m <- unique(data$method)
    l <- data %>% arrange(k) %>% pull (id)
    data$id <- factor(data$id, levels = l)
    data 
  })


test %>%
  ggplot(aes(x=id, y = mean..mean.cor..., fill = (k))) +
  geom_bar(stat = "identity")+
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~(clustering), scale="free") +
  scale_fill_viridis() +
  labs(y = "Value", x = "Clustering strategy")

# Proportionality instead of correlation
## propr

library(propr)

dat <- 
  tmm_data %>%
  mutate(log_tmm = log(tmm + 1))%>%
  select(-tmm) %>%
  left_join(clustering, by = c("enssscg_id" = "gene")) %>%
  filter(value==1)%>%
  select(-value) %>%
  spread(key = enssscg_id, value= log_tmm) %>%
  select(-sample_ID)

a <- propr(dat, metric = "rho")

b <-
  a@matrix  %>%
      as_tibble(rownames = "gene1") %>%
      gather(gene2, cor, -gene1) %>%
      filter (gene1 < gene2)

mean(b$cor)
```




###########

```{r test BHI}


load("data/processed/hclust_multk.Rdata")
clustering <- louvain_multk[[1]][[1]]$cluster

clustering_random <-
  clustering %>%
  mutate(value = sample(value))
r <- 
    clustering_random %>%
    select(gene,value) %>%
    left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(-gene)

  clusters <- as.numeric(as.character(r$value))

  bio_index <- 
    if(require("Biobase") && require("annotate") && require("GO.db") &&
       require("org.Hs.eg.db")) {
      BHI(clusters, annotation="org.Hs.eg.db", names=r$entrez, category="all") 
    }
  
  

results <-
  enrich(clustering = clustering,orthologs = ortholog_info)

library("writexl")

write_xlsx(results$enriched_terms,"enrichment.xlsx")
```

```{r Evaluations based on enrichments}

load("data/processed/louvain_multk.Rdata")

df_kmeans <- 
  to_df(kmeans_multk, add_random = T)

Sys.time()
enrichments_kmeans <- df_kmeans %>%
  map(~enrichments_df(df = .x, grouping = "id", orthologs = ortholog_info))


load("data/processed/enrichments_kmeans.Rdata")

enrichment_scores <- function(df,grouping) {
  
  df %>%
    group_by(.data[[grouping]]) %>%
    do ({
      data <- .
      min_significance <- data %>%
        mutate(log_qvalue = -log10(qvalue)) %>%
                                group_by(Cluster) %>% 
                                summarize(max(log_qvalue),na.rm = T) %>% 
                                pull(`max(log_qvalue)`) %>% na.omit() %>% mean() 
      annotability <- length(data %>%
               filter(qvalue < 0.05) %>% pull(Cluster) %>% sort() %>% unique())/
                           length(data %>% pull(Cluster) %>% sort() %>% unique())
      data.frame(id = unique(data$id), min_significance = min_significance, annotability = annotability)
        
    })
}
# 
# enrichments_louvain[[2]] %>% mutate(log_qvalue =  -log10(qvalue)) %>%
#                                 group_by(Cluster) %>% 
#                                 summarize(max(log_qvalue),na.rm = T) %>% 
#                                 pull(`max(log_qvalue)`) %>% na.omit() %>% mean() 
# to_plot_hclust <-
#   enrichment_scores(df = enrichments_kmeans[[1]], grouping = "id") %>%
#   gather(key = "measure", value = "value", -id) 


to_plot_kmeans_all <-
  map(enrichments_kmeans,
  ~enrichment_scores(df = .x, grouping = "id") %>%
  gather(key = "measure", value = "value", -id))

to_plot_kmeans_all <- bind_rows(to_plot_kmeans_all) %>%
  mutate(title = strsplit(id, " ")[[1]][3],
         subtitle  =paste(strsplit(id, " ")[[1]][1],strsplit(id, " ")[[1]][2]),
           k = strsplit(id, " ")[[1]][4])


l <- sort(unique(as.numeric(to_plot_kmeans_all$k)))
l[7] <- "random"

to_plot_kmeans_all$k <- factor(to_plot_kmeans_all$k,
                                            levels = l)

to_plot_kmeans_all %>%
  ggplot(aes(x= k, y = value, fill = k)) +
  geom_bar(stat = "Identity") +
  theme_minimal() +
  facet_wrap(subtitle~measure, scales="free_y", ncol = 2) +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=15)) + 
  labs(y = "Value", x = "Clustering") +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
  
  


enrichment_scores_kmeans <- to_plot_kmeans_all
save(enrichment_scores_kmeans, file= "enrichment_scores_kmeans.Rdata")














clustering <- louvain_multk[[1]][[1]]$cluster

clustering_random <-
  clustering %>%
  mutate(value = sample(value))

louvain_k <- list(louvain_multk[[1]][[1]], louvain_multk[[2]][[1]], louvain_multk[[3]][[1]],
               louvain_multk[[4]][[1]], louvain_multk[[5]][[1]], louvain_multk[[6]][[1]])

enrichments <- map(louvain_k, 
                   ~enrich(clustering = .x$cluster, orthologs = ortholog_info))

enrichment_random <- enrich(clustering = clustering_random, orthologs = ortholog_info)
# 
# enrichment_random$enriched_terms %>%
#   ggplot(aes(x=p.adjust)) +
#   geom_histogram()
#   geom_bar(stat = "identity")

enrichments_louvain <-
  map(enrichments,
      ~.x$enriched_terms)

enrichments_louvain[[7]] <- enrichment_random$enriched_terms



# ennrichments <- function(clusterings,orthologs) {
#   
#   clustering_random <-
#     clusterings[[1]][[1]]$cluster %>%
#     mutate(value = sample(value))
#   
#   clusterings_k <- list(clusterings[[1]][[1]], clusterings[[2]][[1]], clusterings[[3]][[1]],
#                     clusterings[[4]][[1]], clusterings[[5]][[1]], clusterings[[6]][[1]])
#   
#   enrichments <- map(clusterings_k, 
#                      ~enrich(clustering = .x$cluster, orthologs = ortholog_info))
#   
#   enrichment_random <- enrich(clustering = clustering_random, orthologs = ortholog_info)
# 
#   enrichment_terms <-
#     map(enrichments,
#         ~.x$enriched_terms)
#   
#   enrichment_terms[[7]] <- enrichment_random$enriched_terms
#   
#   return(enrichment_terms)
# 
# }



# Calculate scores
significance_overall <- map(enrichments_louvain, ~.x  %>% pull(qvalue) %>% mean())
significance_filtered <- map(enrichments_louvain, ~.x %>% filter(qvalue < 0.05) %>% pull(qvalue) %>% mean())
significance_overall_perclust <- map(enrichments_louvain, ~.x  %>% group_by(Cluster) 
                                      %>% summarize(mean(qvalue)) %>% pull(`mean(qvalue)`) %>%
                                       mean())
significance_filtered_perclust <- map(enrichments_louvain, ~.x %>% filter(qvalue < 0.05)  %>%
                                        group_by(Cluster) %>% summarize(mean(qvalue)) %>%
                                        pull(`mean(qvalue)`) %>% mean())

num_sig_terms_total <- map(enrichments_louvain, ~.x %>% filter(qvalue < 0.05) %>% pull(qvalue) %>% length)
num_sig_terms_per_enriched_cluster <- map(enrichments_louvain, ~.x %>% 
                                            filter(qvalue < 0.05) %>% group_by(Cluster) %>%
                                            summarize(length(qvalue)) %>% pull(`length(qvalue)`) %>%
                                            mean())

avg_minq_cluster <- map(enrichments_louvain, ~.x %>% group_by(Cluster) %>% 
                  summarize(min(qvalue)) %>% pull(`min(qvalue)`) %>% mean() )

avg_min_logq_cluster <- map(enrichments_louvain, ~.x %>% 
                              mutate(log_qvalue = -log10(qvalue)) %>%
                              group_by(Cluster) %>% 
                              summarize(max(log_qvalue)) %>% 
                              pull(`max(log_qvalue)`) %>% mean() )

perc_enrichmemt <- map(enrichments_louvain, ~
                       length(.x %>%filter(qvalue < 0.05)%>% pull(Cluster) %>% sort() %>% unique())/
                         length(.x %>% pull(Cluster) %>% sort() %>% unique()))

# Calculate BHI
calc_BHI <- function(c, genes) {
  r <- 
    c%>%
    select(gene,value) %>%
    left_join(genes, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(-gene) 
  
  clusters <- as.numeric(as.character(r$value))
  
  bio_index <- 
    if(require("Biobase") && require("annotate") && require("GO.db") &&
       require("org.Hs.eg.db")) {
      BHI(clusters, annotation="org.Hs.eg.db", names=r$entrez, category="BP") 
    }
  return(bio_index)
}
  

all_clusters <- list(louvain_multk[[1]][[1]]$cluster,
                     louvain_multk[[2]][[1]]$cluster, louvain_multk[[3]][[1]]$cluster, 
                     louvain_multk[[4]][[1]]$cluster, louvain_multk[[5]][[1]]$cluster, 
                     louvain_multk[[6]][[1]]$cluster, clustering_random)


BHI <- map(all_clusters, ~calc_BHI(.x, genes=ortholog_info))


clusterings <- c("louvain_66", "louvain_88", "louvain_104", "louvain_119", "louvain_131", "louvain_140", "random")

df_enrichment_louvain <- 
  data.frame(clusterings = clusterings, 
             significance_overall = unlist(significance_overall), 
             significance_filtered = unlist(significance_filtered),
             significance_overall_perclust = unlist(significance_overall_perclust),
             significance_filtered_perclust = unlist(significance_filtered_perclust),
             num_sig_terms_total = unlist(num_sig_terms_total), 
             num_sig_terms_per_enriched_cluster = unlist(num_sig_terms_per_enriched_cluster), 
             avg_minq_cluster = unlist(avg_minq_cluster),
             avg_min_logq_cluster = unlist(avg_min_logq_cluster),
             perc_enrichmemt = unlist(perc_enrichmemt)#,
             #BHI = unlist(BHI)
  ) 

df_enrichment_louvain <- 
  df_enrichment_louvain %>%
  gather(key = measure, value = value, significance_overall:perc_enrichmemt) 

df_enrichment_louvain$clusterings <- factor(df_enrichment_louvain$clusterings,
                                            levels = unique(df_enrichment_louvain$clusterings))

df_enrichment_louvain$measure <- factor(df_enrichment_louvain$measure,
                                        levels = unique(df_enrichment_louvain$measure))



df_enrichment_louvain %>%
  ggplot(aes(x= as.factor(clusterings), y = value, fill = clusterings)) +
  geom_bar(stat = "Identity") +
  theme_minimal() +
  facet_wrap(~measure, scales="free_y") +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=10)) + 
  labs(y = "Value", x = "Clustering") +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 


df_enrichment_louvain %>%
  spread(key = measure, value = value) %>%
  ggplot(aes(x=avg_min_logq_cluster, y = perc_enrichmemt, color = num_sig_terms_per_enriched_cluster)) +
  geom_point() +
  theme_minimal()

save(enrichments_louvain,
     df_enrichment_louvain, file ="data/processed/enrichment_louvain.Rdata")

enrichments_zscore_euclidean_louvain <- enrichments_louvain

names(enrichments_zscore_euclidean_louvain) <- c("zscore euclidean louvain 66",
                               "zscore euclidean louvain 88",
                               "zscore euclidean louvain 104",
                               "zscore euclidean louvain 119",
                               "zscore euclidean louvain 131",
                               "zscore euclidean louvain 140",
                               "zscore euclidean louvain random")

save(enrichments_zscore_euclidean_louvain, 
     file = "data/processed/enrichments_zscore_euclidean_louvain.Rdata")


#########
load("data/processed/test_biological.Rdata")
topList <- RankAggreg(geneLists, 5, N=700, seed=100, convIn=3)

```

```{r merge evals}
load("data/processed/evaluation_som.Rdata")

enrichment_scores_louvain <-
  enrichment_scores_louvain %>%
  spread(key = measure, value = value)


all_enrichment_scores <- 
  bind_rows(enrichment_scores_kmeans,
          enrichment_scores_medoids,
          enrichment_scores_hclust,
          enrichment_scores_louvain,
          enrichment_scores_leiden,
          enrichment_scores_som) %>%
  mutate(scale_dist_clust = paste(title,subtitle))


all_evaluations <- 
  bind_rows(all_kmeans,
            all_medoids,
            all_hclust,
            all_louvain,
            all_leiden,
            all_som)

df <- all_evaluations %>% 
  left_join(all_enrichment_scores, by = c("clustering_id" = "id")) %>%
  select(clustering_id, connectivity_index, avg_silhouette_width, scale_dist_clust, scale_dist, annotability, min_significance, time) %>%
  na.omit() %>%
  group_by(scale_dist_clust) %>%
  do ({
    data <- .
        id <- unique(data$scale_dist_clust)
    bestk <- find_bestk(data)
    head(bestk$bestk, 1)
  }) 

rank <- base::rank


## Explore visualizations

final_rank <- 
  df %>% 
  ungroup() %>% 
  select(clustering_id,connectivity_index, avg_silhouette_width,annotability, min_significance, time) %>%
  find_bestk()

final_rank <- final_rank$bestk

## Time plot
df %>%
  select(scale_dist_clust, time) %>%
  ggplot(aes(x = scale_dist_clust, y = time, fill = scale_dist_clust)) +
 geom_bar(stat = "identity", show.legend = F) +
      theme_minimal()+
    #scale_x_discrete(breaks = NULL) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    scale_fill_viridis(discrete = T, direction = -1) +
    labs(y = "Time (s)", x = "Clustering strategy") +
  theme( # remove the vertical grid lines
           panel.grid.major.x = element_blank() ,
           # explicitly set the horizontal lines (or they will disappear too)
           panel.grid.major.y = element_line( ) 
    )

final_rank %>%
  column_to_rownames("clustering_id") %>%
  mutate(rank_time = rank(-(time))) %>%
  select(rank_asw,rank_connectivity,rank_annotability, rank_min_significance, rank_time) %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))

```


```{r ClusterJudge - Mutual Information}

load("data/processed/hclust_multk.Rdata")

mutinfo_zscore <- function(clustering, orthologs, GO_terms) {
  c <- 
    clustering %>%
    left_join(orthologs, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(ensg_id,value)
  
  clusters <- c$value
  names(clusters) <- c$ensg_id
  
  #clusterJudge(clusters = clusters, entity.attribute = GO_terms)

   zscore_df <- clusterJudge_z_score(clusters = clusters, entity.attribute = GO_terms, 
                                     nmb.randomizations = 30, plot.saveRDS.file=NULL)
   zscore <- zscore_df[1,][[1]]
   return(zscore)
}



hclust_df <- hclust_multk %>% to_df()
Sys.time()
MI_zscores_hclust <- map(hclust_df,
                          ~ .x %>%
                            group_by(id) %>%
                            do ({
                              data <- .
                              zscore <- data %>% 
                                mutinfo_zscore(orthologs = ortholog_info, 
                                               GO_terms = GO[["0.01"]])
                              data.frame(id = unique(data$id), zscore = zscore)
                            }))



Sys.time()
save(MI_zscores_hclust, file="data/processed/MI_zscores_hclust.Rdata")






#GO terms
GOterms <- read_tsv("data/meta/E92_all_GOterms.txt")

GO_terms <- 
  GOterms %>%
  filter(`GO domain` == "biological_process") %>%
  select(ID = `Gene stable ID`, GOID = `GO term accession`) %>%
  na.omit() %>%
   filter(ID %in% genes)  %>%
  distinct()

GO <- consolidate_entity_attribute(
     entity.attribute = as.data.frame(GO_terms)
   , min.entities.per.attr =3
  # , mut.inf=mi.GO.Yeast   ### use precalculated mutual information
   , U.limit = c(0.01,0.001) ### calculate consolidated association for these uncertainty levels
   ) 

validate_association(GO[["0.01"]])


# Function
mutinfo_zscore <- function(clustering, orthologs, GO_terms) {
  c <- 
    clustering %>%
    left_join(orthologs, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(ensg_id,value)
  
  clusters <- c$value
  names(clusters) <- c$ensg_id
  
  #clusterJudge(clusters = clusters, entity.attribute = GO_terms)

   zscore_df <- clusterJudge_z_score(clusters = clusters, entity.attribute = GO_terms, 
                                     nmb.randomizations = 30, plot.saveRDS.file=NULL)
   zscore <- zscore_df[1,][[1]]
   return(zscore)
}



#What about looking at the mutual information not z score????

load("data/processed/louvain_multk.Rdata")


louvain_df <- louvain_multk %>% to_df()

MI_zscores_louvain <- map(louvain_df,
                          ~ .x %>%
                            group_by(id) %>%
                            do ({
                              data <- .
                              zscore <- data %>% 
                                mutinfo_zscore(orthologs = ortholog_info, GO_terms = GO[["0.01"]])
                              data.frame(id = unique(data$id), zscore = zscore)
                            }))


load("data/processed/GO.Rdata")
GO_mutinf <- attribute_mut_inf(entity.attribute = GO[["0.01"]])
str(GO_mutinf)

mutinfo <- function(clustering, orthologs, GO_terms) {
  c <- 
    clustering %>%
    left_join(orthologs, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(ensg_id,value)
  
  clusters <- c$value
  names(clusters) <- c$ensg_id
  
  M <- clusterJudge(clusters = clusters, entity.attribute = GO_terms)

  # zscore_df <- clusterJudge_z_score(clusters = clusters, entity.attribute = GO_terms, 
       #                              nmb.randomizations = 30, plot.saveRDS.file=NULL)
   #zscore <- zscore_df[1,][[1]]
   return(M[1,1])
}



library(ClusterJudge)
MI_louvain <- map(louvain_df,
                          ~ .x %>%
                            group_by(id) %>%
                            do ({
                              data <- .
                              MI <- data %>% 
                                mutinfo(orthologs = ortholog_info, GO_terms = GO[["0.01"]])
                              data.frame(id = unique(data$id), MI = MI)
                            }))


save(MI_louvain, file = "data/processed/MI_louvain.Rdata")
MI_zscore[[4]] %>%
  ggplot(aes(x= id, y = zscore, fill = id)) +
  geom_bar(stat = "Identity") +
  theme_minimal() +
#  facet_wrap(~measure, scales="free_y") +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=15)) + 
  labs(y = "Value", x = "Clustering") +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
  
  

# Tests individual
clusters_df <- louvain_multk[[1]][[1]]$cluster %>%
  left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
  na.omit() %>%
  dplyr::select(ensg_id,value) %>%
  mutate(value = value +1)

clusters <- as.integer(clusters_df$value)
names(clusters) <- clusters_df$ensg_id
str(clusters)

library(ClusterJudge)
load("data/processed/GO.Rdata")

MI <- clusterJudge(
    clusters = clusters, 
    entity.attribute=GO[["0.01"]]) 

MI[1,1]
#6.541085             



Z <- clusterJudge_z_score(
    clusters = clusters, 
    entity.attribute=GO[["0.01"]]) 



load("data/processed/MI_zscores_medoids.Rdata")
MI_zscores_louvain
MI_zscores_kmeans[[4]] %>%
  ggplot(aes(x= id, y = abs(zscore), fill = id)) +
  geom_bar(stat = "Identity") +
  theme_minimal() +
#  facet_wrap(~measure, scales="free_y") +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=15)) + 
  labs(y = "Value", x = "Clustering") +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 



# Calculate for all of them (merge all, facet clustering, method 6x4)







```



```{r Guy}
# Prepare data
load("data/processed/kmeans_multk.Rdata")

# kmeans_partitions <- 
#  list(kmeans_multk[[1]][[1]][-2],
#  kmeans_multk[[2]][[1]][-2],
#  kmeans_multk[[3]][[1]][-2],
#  kmeans_multk[[4]][[1]][-2],
#  kmeans_multk[[5]][[1]][-2],
#  kmeans_multk[[6]][[1]][-2])
save(kmeans_partitions, file="kmeans_partitions.Rdata")

change_ids <- function(clustering,orthologs) {
  m <- gregexpr('[0-9]+',clustering$id)
  k <- regmatches(clustering$id,m)[[1]]
  method <- paste("kmeans",k)
    
  result <- clustering$cluster %>%
    select(gene,value) %>%
    left_join(orthologs, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(-gene,-ensg_id) %>%
    rename(gene = entrez,
           cluster = value) %>%
    mutate(method = method) %>%
    relocate(method,gene,cluster)
  
 # result$method <- method
  return(result)
}

kmeans_partitions <- 
  list(change_ids(kmeans_multk[[1]][[1]], orthologs = ortholog_info),
     change_ids(kmeans_multk[[2]][[1]], orthologs = ortholog_info),
     change_ids(kmeans_multk[[3]][[1]], orthologs = ortholog_info),
     change_ids(kmeans_multk[[4]][[1]], orthologs = ortholog_info),
     change_ids(kmeans_multk[[5]][[1]], orthologs = ortholog_info),
     change_ids(kmeans_multk[[6]][[1]], orthologs = ortholog_info))

save(kmeans_partitions, file="kmeans_partitions.Rdata")

#His script

load("kmeans_partitions.Rdata")

m <- gregexpr('[0-9]+',kmeans_multk[[1]][[1]]$id)
  k <- regmatches(kmeans_multk[[1]][[1]]$id,m)[[1]]
  
  


## my test bp
enrich <- function (clustering,ontology_type="BP",orthologs) {
  cluster_list <- c()
  num_clus <- tail(sort(clustering$cluster), n=1)
  num_genes <- length(clustering$cluster)
  for (i in c(0:num_clus)) {
    genes <- clustering %>%
      filter(cluster==i) %>%
      pull(gene) %>%
      as.character()
    cluster_list[[as.character(i)]] <- genes
  }
  
  # Profile clusters (with universe)
  ck <- try(compareCluster(geneClusters = cluster_list, 
                           fun = "enrichGO", 
                           OrgDb = org.Hs.eg.db, 
                           ont = ontology_type, 
                           universe = as.character(clustering$gene),
                           pAdjustMethod = "BH",
                           pvalueCutoff = 0.05, 
                           qvalueCutoff = 0.05))
  
  ck_tibble <- as_tibble(ck) 
  
  bio_index <-
    length(unique(as.numeric(ck_tibble$Cluster)))/(num_clus)
}

bp <-
  map(kmeans_partitions,
    ~enrich(clustering = .x, orthologs = ortholog_info))

bp_random <- enrich(clustering = random_clustering, orthologs = ortholog_info)


```

```{r Retrieve partitions for ARI}

to_df <- function(multk, add_random = T) {
  
  lapply(c(1:4),
         function(i) 
           
           if(add_random == T) {
             random <- as.data.frame(multk[[1]][[i]]$cluster) %>%
               mutate(value = sample(value))
             df <- random %>%
               mutate(id = 
                        paste((gsub('[[:digit:]]+', '', 
                                    multk[[1]][[i]]$id)),"random", 
                              sep = ""))
             
             for (j in c(1:6)) {
               df <- bind_rows(df, as.data.frame(multk[[j]][[i]]$cluster) %>% 
                                 mutate(id = multk[[j]][[i]]$id))
             }
             return(df)
           }
         
         else {
           df <- data.frame()
           for (j in c(1:6)) {
             df <- bind_rows(df, as.data.frame(multk[[j]][[i]]$cluster) %>% 
                               mutate(id = multk[[j]][[i]]$id))
           }
           return(df)
         }
         
  )
}


load("data/processed/kmeans_multk.Rdata")
load("data/processed/medoids_multk.Rdata")
load("data/processed/hclust_multk.Rdata")
load("data/processed/louvain_multk.Rdata")
load("data/processed/leiden_multk.Rdata")
load("data/processed/som_multk.Rdata")

all_dfs <- map(list(kmeans_multk, medoids_multk, hclust_multk, 
                    louvain_multk, leidenn_multk, som_multk),
               ~to_df(.x))

merge_all_df <- function(list_df) {
  df <- data.frame()
  for (i in c(1:6)) {
    for (j in c(1:4)) {
      df <- bind_rows(df, list_df[[i]][[j]])
    }
  }
  
  return(df)
}

all_partitions <- merge_all_df(all_dfs)






best_partitions <-df$clustering_id

filtered <-
  all_partitions %>%
  filter(id %in% best_partitions)



clusterings <- unique(all_partitions$id)
#36 options
res_ari <- matrix(, nrow = 168, ncol = 168)
rownames(res_ari) <- c(clusterings)
colnames(res_ari) <- c(clusterings)

# Calculate ARI matrix
for (i in clusterings) {
  for (j in clusterings) {
    a <- ARI(all_partitions %>% filter(id == i) %>% pull(value), 
             all_partitions %>% filter(id == j) %>% pull(value))
    res_ari[i,j] <- a
    }
  }

save(res_ari, file="data/processed/res_ari.Rdata")
# Pheatmap for all scalings, distances and clustering methods
res_ari %>%
  pheatmap(fontsize = 4,
           clustering_method = "ward.D2")#, 
           color = viridis(20, direction = -1, option = "B"))

# Pheatmap for specific scalings

d <- c()
for (id in clusterings) {
  if (grepl("max",id) & !grepl("min",id)){d <- c(d,id)}
}

d <- c()
for (id in clusterings) {
  if (grepl("pearson",id)){d <- c(d,id)}
}

res_ari %>%
  as_tibble() %>%
  mutate(id = clusterings)%>%
  filter(id %in% d) %>%
  select(d) %>%
  as.matrix() %>%
  set_rownames(d)%>%
  pheatmap(fontsize = 10)


## Filtered


clusterings <- unique(filtered$id)
#36 options
res_ari <- matrix(, nrow = 24, ncol = 24)
rownames(res_ari) <- c(clusterings)
colnames(res_ari) <- c(clusterings)

# Calculate ARI matrix
for (i in clusterings) {
  for (j in clusterings) {
    a <- ARI(filtered %>% filter(id == i) %>% pull(value), 
             filtered %>% filter(id == j) %>% pull(value))
    res_ari[i,j] <- a
    }
  }

res_ari_filtered <- res_ari
save(res_ari_filtered, file="data/processed/res_ari_filtered.Rdata")
# Pheatmap for all scalings, distances and clustering methods
res_ari_filtered %>%
  pheatmap(fontsize = 10,
           clustering_method = "ward.D2")#, 
           color = viridis(20, direction = -1, option = "B"))


```

```{r umaps}

umap_res %>%
      mutate(enssscg_id = gene_names) %>%
      left_join(color_groups, by = c("enssscg_id" = "gene")) %>%
      ggplot(aes(V1,V2, color = as.factor(value))) +
      geom_point(show.legend = F) +
      theme_minimal() +
      scale_color_manual(values = pal) +
      coord_fixed()

umaps <- map(
  pcas,
  ~.x$scores %>%
    set_rownames(gene_names) %>%
    as.data.frame() %>%
    rownames_to_column("enssscg_id") %>%
    select((1:30)) %>%
    umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2, n_ep = 1000) 
)

save(umaps, file = "data/processed/umaps_zscore_max.Rdata")

clustering_umaps <- function(id, umaps, partitions) { #, 
  partition <-
    partitions %>% 
    filter(id == id)
  
  if(grepl("zscore",id)) {
    umap <- umaps[[1]]
  }
  else if(grepl("max",id)) {
    umap <- umaps[[2]]
  }

  umap_plot(umap_res = umap, color_by = "cluster", color_groups = partition) 
  
}



clustering_umaps(id = "max euclidean fasthclust 50", umaps = umaps, partitions = all_partitions)
clustering_umaps(id = "zscore euclidean fasthclust 50", umaps = umaps, partitions = all_partitions)

# Individual clusters
load("data/processed/umaps_distances.Rdata")


```

```{r}

```

```{r Linear model}
#Input df: ID, scaling, distance, clustering, k, score

to_model <- 
  df %>% 
  mutate(clustering_id = as.character(clustering_id)) %>%
  select(clustering_id, total_rank) %>%
  mutate(scaling = strsplit(clustering_id, " ")[[1]][1],
        distance = strsplit(clustering_id, " ")[[1]][2],
        clustering = strsplit(clustering_id, " ")[[1]][3],
        k = as.numeric(strsplit(clustering_id, " ")[[1]][4])) %>%
  select(-scale_dist_clust,-clustering_id)



model <- lm(total_rank ~ scaling + distance + clustering + k, data = to_model) # interaction

anova(model)

```

```{r ENRICHMENTS CLASSIFICATION}
gene_info

clus <- 
  all_partitions %>%
  filter(id == "zscore euclidean fasthclust 50") %>%
  select(gene, value)

clas <- 
  gene_info %>%
  select(gene = ensg_id, specificity_category)

# UMAP by classification
umaps[[1]] %>%
  rename(gene = features)%>%
  left_join(clus) %>%
  left_join(clas) %>%
      ggplot(aes(V1,V2, color = specificity_category )) +
      geom_point() +
      theme_minimal() +
   #   scale_color_manual(values = tissue_colors) +
      coord_fixed()


# UMAP by tissue
umaps[[1]] %>%
  rename(gene = features)%>%
  left_join(clus) %>%
  left_join(tissue_info, by = c("gene" = "ensg_id")) %>%
      ggplot(aes(V1,V2, color = enhanced_tissues)) +
      geom_point(show.legend = F) +
      theme_minimal() +
      scale_color_manual(values = tissue_colors) +
      coord_fixed()

## HT to tissue

tissue_info
clus %>%
  
# HT to category (enriched / enhanced)

clas
clus

# ----- Hypergeometric test ------

cluster_ht <- 
  clus %>%
  rename(cluster=value) %>%
  left_join(gene_info, by = c("gene" = "ensg_id")) %>% 
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues), 
                                   "not enriched", enhanced_tissues)) %>%
  separate_rows(enhanced_tissues, sep = ",") %>%
  
  {
    # q is the number of successes
    q <- 
      group_by(., enhanced_tissues, cluster, .drop = F) %>%
      summarise(q = n_distinct(gene))
    # k is the number of tries - i.e. the cluster size
    k <- 
      group_by(., cluster, .drop = F) %>%
      summarise(k = n_distinct(gene))
    
    # m is the number of possible successes
    m <- 
      group_by(., enhanced_tissues, .drop = F) %>%
      summarise(m = n_distinct(gene))
    
    # n is the population size - i.e. the number of genes
    n <- n_distinct(.$gene)
    
    q %>% 
      left_join(k) %>% 
      left_join(m) %>%
      # n is the population size - i.e. the number of genes
      mutate(n = n - m) 
    
  } 

cluster_ht_res <- 
  cluster_ht %>%
  group_by(enhanced_tissues, cluster) %>%
  # Testing the chance of getting the observed number or higher per cluster and tissue type
  summarise(p_value = phyper(q - 1, m, n, k, lower.tail = F)) %>%
  ungroup() %>%
  mutate(p_value = ifelse(p_value == 0, .Machine$double.xmin, p_value),
         adj_pval = p.adjust(p_value, method = "BH"),
         sign_level = case_when(adj_pval < 0.00000001 ~ 8,
                                adj_pval < 0.0000001 ~ 7,
                                adj_pval < 0.000001 ~ 6,
                                adj_pval < 0.00001 ~ 5,
                                adj_pval < 0.0001 ~ 4,
                                adj_pval < 0.001 ~ 3,
                                adj_pval < 0.01 ~ 2,
                                adj_pval < 0.05 ~ 1,
                                T ~ 0),
         log_p = -log10(adj_pval))

# a <- 
#   pig_gene_cluster_enrich_hyper %>% 
#   select(1:2, log_p) %>%
#   spread(enhanced_tissues, log_p) %>% 
#   column_to_rownames("cluster") %>% 
#   umap_calc(npcs = 2)
# 
# a %>%
#   as_tibble(rownames = "cluster") %>% 
#   ggplot(aes(V1, V2, label = cluster)) +
#   geom_text()
# 

cluster_ht_res %>% 
  select(enhanced_tissues, cluster, adj_pval, log_p) %>%
  group_by(enhanced_tissues, cluster) %>%
  mutate(log_p = min(log_p, 20)) %>%
  ungroup() %>% 
  mutate(log_p = ifelse(adj_pval < 0.05, 1, 0)) %>%
  select(-adj_pval) %>%
  spread(cluster, log_p, fill = 0) %>% 
  column_to_rownames("enhanced_tissues") %>%
  pheatmap(clustering_method = "ward.D2",
           #cutree_cols = n_distinct(similar_clusters$similar_cluster),
          # annotation_col = similar_clusters %>% 
          #   column_to_rownames("cluster"), 
           #filename = savepath("UMAP similar clusters heatmap.pdf"),
           width = 15, 
           height = 6)

##
cluster_ht_res %>% 
  select(enhanced_tissues, cluster, log_p) %>%
  ggplot(aes(x=cluster,y=enhanced_tissues,size = log_p)) +
  geom_point() +
  theme_bw()
  
  
  ###
  group_by(enhanced_tissues, cluster) %>%
  mutate(log_p = min(log_p, 20)) %>%
  ungroup() %>%
  mutate(log_p = ifelse(log_p < 3, 0, 1)) %>%
  spread(cluster, log_p) %>%
  column_to_rownames("enhanced_tissues") %>%
  pheatmap(clustering_method = "ward.D2")

```

```{r}
library(GOSemSim)

GO_data <- godata("org.Hs.eg.db", ont = "BP")

# sim < - mgeneSim(eg, semData = hsGO, measure = "Wang", drop =
# "IEA", combine = "BMA")
# > rownames(sim) < - genes[rownames(sim)]
# > colnames(sim) < - genes[colnames(sim)]

# Docu
# cluster1 <- c("835", "5261","241", "994")
# cluster2 <- c("307", "308", "317", "321", "506", "540", "378", "388", "396")
# clusterSim(cluster1, cluster2, semData=GO_data, measure="Wang")

clustering <- 
  all_partitions %>%
  filter(id == "zscore euclidean fasthclust 50") %>%
  select(gene, value) %>%
  left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
  select(entrez,value) %>%
  na.omit()



num_clusters <- length(unique(clustering$value)) 


# Manual matrix
# sim_df <- data.frame()
# 
# Sys.time()
# for (i in (1:num_clusters)) {
#   for(j in (1:num_clusters)) {
#     cluster_i <- clustering %>% filter(value == i) %>% pull(entrez) %>% as.character()
#     cluster_j <-  clustering %>% filter(value == j) %>% pull(entrez)  %>% as.character()
#     sim_value <- clusterSim(cluster_i, cluster_j, semData=GO_data, measure="Wang")
#     sim_df <- bind_rows(sim_df,data.frame(cluster1 = i, cluster2 = j, sim = sim_value)) 
#   }
# }
# 
# Sys.time()

  #11,05
    
sim_df%>%
  as_tibble(rownames = "cluster1") %>%    
  gather(cluster2, sim, -cluster1) %>%
  filter (cluster1 < cluster2) #%>%
 # mutate(cluster = cluster)



# mclsterSim

cluster_list <- c()
num_clus <- length(unique(clustering$value))
num_genes <- length(clustering$value)
for (i in c(0:num_clus)) {
  genes <- clustering %>%
    filter(value==i) %>%
    pull(entrez) %>%
    as.character()
  cluster_list[[as.character(i)]] <- genes
}

Sys.time()
sim_matrix <- mclusterSim(clusters= cluster_list, semData = GO_data, measure = "Wang") #1.30 h 
Sys.time()


sim_matrix %>%
  pheatmap(border_color = NA, 
            clustering_method = "ward.D2", 
           color = colorRampPalette(c("white","#00846b"))(50)) #cutree_rows = 5,
save(sim_matrix, file="data/processed/sim_matrix_zscore_euclidean_fasthclust_50.Rdata")
```


```{r umap?}
pig_gene_umap %>%
  group_by(merged_cluster) %>%
  mutate(mean_V1 = mean(V1),
         mean_V2 = mean(V2),
         dist = sqrt((V1 - mean_V1)^2 + (V2 - mean_V2)^2),
         n = n()) %>%
  ggplot(aes(V1, V2)) +
  geom_hex(aes(fill = 1),
           fill = "gray90",
           bins = plot_bins) +
  geom_text(data = . %>% 
              select(merged_cluster, mean_V1, mean_V2, n) %>% 
              distinct(),
            aes(mean_V1, mean_V2, label = merged_cluster),
            size = 2,
            alpha = 0.6,
            show.legend = F) +
  geom_encircle(aes(group = cluster),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                fill = "gray",
                color = "black",
                show.legend = F) +
  geom_encircle(data = . %>%
                  filter(grepl("^mg", merged_cluster_old)),
                aes(group = merged_cluster, 
                    color = merged_cluster,
                    fill = merged_cluster),
                
                alpha = 0.5, 
                s_shape=1, 
                expand=0,
                # color = "white",
                show.legend = F) +
  stripped_theme 
ggsave(savepath("UMAP merged clusters.pdf"), width = 6, height = 6)

```

```{r Max code for HT}

Max HT

# ----- Hypergeometric test ------

pig_gene_cluster_enrich_hyper_test_data <- 
  pig_gene_umap_temp2 %>%
  # head(1000) %>%
  select(enssscg_id, cluster) %>%
  left_join(pig_gene_classification) %>% 
  mutate(enhanced_tissues = ifelse(is.na(enhanced_tissues), 
                                   "not enriched", enhanced_tissues)) %>%
  separate_rows(enhanced_tissues, sep = ";") %>%
  
  {
    # q is the number of successes
    q <- 
      group_by(., enhanced_tissues, cluster, .drop = F) %>%
      summarise(q = n_distinct(enssscg_id))
    # k is the number of tries - i.e. the cluster size
    k <- 
      group_by(., cluster, .drop = F) %>%
      summarise(k = n_distinct(enssscg_id))
    
    # m is the number of possible successes
    m <- 
      group_by(., enhanced_tissues, .drop = F) %>%
      summarise(m = n_distinct(enssscg_id))
    
    # n is the population size - i.e. the number of genes
    n <- n_distinct(.$enssscg_id)
    
    q %>% 
      left_join(k) %>% 
      left_join(m) %>%
      # n is the population size - i.e. the number of genes
      mutate(n = n - m) 
    
  } 

pig_gene_cluster_enrich_hyper <- 
  pig_gene_cluster_enrich_hyper_test_data %>%
  group_by(enhanced_tissues, cluster) %>%
  # Testing the chance of getting the observed number or higher per cluster and tissue type
  summarise(p_value = phyper(q - 1, m, n, k, lower.tail = F)) %>%
  ungroup() %>%
  mutate(p_value = ifelse(p_value == 0, .Machine$double.xmin, p_value),
         adj_pval = p.adjust(p_value, method = "BH"),
         sign_level = case_when(adj_pval < 0.00000001 ~ 8,
                                adj_pval < 0.0000001 ~ 7,
                                adj_pval < 0.000001 ~ 6,
                                adj_pval < 0.00001 ~ 5,
                                adj_pval < 0.0001 ~ 4,
                                adj_pval < 0.001 ~ 3,
                                adj_pval < 0.01 ~ 2,
                                adj_pval < 0.05 ~ 1,
                                T ~ 0),
         log_p = -log10(adj_pval))

# a <- 
#   pig_gene_cluster_enrich_hyper %>% 
#   select(1:2, log_p) %>%
#   spread(enhanced_tissues, log_p) %>% 
#   column_to_rownames("cluster") %>% 
#   umap_calc(npcs = 2)
# 
# a %>%
#   as_tibble(rownames = "cluster") %>% 
#   ggplot(aes(V1, V2, label = cluster)) +
#   geom_text()
# 

pig_gene_cluster_enrich_hyper %>% 
  select(enhanced_tissues, cluster, adj_pval, log_p) %>%
  group_by(enhanced_tissues, cluster) %>%
  mutate(log_p = min(log_p, 20)) %>%
  ungroup() %>% 
  mutate(log_p = ifelse(adj_pval < 0.05, 1, 0)) %>%
  select(-adj_pval) %>%
  spread(cluster, log_p, fill = 0) %>% 
  column_to_rownames("enhanced_tissues") %>%
  pheatmap(clustering_method = "ward.D2",
           cutree_cols = n_distinct(similar_clusters$similar_cluster),
           annotation_col = similar_clusters %>% 
             column_to_rownames("cluster"), 
           filename = savepath("UMAP similar clusters heatmap.pdf"),
           width = 15, 
           height = 6)
```

# Extra
```{r # Comparison to Max clustering}
# Max clustering
max_clustering <- 
  read_excel("data/processed/UMAP cluster annotation 20201126.xlsx", 
             sheet = "Cluster genes") %>%
  select(`Cluster nr`, `Ensembl 92 ID`) %>%
  rename(value = `Cluster nr`, 
         gene = `Ensembl 92 ID`) %>%
  mutate(value = as.numeric(value)) 

enriched_terms <- enrich(clustering = max_clustering, ontology_type = "BP", orthologs = ortholog_info)




test <- clust(dist = d[[1]]$distance, genes = gene_names, 
                   m = "fasthclust", k=50, id = d[[1]]$id)


#Random
clustering <- louvain_multk[[1]][[1]]$cluster
louvain_random <- clustering
louvain_random <-
  louvain_random %>%
  mutate(value = sample(value)) 

enriched_terms_random <- enrich(clustering = louvain_random, ontology_type = "BP", orthologs = ortholog_info)


# Louvain 66
clustering <- louvain_multk[[1]][[1]]$cluster 

enriched_terms_louvain <- enrich(clustering = clustering, ontology_type = "BP", orthologs = ortholog_info)
```

```{r # Comparison to Max clustering}
biological_indices <- data.frame(
  id=c("Random", "Louvain clustering", "2D kmeans clustering"), 
  biological_index = c(bio_index_random, bio_index_louvain, bio_index_max))
  #,
  p_values = c(list(pvalues_random), list(pvalues_max), list(pvalues_louvain)))


bio <- biological_indices %>%
  ggplot(aes(x=id, y=biological_index, group =1, fill = as.factor(id))) +
  geom_bar(stat = "identity", show.legend = F)+
  theme_minimal() +
  #ggtitle("Hierarchical clustering") +
  #facet_wrap(scale_dist~measure, scales="free_y") +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=13)) + 
  labs(y = "BP enrichment", x = "Clustering strategy")


p_values <- bind_rows(data.frame(id = "Random", p_value = pvalues_random), 
                      data.frame(id = "2D kmeans clustering", 
                                           p_value = pvalues_max),
                      data.frame(id = "Louvain clustering", 
                                           p_value = pvalues_louvain))
pval<-
  p_values %>%
  #filter(p_value < 0.01) %>%
  ggplot(aes(x = id, y = -log10(p_value), fill = as.factor(id))) +
  geom_violin(show.legend = F, draw_quantiles = 0.5) +
  theme_minimal() +
  scale_fill_viridis(discrete = T) +
  scale_color_viridis(discrete = T)+
  theme(text = element_text(size=13)) + 
  labs(y = "p values", x = "Clustering strategy")

sort(pvalues_louvain)

bio + pval


intersect <- intersect(ck_max_tibble$Description, ck_louvain_tibble$Description)



  scale_fill_viridis(option="B", direction = -1) 

  p_values2 <- bind_rows(data.frame(clust_id = "Initial clustering", 
                                    p_value = ck_max_tibble$p.adjust,
                                    term = ck_max_tibble$Description),
                         data.frame(clust_id = "Alternative clustering", 
                                    p_value = ck_louvain_tibble$p.adjust,
                                    term = ck_louvain_tibble$Description)) %>%
  filter(term %in% intersect) %>%
  group_by(clust_id, term) %>%
  top_n(1, p_value)
  #select(clust_id, term, max(p_value))
    #3215

  #unique(p_values2$term)

  p_values2 %>%
    spread(key= clust_id, value= p_value) %>%
    ggplot(aes(x=-log10(`Alternative clustering`), y = -log10(`Initial clustering`))) +
    geom_point() +
    #scale_fill_viridis(discrete = T) +
    #scale_color_viridis(continuous = T)+
    theme(text = element_text(size=10)) + 
    #labs(y = "p values", x = "Clustering strategy") +
    geom_abline(slope = 1)+
    theme_minimal()

```

```{r Max UMAP}
load("data/processed/max_umap.Rdata")

max_clustering %>%
  left_join(ortholog_info) %>%
  select(value, enssscg_id) %>%
  left_join(pig_gene_umap_cl %>% as_tibble(rownames = "enssscg_id")) %>%
    #filter(value==i) %>%
    ggplot(aes(V1,V2, color= as.factor(value))) +
   # geom_point(data = t, aes(V1,V2), inherit.aes = F, 
    #           size = 0.5, color= "grey") +
    geom_point(show.legend = F, size =0.5) +
    theme_minimal() +
    coord_fixed() +
  scale_color_manual(values = p2)

p2 <- sample(p2)
p2 <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(83)

louvain_random %>%
  left_join(ortholog_info) %>%
  select(value, enssscg_id) %>%
  left_join(pig_gene_umap_cl %>% as_tibble(rownames = "enssscg_id")) %>%
    #filter(value==i) %>%
    ggplot(aes(V1,V2, color= as.factor(value))) +
   # geom_point(data = t, aes(V1,V2), inherit.aes = F, 
    #           size = 0.5, color= "grey") +
    geom_point(show.legend = F, size =0.5) +
    theme_bw() +
    coord_fixed() +
  scale_color_manual(values = p2) +
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))



```

```{r UMAP plot by cluster (PCA / PLS)}
clustering <- louvain_multk[[1]][[1]]$cluster

# PCA gene matrix to UMAP (Alternative 1)
gene_matrix <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) 

umap_louv <- gene_matrix %>%
  select((1:30)) %>%
  rownames_to_column("gene") %>%
  umap_calc(n_neigh = 15, row_names = "gene")

# PLS gene matrix to UMAP (Alternative 2)
gene_matrix <- scaled[[1]]$scaled %>% column_to_rownames("enssscg_id") 

pls_louv <- plsda(X= gene_matrix, Y = clustering$value, ncomp = 30)

umap_pls_louv <- pls_louv$variates$X%>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  umap_calc(n_neigh = 15, row_names = "gene")


## COMMON PART
# Data to plot, create "grey" UMAP background
to_plot <- 
  umap_louv %>%
  left_join(clustering, by = c("features" = "gene"))
t <- to_plot %>%
  select(-value)

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(67)

#Rename clusters 0->1
to_plot <- to_plot %>%
  mutate(value = value+1)

# Highlight each cluster
plots <-
  lapply(sort(unique(to_plot$value)),
function(i) {
  to_plot%>%
    filter(value==i) %>%
    ggplot(aes(V1,V2)) +
    geom_point(data = t, aes(V1,V2), inherit.aes = F, 
               size = 0.5, color= "grey") +
    geom_point(show.legend = F, size =0.5, color= p[i]) +
    theme_minimal() +
    coord_fixed() +
    ggtitle(paste("cluster",i))
})

pdf("umaps.pdf")
plots
dev.off()


# Plot 1 cluster
to_plot%>%
    filter(value==57) %>%
    ggplot(aes(V1,V2)) +
    geom_point(data = t, aes(V1,V2), inherit.aes = F, 
               size = 0.5, color= "grey") +
    geom_point(show.legend = F, size =0.5, color= p[57]) +
    theme_minimal() +
    coord_fixed()

# Distribution of genes in clusters
distr <- to_plot %>%
  group_by(value) %>%
  count()

to_plot%>%
    ggplot(aes(V1,V2, color= as.factor(value))) +
    geom_point(show.legend = F, size =0.4) +
    theme_minimal() +
    coord_fixed() +
  scale_color_manual(values = p)

```
